<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2025-05-26">

<title>Gviz – 2025 MBMM</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-de84f8d6bb715db06a919283c2d1e787.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-de2f664832f57b2c83139acc48228f9e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">2025 MBMM</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../pages/index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../IntroGR/pages/Intro_to_R.html"> 
<span class="menu-text">Intro_to_R</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../IntroGR/pages/GRanges_gtf.html"> 
<span class="menu-text">Working with genomic coordinates</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_principles.html"> 
<span class="menu-text">ChIP-seq principles</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_analysis_part1.html"> 
<span class="menu-text">ChIP-seq analysis Part1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_analysis_part2.html"> 
<span class="menu-text">ChIP-seq analysis Part2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../Gviz/pages/Genomic_overviews_with_Gviz.html" aria-current="page"> 
<span class="menu-text">Genomic visualization</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../PBMC3k/pages/scRNA-seq_PBMC3k.html"> 
<span class="menu-text">scRNA-seq_PBMC3k</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/paganilab"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/LabPagani"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#concepts" id="toc-concepts" class="nav-link" data-scroll-target="#concepts">Concepts</a></li>
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries">Load libraries</a></li>
  <li><a href="#annotation-track-annotationtrack" id="toc-annotation-track-annotationtrack" class="nav-link" data-scroll-target="#annotation-track-annotationtrack">Annotation Track: AnnotationTrack()</a>
  <ul class="collapse">
  <li><a href="#import-data-on-cpg-islands" id="toc-import-data-on-cpg-islands" class="nav-link" data-scroll-target="#import-data-on-cpg-islands">Import data on CpG islands</a></li>
  <li><a href="#inspect-granges-object" id="toc-inspect-granges-object" class="nav-link" data-scroll-target="#inspect-granges-object">Inspect GRanges object</a></li>
  <li><a href="#retrieve-information-of-genome-and-sequence" id="toc-retrieve-information-of-genome-and-sequence" class="nav-link" data-scroll-target="#retrieve-information-of-genome-and-sequence">Retrieve information of genome and sequence</a></li>
  <li><a href="#generate-annotation-track" id="toc-generate-annotation-track" class="nav-link" data-scroll-target="#generate-annotation-track">Generate Annotation Track</a></li>
  <li><a href="#plot" id="toc-plot" class="nav-link" data-scroll-target="#plot">Plot</a></li>
  </ul></li>
  <li><a href="#genome-axis-track-genomeaxistrack" id="toc-genome-axis-track-genomeaxistrack" class="nav-link" data-scroll-target="#genome-axis-track-genomeaxistrack">Genome Axis Track: GenomeAxisTrack()</a>
  <ul class="collapse">
  <li><a href="#generate-genomeaxistrack" id="toc-generate-genomeaxistrack" class="nav-link" data-scroll-target="#generate-genomeaxistrack">Generate GenomeAxisTrack()</a></li>
  <li><a href="#plot-1" id="toc-plot-1" class="nav-link" data-scroll-target="#plot-1">Plot</a></li>
  </ul></li>
  <li><a href="#chromosome-ideogram-track-ideogramtrack" id="toc-chromosome-ideogram-track-ideogramtrack" class="nav-link" data-scroll-target="#chromosome-ideogram-track-ideogramtrack">Chromosome Ideogram Track: IdeogramTrack()</a></li>
  <li><a href="#lets-explore-the-environment-and-objects-so-far" id="toc-lets-explore-the-environment-and-objects-so-far" class="nav-link" data-scroll-target="#lets-explore-the-environment-and-objects-so-far">Let’s explore the environment and objects so far!</a>
  <ul class="collapse">
  <li><a href="#exercise-1" id="toc-exercise-1" class="nav-link" data-scroll-target="#exercise-1">Exercise 1</a></li>
  </ul></li>
  <li><a href="#gene-model-information-generegiontrack" id="toc-gene-model-information-generegiontrack" class="nav-link" data-scroll-target="#gene-model-information-generegiontrack">Gene Model information: GeneRegionTrack()</a>
  <ul class="collapse">
  <li><a href="#load-gene-model-data-from-a-stored-data.frame" id="toc-load-gene-model-data-from-a-stored-data.frame" class="nav-link" data-scroll-target="#load-gene-model-data-from-a-stored-data.frame">Load gene model data from a stored data.frame</a></li>
  <li><a href="#generate-track-and-plot" id="toc-generate-track-and-plot" class="nav-link" data-scroll-target="#generate-track-and-plot">Generate Track and plot</a></li>
  <li><a href="#zoom-in-to-the-region" id="toc-zoom-in-to-the-region" class="nav-link" data-scroll-target="#zoom-in-to-the-region">Zoom in to the region</a></li>
  </ul></li>
  <li><a href="#display-the-genomic-sequence-at-a-given-position-sequencetrack" id="toc-display-the-genomic-sequence-at-a-given-position-sequencetrack" class="nav-link" data-scroll-target="#display-the-genomic-sequence-at-a-given-position-sequencetrack">Display the Genomic Sequence at a given position: SequenceTrack()</a></li>
  <li><a href="#incoorporate-data-datatrack" id="toc-incoorporate-data-datatrack" class="nav-link" data-scroll-target="#incoorporate-data-datatrack">Incoorporate Data: DataTrack()</a></li>
  <li><a href="#global-control-parameters" id="toc-global-control-parameters" class="nav-link" data-scroll-target="#global-control-parameters">Global Control Parameters</a>
  <ul class="collapse">
  <li><a href="#specify-parameters-during-track-generation" id="toc-specify-parameters-during-track-generation" class="nav-link" data-scroll-target="#specify-parameters-during-track-generation">1. Specify parameters during Track generation</a></li>
  <li><a href="#modify-parameters-with-displaypars" id="toc-modify-parameters-with-displaypars" class="nav-link" data-scroll-target="#modify-parameters-with-displaypars">2. Modify parameters with <em>displayPars()</em></a></li>
  <li><a href="#set-display-parameters-for-a-single-plotting-operation" id="toc-set-display-parameters-for-a-single-plotting-operation" class="nav-link" data-scroll-target="#set-display-parameters-for-a-single-plotting-operation">3. Set display parameters for a single plotting operation</a></li>
  <li><a href="#exercise-2" id="toc-exercise-2" class="nav-link" data-scroll-target="#exercise-2">Exercise 2</a></li>
  </ul></li>
  <li><a href="#parameters-for-data-track" id="toc-parameters-for-data-track" class="nav-link" data-scroll-target="#parameters-for-data-track">Parameters for Data Track</a>
  <ul class="collapse">
  <li><a href="#input" id="toc-input" class="nav-link" data-scroll-target="#input">Input</a></li>
  <li><a href="#import-data-on-multiple-replicates-for-control-and-treated-samples" id="toc-import-data-on-multiple-replicates-for-control-and-treated-samples" class="nav-link" data-scroll-target="#import-data-on-multiple-replicates-for-control-and-treated-samples">Import data on multiple replicates for control and treated samples</a></li>
  <li><a href="#integrate-information-of-data-grouping" id="toc-integrate-information-of-data-grouping" class="nav-link" data-scroll-target="#integrate-information-of-data-grouping">Integrate information of data grouping</a></li>
  <li><a href="#exercise-3" id="toc-exercise-3" class="nav-link" data-scroll-target="#exercise-3">Exercise 3</a></li>
  </ul></li>
  <li><a href="#parameters-for-alignment-track" id="toc-parameters-for-alignment-track" class="nav-link" data-scroll-target="#parameters-for-alignment-track">Parameters for Alignment Track</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Gviz</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 26, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p><strong>PLotting data and annotation information along genomic coordinates</strong></p>
<blockquote class="blockquote">
<p>Based on User Guide by Florian Hahne<sup>1</sup> and Robert Ivánek<sup>2,3</sup> <sup>1</sup> Novartis Institute for Biomedical Research, Basel, Switzerland <sup>2</sup> Department of Biomedicine, University of Basel, Basel, Switzerland <sup>3</sup> Swiss Institute of Bioinformatics, Basel, Switzerland</p>
</blockquote>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>The Gviz package (Hahne and Ivanek 2016) aims to close this gap by providing <strong>a structured visualization framework to plot any type of data along genomic coordinates.</strong> It is loosely based on the GenomeGraphs package by Steffen Durinck and James Bullard, however the complete class hierarchy as well as all the plotting methods have been restructured in order to increase performance and flexibility. All plotting is done using the grid graphics system, and several <strong>specialized annotation classes allow to integrate publicly available genomic annotation data from sources like UCSC or ENSEMBL.</strong></p>
</section>
<section id="concepts" class="level2">
<h2 class="anchored" data-anchor-id="concepts">Concepts</h2>
<ol type="1">
<li>Individual types of genomic features or data are represented by separate tracks.</li>
</ol>
<ul>
<li>AnnotationTrack</li>
<li>GenomeAxisTrack</li>
<li>IdeogramTrack</li>
<li>GeneRegionTrack</li>
<li>SequenceTrack</li>
<li>DataTrack</li>
</ul>
<ol start="2" type="1">
<li><p>When combining multiple objects, the individual tracks will always share the same genomic coordinate system.</p></li>
<li><p>Tracks may contain information for multiple chromosomes, however most of this is hidden except for the currently active chromosome during a given plotting operation.</p></li>
<li><p>In Bioconductor, genomic features consisting of start and stop coordinates are most often represented as run-length encoded vectors, e.g.&nbsp;IRanges and GRanges classes.</p></li>
</ol>
<blockquote class="blockquote">
<p><strong>Bioconductor</strong>: open source software for Bioinformatics</p>
</blockquote>
</section>
<section id="load-libraries" class="level2">
<h2 class="anchored" data-anchor-id="load-libraries">Load libraries</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># suppressWarnings({</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   if (!require("BiocManager"))</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#     install.packages("BiocManager", "Gviz")</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># })</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>, <span class="st">"Gviz"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install("dplyr")</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install("Gviz")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressWarnings</span>({</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(BiocManager)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(GenomicRanges)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(Gviz)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(biomaRt)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(rtracklayer)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="annotation-track-annotationtrack" class="level2">
<h2 class="anchored" data-anchor-id="annotation-track-annotationtrack">Annotation Track: AnnotationTrack()</h2>
<p><strong>Generate Track for CpG Islands and plot</strong></p>
<section id="import-data-on-cpg-islands" class="level3">
<h3 class="anchored" data-anchor-id="import-data-on-cpg-islands">Import data on CpG islands</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(cpgIslands)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(cpgIslands)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GRanges"
attr(,"package")
[1] "GenomicRanges"</code></pre>
</div>
</div>
</section>
<section id="inspect-granges-object" class="level3">
<h3 class="anchored" data-anchor-id="inspect-granges-object">Inspect GRanges object</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>cpgIslands</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 0 metadata columns:
       seqnames            ranges strand
          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
   [1]     chr7 26549019-26550183      *
   [2]     chr7 26564119-26564500      *
   [3]     chr7 26585667-26586158      *
   [4]     chr7 26591772-26593309      *
   [5]     chr7 26594192-26594570      *
   [6]     chr7 26623835-26624150      *
   [7]     chr7 26659284-26660352      *
   [8]     chr7 26721294-26721717      *
   [9]     chr7 26821518-26823297      *
  [10]     chr7 26991322-26991841      *
  -------
  seqinfo: 1 sequence from hg19 genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="retrieve-information-of-genome-and-sequence" class="level3">
<h3 class="anchored" data-anchor-id="retrieve-information-of-genome-and-sequence">Retrieve information of genome and sequence</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>chr <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(<span class="fu">seqnames</span>(cpgIslands)))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>gen <span class="ot">&lt;-</span> <span class="st">"hg19"</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>chr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "chr7"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>gen</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "hg19"</code></pre>
</div>
</div>
</section>
<section id="generate-annotation-track" class="level3">
<h3 class="anchored" data-anchor-id="generate-annotation-track">Generate Annotation Track</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>atrack <span class="ot">&lt;-</span> <span class="fu">AnnotationTrack</span>(cpgIslands, <span class="at">name =</span> <span class="st">"CpG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot" class="level3">
<h3 class="anchored" data-anchor-id="plot">Plot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(atrack)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/plot-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="genome-axis-track-genomeaxistrack" class="level2">
<h2 class="anchored" data-anchor-id="genome-axis-track-genomeaxistrack">Genome Axis Track: GenomeAxisTrack()</h2>
<ul>
<li>Display the genomic coordinates to provide reference.</li>
<li>A GenomeAxisTrack object is always relative to the other tracks that are plotted.</li>
<li>No need to know in advance about a particular genomic location when constructing the object. The displayed coordinates will be determined from the context, e.g., from the from and to arguments of the plotTracks function, or, when plotted together with other track objects, from their genomic locations.</li>
</ul>
<section id="generate-genomeaxistrack" class="level3">
<h3 class="anchored" data-anchor-id="generate-genomeaxistrack">Generate GenomeAxisTrack()</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>gtrack <span class="ot">&lt;-</span> <span class="fu">GenomeAxisTrack</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-1" class="level3">
<h3 class="anchored" data-anchor-id="plot-1">Plot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="chromosome-ideogram-track-ideogramtrack" class="level2">
<h2 class="anchored" data-anchor-id="chromosome-ideogram-track-ideogramtrack">Chromosome Ideogram Track: IdeogramTrack()</h2>
<ul>
<li>An ideogram is a simplified visual representation of a chromosome, with the different chromosomal staining bands indicated by color, and the centromere (if present) indicated by the shape.</li>
<li>We can fetch the chromosome ideogram information directly from UCSC (need an established internet connection for this to work)</li>
<li>Need to provide the information about a valid UCSC genome and chromosome.</li>
<li>A plotted track is always defined for exactly one chromosome on a particular genome.</li>
<li>The chromosomal data necessary to draw the ideogram is not part of the Gviz package itself, instead it is downloaded from an online source (UCSC).</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#options(Gviz.ucscUrl="http://genome-euro.ucsc.edu/cgi-bin/")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check if you are able to generate the ideogram track. If yes, you can add this track every time you call the function <code>plotTracks()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#itrack &lt;- IdeogramTrack(genome = gen, chromosome = chr)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plotTracks(list(itrack, gtrack, atrack))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="lets-explore-the-environment-and-objects-so-far" class="level2">
<h2 class="anchored" data-anchor-id="lets-explore-the-environment-and-objects-so-far">Let’s explore the environment and objects so far!</h2>
<ol type="1">
<li>Check the environment</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "atrack"     "chr"        "cpgIslands" "gen"        "gtrack"    </code></pre>
</div>
</div>
<ol start="2" type="1">
<li>Check the objects</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>atrack</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>AnnotationTrack 'CpG'
| genome: hg19
| active chromosome: chr7
| annotation features: 10</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>atrack<span class="sc">@</span>range</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 4 metadata columns:
       seqnames            ranges strand |     feature       group          id
          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt;
   [1]     chr7 26549019-26550183      * |     unknown           1     unknown
   [2]     chr7 26564119-26564500      * |     unknown           2   unknown.1
   [3]     chr7 26585667-26586158      * |     unknown           3   unknown.2
   [4]     chr7 26591772-26593309      * |     unknown           4   unknown.3
   [5]     chr7 26594192-26594570      * |     unknown           5   unknown.4
   [6]     chr7 26623835-26624150      * |     unknown           6   unknown.5
   [7]     chr7 26659284-26660352      * |     unknown           7   unknown.6
   [8]     chr7 26721294-26721717      * |     unknown           8   unknown.7
   [9]     chr7 26821518-26823297      * |     unknown           9   unknown.8
  [10]     chr7 26991322-26991841      * |     unknown          10   unknown.9
         density
       &lt;numeric&gt;
   [1]         1
   [2]         1
   [3]         1
   [4]         1
   [5]         1
   [6]         1
   [7]         1
   [8]         1
   [9]         1
  [10]         1
  -------
  seqinfo: 1 sequence from hg19 genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#itrack@bandTable</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#itrack@range</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="exercise-1" class="level3">
<h3 class="anchored" data-anchor-id="exercise-1">Exercise 1</h3>
<p>Change the order of the objects plotted</p>
</section>
</section>
<section id="gene-model-information-generegiontrack" class="level2">
<h2 class="anchored" data-anchor-id="gene-model-information-generegiontrack">Gene Model information: GeneRegionTrack()</h2>
<ul>
<li>Display information on gene structure from local source or online resources like UCSC and Ensembl</li>
</ul>
<section id="load-gene-model-data-from-a-stored-data.frame" class="level3">
<h3 class="anchored" data-anchor-id="load-gene-model-data-from-a-stored-data.frame">Load gene model data from a stored data.frame</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(geneModels)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(geneModels,<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  chromosome    start      end width strand feature            gene
1       chr7 26591441 26591829   389      + lincRNA ENSG00000233760
2       chr7 26591458 26591829   372      + lincRNA ENSG00000233760
             exon      transcript     symbol
1 ENSE00001693369 ENST00000420912 AC004947.2
2 ENSE00001596777 ENST00000457000 AC004947.2</code></pre>
</div>
</div>
</section>
<section id="generate-track-and-plot" class="level3">
<h3 class="anchored" data-anchor-id="generate-track-and-plot">Generate Track and plot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>grtrack <span class="ot">&lt;-</span> <span class="fu">GeneRegionTrack</span>(geneModels, <span class="at">genome =</span> gen,</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">chromosome =</span> chr, <span class="at">name =</span> <span class="st">"Gene Model"</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co">#plotTracks(list(itrack, gtrack, atrack, grtrack))</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>( gtrack, atrack, grtrack))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(geneModels,<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  chromosome    start      end width strand feature            gene
1       chr7 26591441 26591829   389      + lincRNA ENSG00000233760
2       chr7 26591458 26591829   372      + lincRNA ENSG00000233760
             exon      transcript     symbol
1 ENSE00001693369 ENST00000420912 AC004947.2
2 ENSE00001596777 ENST00000457000 AC004947.2</code></pre>
</div>
</div>
</section>
<section id="zoom-in-to-the-region" class="level3">
<h3 class="anchored" data-anchor-id="zoom-in-to-the-region">Zoom in to the region</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack, grtrack),</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">from =</span> <span class="dv">26700000</span>, <span class="at">to =</span> <span class="dv">26750000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>The layout of the gene model track has changed depending on the zoom level. This is a feature of the Gviz package, which automatically tries to find the optimal visualization settings to make best use of the available space.</p>
</blockquote>
</section>
</section>
<section id="display-the-genomic-sequence-at-a-given-position-sequencetrack" class="level2">
<h2 class="anchored" data-anchor-id="display-the-genomic-sequence-at-a-given-position-sequencetrack">Display the Genomic Sequence at a given position: SequenceTrack()</h2>
<ul>
<li>The track class SequenceTrack can draw the necessary sequence information from one of the BSgenome packages.</li>
<li>Create track for a specific chromosome</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenome.Hsapiens.UCSC.hg19)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>strack <span class="ot">&lt;-</span> <span class="fu">SequenceTrack</span>(Hsapiens, <span class="at">chromosome =</span> chr)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack, grtrack, strack), </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">from =</span> <span class="dv">26591822</span>, <span class="at">to =</span> <span class="dv">26591852</span>, <span class="at">cex =</span> <span class="fl">0.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="incoorporate-data-datatrack" class="level2">
<h2 class="anchored" data-anchor-id="incoorporate-data-datatrack">Incoorporate Data: DataTrack()</h2>
<ul>
<li>Annotate genomic coordinate plots with real numeric data</li>
<li>Different visualization options for these tracks, from dot plots to histograms to box-and-whisker plots.</li>
<li>The individual rows in a numeric matrix are considered to be different data groups or samples, and the columns are the intervals in the genomic coordinates.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">255</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>lim <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">26700000</span>, <span class="dv">26750000</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">c</span>(lim[<span class="dv">1</span>], </span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">sample</span>(<span class="fu">seq</span>(<span class="at">from =</span> lim[<span class="dv">1</span>], <span class="at">to =</span> lim[<span class="dv">2</span>]), <span class="dv">99</span>), </span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>                 lim[<span class="dv">2</span>]))</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">100</span>, <span class="at">min =</span> <span class="sc">-</span><span class="dv">10</span>, <span class="at">max =</span> <span class="dv">10</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>dtrack <span class="ot">&lt;-</span> <span class="fu">DataTrack</span>(<span class="at">data =</span> dat, <span class="at">start =</span> coords[<span class="sc">-</span><span class="fu">length</span>(coords)],</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">end =</span> coords[<span class="sc">-</span><span class="dv">1</span>], <span class="at">chromosome =</span> chr, <span class="at">genome =</span> gen, </span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>                    <span class="at">name =</span> <span class="st">"Uniform"</span>)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack, grtrack, dtrack), </span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">from =</span> lim[<span class="dv">1</span>], <span class="at">to =</span> lim[<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<p><strong>Change plot type to <em>histogram</em></strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack, grtrack, dtrack), </span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">from =</span> lim[<span class="dv">1</span>], <span class="at">to =</span> lim[<span class="dv">2</span>], <span class="at">type =</span> <span class="st">"histogram"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p><strong>Case Use of the DataTracks</strong> + Display the coverage of NGS reads along a chromosome + Show the measurement values of mapped probes from a microarray experiment</p>
</blockquote>
</section>
<section id="global-control-parameters" class="level2">
<h2 class="anchored" data-anchor-id="global-control-parameters">Global Control Parameters</h2>
<section id="specify-parameters-during-track-generation" class="level3">
<h3 class="anchored" data-anchor-id="specify-parameters-during-track-generation">1. Specify parameters during Track generation</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>grtrack <span class="ot">&lt;-</span> <span class="fu">GeneRegionTrack</span>(geneModels, <span class="at">genome =</span> gen, <span class="at">chromosome =</span> chr, </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">name =</span> <span class="st">"Gene Model"</span>,</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">transcriptAnnotation =</span> <span class="st">"symbol"</span>,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">background.title =</span> <span class="st">"brown"</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack, grtrack))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="modify-parameters-with-displaypars" class="level3">
<h3 class="anchored" data-anchor-id="modify-parameters-with-displaypars">2. Modify parameters with <em>displayPars()</em></h3>
<p>Display track parameters</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">displayPars</span>(grtrack))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Modify parameters for background colors in the GeneRegion track</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">displayPars</span>(grtrack) <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">background.panel =</span> <span class="st">"#FFFEDB"</span>, <span class="at">background.title =</span> <span class="st">"darkorange"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack, grtrack))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="set-display-parameters-for-a-single-plotting-operation" class="level3">
<h3 class="anchored" data-anchor-id="set-display-parameters-for-a-single-plotting-operation">3. Set display parameters for a single plotting operation</h3>
<ul>
<li>By passing in additional arguments to the plotTracks function, instead of the permanent setting in the track object</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">list</span>(gtrack, atrack, grtrack), </span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">background.panel =</span> <span class="st">"#FFFEDB"</span>, <span class="at">background.title =</span> <span class="st">"darkblue"</span>, <span class="at">fontcolor.title=</span><span class="st">'white'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p><strong>Tip!</strong><br>
The availableDisplayPars function prints out the available parameters for a class as well as their default values in a list-like structure.</p>
</blockquote>
</section>
<section id="exercise-2" class="level3">
<h3 class="anchored" data-anchor-id="exercise-2">Exercise 2</h3>
<p>Change the color of the font title: - 1. In the GeneRegion track - 2. In all tracks</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="parameters-for-data-track" class="level2">
<h2 class="anchored" data-anchor-id="parameters-for-data-track">Parameters for Data Track</h2>
<ul>
<li>To display one or several numeric values that are associated to a particular genomic coordinate range</li>
<li>When we need to incorporate sample group information</li>
</ul>
<section id="input" class="level3">
<h3 class="anchored" data-anchor-id="input">Input</h3>
<ul>
<li>a set of ranges, either in the form of an IRanges or GRanges object</li>
<li>a numeric vector (or matrix) of the same length as the number of ranges</li>
</ul>
</section>
<section id="import-data-on-multiple-replicates-for-control-and-treated-samples" class="level3">
<h3 class="anchored" data-anchor-id="import-data-on-multiple-replicates-for-control-and-treated-samples">Import data on multiple replicates for control and treated samples</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(twoGroups)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>twoGroups</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 25 ranges and 6 metadata columns:
       seqnames    ranges strand |   control control.1 control.2   treated
          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
   [1]     chrX      1-30      * | -8.961260  -7.65790  9.879565  -5.84376
   [2]     chrX     42-71      * | -4.211471   4.68826 -1.053306   1.03084
   [3]     chrX    84-113      * |  2.287112   8.01327 -7.121998  -4.46718
   [4]     chrX   125-154      * |  9.209838  -6.23243  8.596822  -6.32041
   [5]     chrX   167-196      * |  0.406842  -7.05442 -0.551974   9.36363
   ...      ...       ...    ... .       ...       ...       ...       ...
  [21]     chrX   833-862      * |  2.712424  7.538538   6.56281  -3.42396
  [22]     chrX   875-904      * |  0.727043 -0.950022  -7.63786 -28.96335
  [23]     chrX   916-945      * |  7.386881  4.817002   3.64004  -7.59088
  [24]     chrX   958-987      * | -5.541625 -2.231180  -4.65625   7.04730
  [25]     chrX 1000-1029      * | -1.139128  2.861125   8.17754  -5.34925
        treated.1 treated.2
        &lt;numeric&gt; &lt;numeric&gt;
   [1]    9.71353  9.993286
   [2]   -6.77430  0.593712
   [3]   -4.05888  8.053195
   [4]   -1.56806  3.511461
   [5]   -4.88057  1.552880
   ...        ...       ...
  [21] -25.015912 -14.86389
  [22] -15.066067 -26.36181
  [23]   5.146919  -5.80432
  [24]  -7.555974  -8.46894
  [25]   0.874496  -6.32031
  -------
  seqinfo: 1 sequence from hg19 genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>dTrack <span class="ot">&lt;-</span> <span class="fu">DataTrack</span>(twoGroups, <span class="at">name =</span> <span class="st">"uniform"</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(dTrack)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(dTrack, <span class="at">type =</span> <span class="fu">c</span>(<span class="st">"boxplot"</span>, <span class="st">"a"</span>, <span class="st">"g"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">mcols</span>(twoGroups))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "control"   "control.1" "control.2" "treated"   "treated.1" "treated.2"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(dTrack, <span class="at">type =</span> <span class="fu">c</span>(<span class="st">"heatmap"</span>), <span class="at">showSampleNames =</span> <span class="cn">TRUE</span>, </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">cex.sampleNames =</span> <span class="fl">0.6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="integrate-information-of-data-grouping" class="level3">
<h3 class="anchored" data-anchor-id="integrate-information-of-data-grouping">Integrate information of data grouping</h3>
<ul>
<li>Samples are grouped together based on a factor variable (e.g.&nbsp;control, treated).</li>
<li>The grouping is reflected in the layout of the respective track types.</li>
<li>Grouping is not supported for all plotting types</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>( dTrack,  </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">groups =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"treated"</span>), <span class="at">each =</span> <span class="dv">3</span>), </span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">type =</span> <span class="fu">c</span>(<span class="st">"a"</span>, <span class="st">"p"</span>, <span class="st">"confint"</span>), </span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">from=</span><span class="dv">1</span>, <span class="at">to=</span><span class="dv">1030</span>, <span class="at">genome=</span><span class="st">'hg19'</span>, <span class="at">chromosome=</span><span class="st">'chrX'</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>           )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../pics/Data_Type_Formats_TheGvizUserGuide.png" class="img-fluid figure-img"></p>
<figcaption>Data Type Formats</figcaption>
</figure>
</div>
</section>
<section id="exercise-3" class="level3">
<h3 class="anchored" data-anchor-id="exercise-3">Exercise 3</h3>
<p>Display tracks for chromosome ideogram and genome coordinates</p>
<blockquote class="blockquote">
<p>Tip: Check the chromosome!</p>
</blockquote>
</section>
</section>
<section id="parameters-for-alignment-track" class="level2">
<h2 class="anchored" data-anchor-id="parameters-for-alignment-track">Parameters for Alignment Track</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>afrom <span class="ot">&lt;-</span> <span class="dv">44945200</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>ato <span class="ot">&lt;-</span> <span class="dv">44947200</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>alTrack <span class="ot">&lt;-</span> <span class="fu">AlignmentsTrack</span>(</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="at">package =</span> <span class="st">"Gviz"</span>, <span class="st">"extdata"</span>, <span class="st">"snps.bam"</span>), <span class="at">isPaired =</span> <span class="cn">TRUE</span>)</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(alTrack, <span class="at">chromosome =</span> <span class="st">"chr21"</span>, <span class="at">from =</span> afrom, <span class="at">to =</span> ato)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(alTrack, strack, <span class="at">chromosome =</span> <span class="st">"chr21"</span>, <span class="at">from =</span> afrom, <span class="at">to =</span> ato)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">c</span>(alTrack, strack), <span class="at">chromosome =</span> <span class="st">"chr21"</span>,</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">from =</span> <span class="dv">44946590</span>, <span class="at">to =</span> <span class="dv">44946660</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(<span class="fu">c</span>(alTrack, strack), <span class="at">chromosome =</span> <span class="st">"chr21"</span>, <span class="at">from =</span> <span class="dv">44946590</span>, </span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">to =</span> <span class="dv">44946660</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">min.height =</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Genomic_overviews_with_Gviz_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>