<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2025-05-26">

<title>GRanges – 2025 MBMM</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-de84f8d6bb715db06a919283c2d1e787.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-de2f664832f57b2c83139acc48228f9e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">2025 MBMM</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../pages/index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../IntroGR/pages/Intro_to_R.html"> 
<span class="menu-text">Intro_to_R</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../IntroGR/pages/GRanges_gtf.html" aria-current="page"> 
<span class="menu-text">Working with genomic coordinates</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_principles.html"> 
<span class="menu-text">ChIP-seq principles</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_analysis_part1.html"> 
<span class="menu-text">ChIP-seq analysis Part1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_analysis_part2.html"> 
<span class="menu-text">ChIP-seq analysis Part2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../Gviz/pages/Genomic_overviews_with_Gviz.html"> 
<span class="menu-text">Genomic visualization</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../PBMC3k/pages/scRNA-seq_PBMC3k.html"> 
<span class="menu-text">scRNA-seq_PBMC3k</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/paganilab"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/LabPagani"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#install-and-load" id="toc-install-and-load" class="nav-link active" data-scroll-target="#install-and-load">Install and load</a></li>
  <li><a href="#the-genomic-range-object" id="toc-the-genomic-range-object" class="nav-link" data-scroll-target="#the-genomic-range-object">The Genomic Range Object</a>
  <ul class="collapse">
  <li><a href="#extract-idividual-components-of-the-gr-object" id="toc-extract-idividual-components-of-the-gr-object" class="nav-link" data-scroll-target="#extract-idividual-components-of-the-gr-object">Extract idividual components of the GR object</a></li>
  <li><a href="#basic-interval-operations" id="toc-basic-interval-operations" class="nav-link" data-scroll-target="#basic-interval-operations">Basic interval operations</a></li>
  <li><a href="#extracting-annotations" id="toc-extracting-annotations" class="nav-link" data-scroll-target="#extracting-annotations">Extracting annotations</a></li>
  <li><a href="#subset-granges-objects" id="toc-subset-granges-objects" class="nav-link" data-scroll-target="#subset-granges-objects">Subset GRanges objects</a></li>
  </ul></li>
  <li><a href="#plyranges" id="toc-plyranges" class="nav-link" data-scroll-target="#plyranges"><code>plyranges</code></a></li>
  <li><a href="#import-gtf-for-homo-sapiens" id="toc-import-gtf-for-homo-sapiens" class="nav-link" data-scroll-target="#import-gtf-for-homo-sapiens">Import GTF for Homo sapiens</a>
  <ul class="collapse">
  <li><a href="#inspect-the-gtf-file" id="toc-inspect-the-gtf-file" class="nav-link" data-scroll-target="#inspect-the-gtf-file">Inspect the GTF file</a></li>
  <li><a href="#drawing-plots-of-metadata-information" id="toc-drawing-plots-of-metadata-information" class="nav-link" data-scroll-target="#drawing-plots-of-metadata-information">Drawing plots of metadata information</a></li>
  </ul></li>
  <li><a href="#defining-the-tss-of-transcripts" id="toc-defining-the-tss-of-transcripts" class="nav-link" data-scroll-target="#defining-the-tss-of-transcripts">Defining the TSS of transcripts</a></li>
  <li><a href="#intersect-with-tf-bound-regions" id="toc-intersect-with-tf-bound-regions" class="nav-link" data-scroll-target="#intersect-with-tf-bound-regions">Intersect with TF-bound regions</a>
  <ul class="collapse">
  <li><a href="#retrieve-data-from-geo" id="toc-retrieve-data-from-geo" class="nav-link" data-scroll-target="#retrieve-data-from-geo">Retrieve data from GEO</a></li>
  </ul></li>
  <li><a href="#convert-seqnames-of-batf-gr-object" id="toc-convert-seqnames-of-batf-gr-object" class="nav-link" data-scroll-target="#convert-seqnames-of-batf-gr-object">Convert seqnames of BATF GR object</a></li>
  <li><a href="#range-based-transformations-are-grouped-in-2-categories" id="toc-range-based-transformations-are-grouped-in-2-categories" class="nav-link" data-scroll-target="#range-based-transformations-are-grouped-in-2-categories">Range-based transformations are grouped in 2 categories:</a>
  <ul class="collapse">
  <li><a href="#intra-range-functions" id="toc-intra-range-functions" class="nav-link" data-scroll-target="#intra-range-functions">Intra-range functions</a></li>
  <li><a href="#inter-range-functions" id="toc-inter-range-functions" class="nav-link" data-scroll-target="#inter-range-functions">Inter-range functions</a></li>
  </ul></li>
  <li><a href="#interval-set-operations-for-granges-objects" id="toc-interval-set-operations-for-granges-objects" class="nav-link" data-scroll-target="#interval-set-operations-for-granges-objects">Interval set operations for GRanges objects</a></li>
  <li><a href="#splitting-and-combining-granges-objects" id="toc-splitting-and-combining-granges-objects" class="nav-link" data-scroll-target="#splitting-and-combining-granges-objects">Splitting and combining GRanges objects</a></li>
  <li><a href="#online-resources-and-databases-for-omics-data" id="toc-online-resources-and-databases-for-omics-data" class="nav-link" data-scroll-target="#online-resources-and-databases-for-omics-data">Online Resources and Databases for omics data</a>
  <ul class="collapse">
  <li><a href="#poor-mans-guide-to-omics" id="toc-poor-mans-guide-to-omics" class="nav-link" data-scroll-target="#poor-mans-guide-to-omics">Poor man’s guide to omics</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">GRanges</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 26, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p><strong>GRanges</strong></p>
<p>The GenomicRanges package serves as the foundation for representing genomic locations. We will use the operations implemented in the GenomicRanges and related packages to run jobs typically performed in the context of a high throughput sequence analysis.</p>
<p><strong>Objectives</strong></p>
<ul>
<li>Introduction to GRanges class</li>
<li>Explore main functions for handling GRanges objects</li>
<li>Work with GRanges and dplyr</li>
<li>Import and inspect a GTF file</li>
<li>Define TSS genomic positions</li>
<li>Intersect TSS windows with TF-bound genomic intervals</li>
</ul>
<section id="install-and-load" class="level2">
<h2 class="anchored" data-anchor-id="install-and-load">Install and load</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"GenomicRanges"</span>))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"GenomicRanges"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curl)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#library(plyranges)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="the-genomic-range-object" class="level2">
<h2 class="anchored" data-anchor-id="the-genomic-range-object">The Genomic Range Object</h2>
<p>The GRanges class defines a set of genomic ranges, each characterized by a single start and end position on the genome. It is commonly used to store the locations of genomic features such as contiguous binding sites, transcripts, and exons. GRanges objects can be created using the GRanges constructor function. For instance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqnames =</span> <span class="fu">Rle</span>(<span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span>, <span class="st">"chr3"</span>), <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">4</span>)),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">101</span><span class="sc">:</span><span class="dv">110</span>, <span class="at">end =</span> <span class="dv">111</span><span class="sc">:</span><span class="dv">120</span>, <span class="at">names =</span> <span class="fu">head</span>(letters, <span class="dv">10</span>)),</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">strand =</span> <span class="fu">Rle</span>(<span class="fu">strand</span>(<span class="fu">c</span>(<span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span>)), <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">2</span>)), <span class="co"># strand --&gt; without double quotes</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">score =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">GC =</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="at">length=</span><span class="dv">10</span>))</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  d     chr2   104-114      * |         4  0.666667
  e     chr1   105-115      * |         5  0.555556
  f     chr1   106-116      + |         6  0.444444
  g     chr3   107-117      + |         7  0.333333
  h     chr3   108-118      + |         8  0.222222
  i     chr3   109-119      - |         9  0.111111
  j     chr3   110-120      - |        10  0.000000
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>The output of the GRanges displays information in two sections separated by a <code>|</code> symbol. Genomic coordinates (including <code>seqnames</code>, <code>ranges</code>, <code>strand</code>, <code>names</code>) appear on the left-hand side, while metadata columns (annotations) are shown on the right. In the above example, the metadata includes <code>score</code> and <code>GC</code> content, although the metadata section of a GRanges object can store a wide variety of information.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GRanges"
attr(,"package")
[1] "GenomicRanges"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10</code></pre>
</div>
</div>
<p>An Rle (run-length-encoded) vector is a specific representation of a vector. For more information have a look at: <a href="https://kasperdanielhansen.github.io/genbioconductor/html/GenomicRanges_Rle.html" class="uri">https://kasperdanielhansen.github.io/genbioconductor/html/GenomicRanges_Rle.html</a>. The IRanges function defines interval ranges.</p>
<section id="extract-idividual-components-of-the-gr-object" class="level3">
<h3 class="anchored" data-anchor-id="extract-idividual-components-of-the-gr-object">Extract idividual components of the GR object</h3>
<p>The genomic coordinate components of a GRanges object can be accessed using the <code>seqnames</code>, <code>names</code>, and <code>strand</code> functions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract chromosome names</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">seqnames</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>factor-Rle of length 10 with 4 runs
  Lengths:    1    3    2    4
  Values : chr1 chr2 chr1 chr3
Levels(3): chr1 chr2 chr3</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the strand</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">strand</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>factor-Rle of length 10 with 5 runs
  Lengths: 1 2 2 3 2
  Values : - + * + -
Levels(3): + - *</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the names</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the start position</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">start</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 101 102 103 104 105 106 107 108 109 110</code></pre>
</div>
</div>
</section>
<section id="basic-interval-operations" class="level3">
<h3 class="anchored" data-anchor-id="basic-interval-operations">Basic interval operations</h3>
<p>The genomic ranges can be extracted without corresponding metadata with <code>ranges</code> or <code>granges</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ranges</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IRanges object with 10 ranges and 0 metadata columns:
        start       end     width
    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  a       101       111        11
  b       102       112        11
  c       103       113        11
  d       104       114        11
  e       105       115        11
  f       106       116        11
  g       107       117        11
  h       108       118        11
  i       109       119        11
  j       110       120        11</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract genomic ranges without annotations</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">granges</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 0 metadata columns:
    seqnames    ranges strand
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  a     chr1   101-111      -
  b     chr2   102-112      +
  c     chr2   103-113      +
  d     chr2   104-114      *
  e     chr1   105-115      *
  f     chr1   106-116      +
  g     chr3   107-117      +
  h     chr3   108-118      +
  i     chr3   109-119      -
  j     chr3   110-120      -
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>What is the length of the genomic regions? One might want to plot the peak length distribution for called peaks. GRanges can provide this in one function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the width</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">width</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 11 11 11 11 11 11 11 11 11 11</code></pre>
</div>
</div>
</section>
<section id="extracting-annotations" class="level3">
<h3 class="anchored" data-anchor-id="extracting-annotations">Extracting annotations</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract annotations</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 10 rows and 2 columns
      score        GC
  &lt;integer&gt; &lt;numeric&gt;
a         1  1.000000
b         2  0.888889
c         3  0.777778
d         4  0.666667
e         5  0.555556
f         6  0.444444
g         7  0.333333
h         8  0.222222
i         9  0.111111
j        10  0.000000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(gr)<span class="sc">$</span>GC</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 1.0000000 0.8888889 0.7777778 0.6666667 0.5555556 0.4444444 0.3333333
 [8] 0.2222222 0.1111111 0.0000000</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reverse GRanges</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co">#rev(gr)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="subset-granges-objects" class="level3">
<h3 class="anchored" data-anchor-id="subset-granges-objects">Subset GRanges objects</h3>
<p>Standard subsetting methods similar to R vectors and lists can be used to subset GRanges objects.</p>
<p>Subset specific intervals</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>gr[<span class="fu">c</span>(<span class="st">'a'</span>,<span class="st">'d'</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  d     chr2   104-114      * |         4  0.666667
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>gr[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Select and filter specific metadata</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>gr[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>, <span class="fu">c</span>(<span class="st">'GC'</span>,<span class="st">'score'</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 2 metadata columns:
    seqnames    ranges strand |        GC     score
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt;
  b     chr2   102-112      + |  0.888889         2
  c     chr2   103-113      + |  0.777778         3
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>gr[gr<span class="sc">$</span>GC<span class="sc">&gt;</span><span class="fl">0.7</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
</section>
<section id="plyranges" class="level2">
<h2 class="anchored" data-anchor-id="plyranges"><code>plyranges</code></h2>
<p>Can we use dplyr functions with GRanges objects?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>gr <span class="sc">%&gt;%</span> <span class="fu">select</span>(GC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In general, <code>dplyr</code> functions do not directly apply on a GRanges object, because GRanges is a special Bioconductor object and not a standard data frame or tibble. One option is to convert a GRanges object to a tibble or data.frame using as.data.frame() or as_tibble(), and then use dplyr functions like select(), filter(), mutate(), etc. Following manipulation, it can be converted back to a GRanges object if needed.</p>
<p><code>makeGRangesFromDataFrame</code> uses as input a data-frame-like object and automatically finds the columns that describe genomic ranges. It then returns them as a GRanges object. A number of arguments control the behaviour of the function and modify the output GRange object.</p>
<p>What is the purpose of the <code>keep.extra.columns</code> option?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a dataframe, filter with dplyr, and re-convert to a GRanges object</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>gr_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(gr)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>gr_df <span class="ot">&lt;-</span> gr_df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(GC<span class="sc">&gt;</span><span class="fl">0.7</span>) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>strand)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert data.frame to GRanges object </span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="fu">makeGRangesFromDataFrame</span>(gr_df, <span class="at">keep.extra.columns =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      * |         1  1.000000
  b     chr2   102-112      * |         2  0.888889
  c     chr2   103-113      * |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Alternatively, one can use <code>plyranges</code>, a Bioconductor package that extends dplyr-style syntax directly to GRanges objects.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install("plyranges")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyranges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'plyranges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    between, n, n_distinct</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:IRanges':

    slice</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:stats':

    filter</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>gr <span class="sc">%&gt;%</span> <span class="fu">filter</span>(GC<span class="sc">&gt;</span><span class="fl">0.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert GRanges object to a dataframe </span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>gr <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  seqnames start end width strand score        GC
a     chr1   101 111    11      -     1 1.0000000
b     chr2   102 112    11      +     2 0.8888889
c     chr2   103 113    11      +     3 0.7777778
d     chr2   104 114    11      *     4 0.6666667
e     chr1   105 115    11      *     5 0.5555556
f     chr1   106 116    11      +     6 0.4444444</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>Not all <code>dplyr</code> functions are compatible with GRanges object. In some cases you need to convert to a tibble and then back to a GR object.</p>
</blockquote>
</section>
<section id="import-gtf-for-homo-sapiens" class="level2">
<h2 class="anchored" data-anchor-id="import-gtf-for-homo-sapiens">Import GTF for Homo sapiens</h2>
<p>GRanges provides a flexible framework for importing, inspecting, and processing GTF (Gene Transfer Format) files. Here, we will retrieve a GTF file containing genomic annotations for Homo sapiens from the Ensembl FTP server. We will use the <code>import</code> function from the rtracklayer package to fetch the gtf file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import library</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the FTP URL</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>ftp_url <span class="ot">=</span> <span class="st">"ftp://ftp.ensembl.org/pub/release-106/gtf/homo_sapiens/Homo_sapiens.GRCh38.106.chr.gtf.gz"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import gtf as GRange object</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="dv">1000</span>)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(ftp_url)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View GRanges </span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>gtf <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 6 ranges and 22 metadata columns:
      seqnames          ranges strand |         source        type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;factor&gt;    &lt;factor&gt; &lt;numeric&gt;
  [1]        1 1211340-1214153      - | ensembl_havana gene               NA
  [2]        1 1211340-1214153      - | ensembl_havana transcript         NA
  [3]        1 1213983-1214153      - | ensembl_havana exon               NA
  [4]        1 1213983-1214127      - | ensembl_havana CDS                NA
  [5]        1 1214125-1214127      - | ensembl_havana start_codon        NA
  [6]        1 1213663-1213785      - | ensembl_havana exon               NA
          phase         gene_id gene_version   gene_name    gene_source
      &lt;integer&gt;     &lt;character&gt;  &lt;character&gt; &lt;character&gt;    &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000186827           11     TNFRSF4 ensembl_havana
  [2]      &lt;NA&gt; ENSG00000186827           11     TNFRSF4 ensembl_havana
  [3]      &lt;NA&gt; ENSG00000186827           11     TNFRSF4 ensembl_havana
  [4]         0 ENSG00000186827           11     TNFRSF4 ensembl_havana
  [5]         0 ENSG00000186827           11     TNFRSF4 ensembl_havana
  [6]      &lt;NA&gt; ENSG00000186827           11     TNFRSF4 ensembl_havana
        gene_biotype   transcript_id transcript_version transcript_name
         &lt;character&gt;     &lt;character&gt;        &lt;character&gt;     &lt;character&gt;
  [1] protein_coding            &lt;NA&gt;               &lt;NA&gt;            &lt;NA&gt;
  [2] protein_coding ENST00000379236                  4     TNFRSF4-201
  [3] protein_coding ENST00000379236                  4     TNFRSF4-201
  [4] protein_coding ENST00000379236                  4     TNFRSF4-201
  [5] protein_coding ENST00000379236                  4     TNFRSF4-201
  [6] protein_coding ENST00000379236                  4     TNFRSF4-201
      transcript_source transcript_biotype         tag     ccds_id
            &lt;character&gt;        &lt;character&gt; &lt;character&gt; &lt;character&gt;
  [1]              &lt;NA&gt;               &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [2]    ensembl_havana     protein_coding       basic      CCDS11
  [3]    ensembl_havana     protein_coding       basic      CCDS11
  [4]    ensembl_havana     protein_coding       basic      CCDS11
  [5]    ensembl_havana     protein_coding       basic      CCDS11
  [6]    ensembl_havana     protein_coding       basic      CCDS11
      transcript_support_level exon_number         exon_id exon_version
                   &lt;character&gt; &lt;character&gt;     &lt;character&gt;  &lt;character&gt;
  [1]                     &lt;NA&gt;        &lt;NA&gt;            &lt;NA&gt;         &lt;NA&gt;
  [2]   1 (assigned to previ..        &lt;NA&gt;            &lt;NA&gt;         &lt;NA&gt;
  [3]   1 (assigned to previ..           1 ENSE00001832731            2
  [4]   1 (assigned to previ..           1            &lt;NA&gt;         &lt;NA&gt;
  [5]   1 (assigned to previ..           1            &lt;NA&gt;         &lt;NA&gt;
  [6]   1 (assigned to previ..           2 ENSE00001333068            1
           protein_id protein_version
          &lt;character&gt;     &lt;character&gt;
  [1]            &lt;NA&gt;            &lt;NA&gt;
  [2]            &lt;NA&gt;            &lt;NA&gt;
  [3]            &lt;NA&gt;            &lt;NA&gt;
  [4] ENSP00000368538               3
  [5]            &lt;NA&gt;            &lt;NA&gt;
  [6]            &lt;NA&gt;            &lt;NA&gt;
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<section id="inspect-the-gtf-file" class="level3">
<h3 class="anchored" data-anchor-id="inspect-the-gtf-file">Inspect the GTF file</h3>
<p>Check the metadata information contained in the GTF file</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the metadata information</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(gtf) <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "source"                   "type"                    
 [3] "score"                    "phase"                   
 [5] "gene_id"                  "gene_version"            
 [7] "gene_name"                "gene_source"             
 [9] "gene_biotype"             "transcript_id"           
[11] "transcript_version"       "transcript_name"         
[13] "transcript_source"        "transcript_biotype"      
[15] "tag"                      "ccds_id"                 
[17] "transcript_support_level" "exon_number"             
[19] "exon_id"                  "exon_version"            
[21] "protein_id"               "protein_version"         </code></pre>
</div>
</div>
<p>Notice that GRanges was able to properly tabulate the information contained in the 9th column of the GTF file. This wouldn’t be as efficient using the <code>read.table()</code> function.</p>
<p>What are the different type of intervals?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the levels in the `type` column</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="co">#gtf %&gt;% count(type)</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>gtf <span class="sc">%&gt;%</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(type) <span class="sc">%&gt;%</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n =</span> <span class="fu">n</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 9 rows and 2 columns
             type         n
         &lt;factor&gt; &lt;integer&gt;
1 gene                61498
2 transcript         246452
3 exon              1572337
4 CDS                851032
5 start_codon         95255
6 stop_codon          88853
7 five_prime_utr     168139
8 three_prime_utr    195144
9 Selenocysteine        119</code></pre>
</div>
</div>
<p>How many unique protein-coding genes there are in the gtf?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Using dplyr with additionally removing duplicate genes</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>pcg <span class="ot">&lt;-</span> gtf <span class="sc">%&gt;%</span> </span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>      <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>(type<span class="sc">==</span><span class="st">'gene'</span> <span class="sc">&amp;</span> gene_biotype<span class="sc">==</span><span class="st">'protein_coding'</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(gene_name)) <span class="sc">%&gt;%</span></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">distinct</span>(gene_name, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">makeGRangesFromDataFrame</span>(<span class="at">keep.extra.columns =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(pcg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19355</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(pcg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GRanges"
attr(,"package")
[1] "GenomicRanges"</code></pre>
</div>
</div>
</section>
<section id="drawing-plots-of-metadata-information" class="level3">
<h3 class="anchored" data-anchor-id="drawing-plots-of-metadata-information">Drawing plots of metadata information</h3>
<p>Tabulate the number of genes per chromosome</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqnames</span>(pcg) <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>.
   1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 
2005 1220 1043  741  865 1011  894  668  750  713 1283  999  314  591  575  816 
  17   18   19   20   21   22    X    Y   MT 
1146  263 1389  529  215  427  839   46   13 </code></pre>
</div>
</div>
<p>Plot the number of unique protein-coding genes per chromosome. In this case, we do not filter for protein-coding genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PLot by chromosome</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">28</span>, <span class="at">repr.plot.height=</span><span class="dv">4</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>pcg <span class="sc">%&gt;%</span> </span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">#group_by(seqnames, gene_biotype) %&gt;%</span></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">.by=</span><span class="fu">c</span>(seqnames, gene_biotype), <span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> seqnames, <span class="at">y =</span> count, <span class="at">fill=</span>gene_biotype)) <span class="sc">+</span></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">'identity'</span>, <span class="at">position =</span> <span class="st">'dodge'</span>) <span class="sc">+</span></span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>),</span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title=</span><span class="fu">element_blank</span>(), <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>),      </span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust=</span><span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="GRanges_gtf_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Plot the different genes types per chromosome</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PLot by chromosome</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>gtf <span class="sc">%&gt;%</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(type<span class="sc">==</span><span class="st">'gene'</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(gene_name)) <span class="sc">%&gt;%</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(seqnames, gene_biotype) <span class="sc">%&gt;%</span></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> seqnames, <span class="at">y =</span> count, <span class="at">fill=</span>gene_biotype)) <span class="sc">+</span></span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">'identity'</span>, <span class="at">position =</span> <span class="st">'dodge'</span>) <span class="sc">+</span></span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>),</span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title=</span><span class="fu">element_blank</span>(), <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="at">legend.position =</span> <span class="st">'bottom'</span>,     </span>
<span id="cb65-13"><a href="#cb65-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust=</span><span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'seqnames'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="GRanges_gtf_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
</div>
<p>Plot the distribution across chromosomes per each gene type. Which gene type is the most abundant?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">20</span>, <span class="at">repr.plot.height=</span><span class="dv">8</span>)</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>gtf <span class="sc">%&gt;%</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(type<span class="sc">==</span><span class="st">'gene'</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(gene_name)) <span class="sc">%&gt;%</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(seqnames, gene_biotype) <span class="sc">%&gt;%</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_biotype, <span class="at">y =</span> count, <span class="at">fill=</span>seqnames)) <span class="sc">+</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">'identity'</span>, <span class="at">position =</span> <span class="st">'dodge'</span>) <span class="sc">+</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>),</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title=</span><span class="fu">element_blank</span>(), <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>),      </span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust=</span><span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'seqnames'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="GRanges_gtf_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="1920"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="defining-the-tss-of-transcripts" class="level2">
<h2 class="anchored" data-anchor-id="defining-the-tss-of-transcripts">Defining the TSS of transcripts</h2>
<p>In genomic analyses, we often need to generate a file containing <strong>transcription start site (TSS) positions or windows surrounding the TSS</strong>. This regions are important for <strong>defining promoter areas</strong>, where the regulation of gene expression primarily occurs. TSS-centered regions are also crucial for performing downstream analyses, e.g.&nbsp;motif discovery for transcription factors, or assessing the overlap with accessible regions (derived from ATAC-seq data) or regions of histone mark occupancy and transcription factor binding (ChIP-seq data).</p>
<p>GTF files are the primary source from where we can extract TSS positions. Here, we will identify the TSS of transcripts and create a window around them that spans 1000 bases upstream and 500 bases downstream of the TSS.</p>
<p>First we will retrieve all transcripts. We will filter for <strong>ensembl_havana annotated transcripts</strong>. HAVANA stands for Human And Vertebrate Analysis and Annotation. These are manually curated annotations (Wellcome Sanger Institute) integrated with the automatic annotation of Ensembl.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>tx <span class="ot">&lt;-</span> gtf <span class="sc">%&gt;%</span> </span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(type <span class="sc">==</span> <span class="st">"transcript"</span> <span class="sc">&amp;</span> source <span class="sc">==</span> <span class="st">"ensembl_havana"</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(transcript_name)) <span class="sc">%&gt;%</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">#distinct(transcript_name, .keep_all = TRUE) %&gt;%</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">makeGRangesFromDataFrame</span>(, <span class="at">keep.extra.columns =</span> <span class="cn">TRUE</span>)</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(tx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 26624</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tx,<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 22 metadata columns:
      seqnames          ranges strand |         source       type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt;
  [1]        1 1211340-1214153      - | ensembl_havana transcript        NA
  [2]        1 1203508-1206571      - | ensembl_havana transcript        NA
          phase         gene_id gene_version   gene_name    gene_source
      &lt;integer&gt;     &lt;character&gt;  &lt;character&gt; &lt;character&gt;    &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000186827           11     TNFRSF4 ensembl_havana
  [2]      &lt;NA&gt; ENSG00000186891           14    TNFRSF18 ensembl_havana
        gene_biotype   transcript_id transcript_version transcript_name
         &lt;character&gt;     &lt;character&gt;        &lt;character&gt;     &lt;character&gt;
  [1] protein_coding ENST00000379236                  4     TNFRSF4-201
  [2] protein_coding ENST00000328596                 10    TNFRSF18-201
      transcript_source transcript_biotype         tag     ccds_id
            &lt;character&gt;        &lt;character&gt; &lt;character&gt; &lt;character&gt;
  [1]    ensembl_havana     protein_coding       basic      CCDS11
  [2]    ensembl_havana     protein_coding       basic       CCDS9
      transcript_support_level exon_number     exon_id exon_version  protein_id
                   &lt;character&gt; &lt;character&gt; &lt;character&gt;  &lt;character&gt; &lt;character&gt;
  [1]   1 (assigned to previ..        &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;        &lt;NA&gt;
  [2]                        1        &lt;NA&gt;        &lt;NA&gt;         &lt;NA&gt;        &lt;NA&gt;
      protein_version
          &lt;character&gt;
  [1]            &lt;NA&gt;
  [2]            &lt;NA&gt;
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>To create TSS genomic coordinates, we will define TSS as start position for the ‘+’ strand and end position for the ‘-’ strand. Here, we use the <code>ifelse()</code> function to define the position: if strand is positive, then start equals to tx start position, otherwise (when strand is not positive) start equals to tx end position.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define TSS as start position for '+' strand and end position for '-' strand</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>tss <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">seqnames =</span> <span class="fu">seqnames</span>(tx),</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">ranges =</span> <span class="fu">IRanges</span>(</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">start =</span> <span class="fu">ifelse</span>(<span class="fu">strand</span>(tx) <span class="sc">==</span> <span class="st">"+"</span>, <span class="fu">start</span>(tx) , <span class="fu">end</span>(tx)),</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">end =</span> <span class="fu">ifelse</span>(<span class="fu">strand</span>(tx) <span class="sc">==</span> <span class="st">"+"</span>, <span class="fu">start</span>(tx), <span class="fu">end</span>(tx))</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">strand =</span> <span class="fu">strand</span>(tx),</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_name =</span> tx<span class="sc">$</span>gene_name,  </span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_id =</span> tx<span class="sc">$</span>gene_id </span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(tss)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 26624</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tss, <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 4 ranges and 2 metadata columns:
      seqnames    ranges strand |   gene_name         gene_id
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;
  [1]        1   1214153      - |     TNFRSF4 ENSG00000186827
  [2]        1   1206571      - |    TNFRSF18 ENSG00000186891
  [3]        1   1206592      - |    TNFRSF18 ENSG00000186891
  [4]        1   1471765      + |      ATAD3B ENSG00000160072
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Export the file with TSS genomic positions in BED format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">export</span>(tss, <span class="st">"../results/TSS.bed"</span>, <span class="at">format =</span> <span class="st">"BED"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>GRanges has dedicated method for doing the same job, called <code>resize()</code>. Let’s use the <code>resize</code> method to define the TSS</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Alternative way to define TSS with the resize method</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>tss<span class="fl">.2</span> <span class="ot">&lt;-</span> gtf <span class="sc">%&gt;%</span> .[<span class="fu">mcols</span>(.)<span class="sc">$</span>type<span class="sc">==</span><span class="st">'transcript'</span>] <span class="sc">%&gt;%</span></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>                 .[<span class="fu">mcols</span>(.)<span class="sc">$</span>source<span class="sc">==</span><span class="st">'ensembl_havana'</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">mcols</span>(.)<span class="sc">$</span>transcript_name)] <span class="sc">%&gt;%</span></span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">resize</span>(., <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>                  .[,<span class="fu">c</span>(<span class="st">'gene_name'</span>,<span class="st">'gene_id'</span>)] </span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a><span class="fu">export</span>(tss<span class="fl">.2</span>, <span class="st">"../results/TSS_v2.bed"</span>, <span class="at">format =</span> <span class="st">"BED"</span>)</span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(tss<span class="fl">.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 26624</code></pre>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tss<span class="fl">.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 6 ranges and 2 metadata columns:
      seqnames    ranges strand |   gene_name         gene_id
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;
  [1]        1   1214153      - |     TNFRSF4 ENSG00000186827
  [2]        1   1206571      - |    TNFRSF18 ENSG00000186891
  [3]        1   1206592      - |    TNFRSF18 ENSG00000186891
  [4]        1   1471765      + |      ATAD3B ENSG00000160072
  [5]        1   3069203      + |      PRDM16 ENSG00000142611
  [6]        1   1307930      - |       ACAP3 ENSG00000131584
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Verify that we got the same results</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(tss, tss<span class="fl">.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Try to define a window spanning the TSS 1000 bases upstream and 500 bases downstream. The upstream and downstream arguments define the number of nucleotides in the 5’ and 3’ direction, respectively.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define a 100 bp window around the TSS</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>tss_window <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">seqnames =</span> <span class="fu">seqnames</span>(tss),</span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ranges =</span> <span class="fu">IRanges</span>(</span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">start =</span> <span class="fu">pmax</span>(<span class="fu">start</span>(tss) <span class="sc">-</span> <span class="dv">1000</span>, <span class="dv">1</span>),  <span class="co"># Ensures start position is &gt;= 1. Returns the maximum value. </span></span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">end =</span> <span class="fu">end</span>(tss) <span class="sc">+</span> <span class="dv">500</span></span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>  <span class="co">#strand = strand(tss)</span></span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_name =</span> tx<span class="sc">$</span>gene_name,  </span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_id =</span> tx<span class="sc">$</span>gene_id </span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a><span class="co"># View result</span></span>
<span id="cb84-15"><a href="#cb84-15" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tss_window, <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 4 ranges and 2 metadata columns:
      seqnames          ranges strand |   gene_name         gene_id
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;
  [1]        1 1213153-1214653      * |     TNFRSF4 ENSG00000186827
  [2]        1 1205571-1207071      * |    TNFRSF18 ENSG00000186891
  [3]        1 1205592-1207092      * |    TNFRSF18 ENSG00000186891
  [4]        1 1470765-1472265      * |      ATAD3B ENSG00000160072
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>We can use the <em>intra range transformation</em> <code>promoter()</code> function of GRanges to do the same job. The output range is defined as</p>
<pre><code>(start(x) - upstream) to (start(x) + downstream - 1)
  </code></pre>
<p>for ranges on the + or * strand, and as</p>
<pre><code>(end(x) - downstream + 1) to (end(x) + upstream)
  </code></pre>
<p>for ranges on the - strand</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>tss_window <span class="ot">&lt;-</span> <span class="fu">promoters</span>(tx, <span class="at">upstream=</span><span class="dv">1000</span>, <span class="at">downstream=</span><span class="dv">500</span>) <span class="sc">%&gt;%</span> .[,<span class="fu">c</span>(<span class="st">'gene_name'</span>)]</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="co">#export(tss_window, "../results/TSS_window.bed", format='BED')</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>tss_window <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 4 ranges and 1 metadata column:
      seqnames          ranges strand |   gene_name
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
  [1]        1 1213654-1215153      - |     TNFRSF4
  [2]        1 1206072-1207571      - |    TNFRSF18
  [3]        1 1206093-1207592      - |    TNFRSF18
  [4]        1 1470765-1472264      + |      ATAD3B
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="intersect-with-tf-bound-regions" class="level2">
<h2 class="anchored" data-anchor-id="intersect-with-tf-bound-regions">Intersect with TF-bound regions</h2>
<p>One common question in the analysis of genomic intervals is what are the genes regulated by various trascription factors, or other members of the transcriptional machinery. In that case, we might want to compare the TSS intervals of genes to regions bound by a transcription factor.</p>
<p>We will determine the overlap between genomic regions bound by BATF, a pioneer transcription factor, and promoter intervals to identify genes potentially regulated by BATF. The BATF genomic regions derive from an experiment where naive CD4+ T cells isolated from human umbilical cord blood were in vitro polarized to Th17-fate for 72h and further subjected to BATF ChIP analysis. Genome-wide ChIP-seq reads were aligned to the hg38 genome assembly. The data is deposited in the public repository Gene Expression Omnibus (GEO) with accession id GSE174810. The latter is a SubSeries which is is part of SuperSeries GSE176065.</p>
<p>[<strong>GEO</strong>]<a href="https://www.ncbi.nlm.nih.gov/geo/" class="uri">https://www.ncbi.nlm.nih.gov/geo/</a></p>
<p><em>GSE174810: Genome-wide binding sites of FOSL1, FOSL2 and BATF in human Th17 cells [ChIP-seq]</em></p>
<center>
<p><img src="../pics/GEOpng.png" class="img-fluid" width="600"></p>
</center>
<section id="retrieve-data-from-geo" class="level3">
<h3 class="anchored" data-anchor-id="retrieve-data-from-geo">Retrieve data from GEO</h3>
<p>We will fetch the supplementary files with the GEOquery function <code>getGEOSuppFiles()</code>. This will download the files in a directory named according to the GEO accession id. We can then untar/unzip the file and import the files we need. We will try different ways to achieve this:</p>
<p><code>getGEO()</code> is the main user-level function in the GEOquery package. For more information, check out the package vignette <a href="https://bioconductor.org/packages/devel/bioc/vignettes/GEOquery/inst/doc/GEOquery.html">GEOquery</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GEOquery)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting options('download.file.method.GEOquery'='auto')</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting options('GEOquery.inmemory.gpl'=FALSE)</code></pre>
</div>
</div>
<ol type="1">
<li>Download supplementary files</li>
</ol>
<p>Check the path to the current directory in order to understand where the file is downloaded.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Download data from a GEO</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download data from a GEO and list files</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="co">#setwd('../')</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a><span class="fu">getGEOSuppFiles</span>(<span class="st">"GSE174810"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check that the tar file has been downloaded</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">'../GSE174810/'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "BATF_peaks.narrowPeak.gz" "GSE174810_RAW"           
[3] "GSE174810_RAW.tar"       </code></pre>
</div>
</div>
<p>Check the list of files in the tar</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">untar</span>(<span class="st">"GSE174810/GSE174810_RAW.tar"</span>, <span class="at">list =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To untar directly from R we can use the built-in <code>untar()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">untar</span>(<span class="st">"../GSE174810/GSE174810_RAW.tar"</span>, <span class="at">exdir =</span> <span class="st">"../GSE174810/GSE174810_RAW"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>List the supplementary files</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">'../GSE174810/GSE174810_RAW/'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GSM5328649_BATF_R1_peaks.narrowPeak.gz"      
[2] "GSM5328651_BATF_R3_peaks.narrowPeak.gz"      
[3] "GSM5328653_FOSL1_C-12_R1_peaks.narrowPeak.gz"
[4] "GSM5328655_FOSL1_C-12_R2_peaks.narrowPeak.gz"
[5] "GSM5328657_FOSL2_R1_peaks.narrowPeak.gz"     
[6] "GSM5328659_FOSL2_R2_peaks.narrowPeak.gz"     </code></pre>
</div>
</div>
<p>Import the file of interest, e.g.&nbsp;BATF_R1_peaks, by providing the connection <code>con</code> from which data is loaded.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="co"># R.utils::gunzip("../GSE174810/GSE174810_RAW/GSM5328649_BATF_R1_peaks.narrowPeak.gz", </span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a><span class="co">#                 overwrite = TRUE)</span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>file_path <span class="ot">&lt;-</span> <span class="st">"../GSE174810/GSE174810_RAW/GSM5328649_BATF_R1_peaks.narrowPeak.gz"</span></span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Does the file exist?</span></span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a><span class="fu">file.exists</span>(file_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="dv">1000</span>)</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>batf <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="at">con=</span>file_path, <span class="at">format =</span> <span class="st">"narrowPeak"</span>)</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a><span class="co">#dim(bed)</span></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>batf <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 6 metadata columns:
      seqnames          ranges strand |           name     score signalValue
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt; &lt;numeric&gt;   &lt;numeric&gt;
  [1]     chr1   992366-992706      * | BATF_R1_peak_1       491    10.86899
  [2]     chr1 1013033-1013271      * | BATF_R1_peak_2        81     5.07551
         pValue    qValue      peak
      &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;
  [1]   52.8283  49.18042       180
  [2]   10.9246   8.12685       170
  -------
  seqinfo: 33 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<ol start="2" type="1">
<li>Retrieve the file from GEO providing a URL.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define URL</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>batf_url <span class="ot">&lt;-</span> <span class="st">"https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5328nnn/GSM5328649/suppl/GSM5328649%5FBATF%5FR1%5Fpeaks%2EnarrowPeak%2Egz"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># error when rtracklayer::import is used directly on a remote .gz file</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Download first</span></span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>destfile <span class="ot">&lt;-</span> <span class="st">"../GSE174810/BATF_peaks.narrowPeak.gz"</span></span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(batf_url, <span class="at">destfile =</span> destfile)</span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-7"><a href="#cb108-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Then import it</span></span>
<span id="cb108-8"><a href="#cb108-8" aria-hidden="true" tabindex="-1"></a>batf <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="at">con =</span> destfile, <span class="at">format =</span> <span class="st">"narrowPeak"</span>)</span>
<span id="cb108-9"><a href="#cb108-9" aria-hidden="true" tabindex="-1"></a>batf <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 6 ranges and 6 metadata columns:
      seqnames          ranges strand |           name     score signalValue
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt; &lt;numeric&gt;   &lt;numeric&gt;
  [1]     chr1   992366-992706      * | BATF_R1_peak_1       491    10.86899
  [2]     chr1 1013033-1013271      * | BATF_R1_peak_2        81     5.07551
  [3]     chr1 1064103-1064463      * | BATF_R1_peak_3       515    16.40878
  [4]     chr1 1098903-1099465      * | BATF_R1_peak_4        39     3.95218
  [5]     chr1 1208075-1208187      * | BATF_R1_peak_5        36     3.40441
  [6]     chr1 1217964-1218107      * | BATF_R1_peak_6        53     4.60597
         pValue    qValue      peak
      &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;
  [1]  52.82834  49.18042       180
  [2]  10.92463   8.12685       170
  [3]  55.24189  51.56105       189
  [4]   6.55161   3.96331       393
  [5]   6.17971   3.61470        59
  [6]   8.03364   5.36412        56
  -------
  seqinfo: 33 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>Let us check the length distribution of the intervals</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(<span class="fu">width</span>(batf))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 306.8564</code></pre>
</div>
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co">#mcols(batf)$width &lt;- width(batf)</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>batf <span class="sc">%&gt;%</span> </span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(width)) <span class="sc">+</span> </span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>() <span class="sc">+</span> </span>
<span id="cb112-7"><a href="#cb112-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">trans=</span><span class="st">'log2'</span>) <span class="sc">+</span></span>
<span id="cb112-8"><a href="#cb112-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept=</span><span class="fu">mean</span>(<span class="fu">width</span>(batf)), <span class="at">color=</span><span class="st">"red"</span>, <span class="at">linewidth=</span><span class="fl">0.1</span> ) <span class="sc">+</span></span>
<span id="cb112-9"><a href="#cb112-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="GRanges_gtf_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="384"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co"># log2(width) returns log2 values on the x-axis, whilst scale_x_continuous() keeps the original values.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the seqnames. You will notice that it contains seqnames not present in the tss_window()</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqnames</span>(batf) <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>.
                   chr1                   chr10                   chr11 
                   4462                    2267                    1958 
                  chr12                   chr13                   chr14 
                   2522                    1347                    1634 
                  chr15                   chr16                   chr17 
                   1450                    1144                    1652 
chr17_GL000205v2_random                   chr18                   chr19 
                      1                     915                     950 
 chr1_KI270706v1_random                    chr2                   chr20 
                      1                    4209                     949 
                  chr21                   chr22 chr22_KI270733v1_random 
                    622                     592                       1 
                   chr3                    chr4                    chr5 
                   3375                    2256                    2491 
                   chr6                    chr7                    chr8 
                   3390                    2148                    2218 
                   chr9        chrUn_GL000195v1        chrUn_GL000219v1 
                   1625                       1                       2 
       chrUn_GL000220v1        chrUn_KI270438v1        chrUn_KI270466v1 
                      1                       1                       1 
       chrUn_KI270467v1                    chrX                    chrY 
                      3                    1047                       8 </code></pre>
</div>
</div>
<p>From the UCSC site:</p>
<p>Chr<em>N</em>_random relates to sequence that is known to be in a particular chromosome, but could not be reliably ordered within the current sequence. ChrUn contains clone contigs that cannot be confidently placed on a specific chromosome. For the chrN_random and chrUn_random files, we essentially just concatenate together all the contigs into short pseudo-chromosomes. The coordinates of these are fairly arbitrary, although the relative positions of the coordinates are good within a contig. You can find more information about the data organization and format on the <a href="https://genome.ucsc.edu/goldenPath/datorg.html">Data Organization and Format page</a>.</p>
<p>To be able to intersect the two GRanges objects, <strong>the seqnames need to match</strong>. We need to modify the seqnames() of the <code>batf GR object</code> to match the levels of the <code>tss_window GR object</code>. To keep it simple, we will only change the seqnames for the autosomal chromosomes (1:22) plus chromosomes X and Y. It is also important to ensure that the genomic positions refer to the same genome assembly, in this case the human build 38!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Seqnames of tss_window</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="fu">seqnames</span>(tss_window) <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>.
   1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 
2694 1761 1535 1031 1146 1365 1221  955 1004  968 1683 1470  419  842  834 1130 
  17   18   19   20   21   22    X    Y   MT 
1592  403 1795  729  312  591 1071   73    0 </code></pre>
</div>
</div>
</section>
</section>
<section id="convert-seqnames-of-batf-gr-object" class="level2">
<h2 class="anchored" data-anchor-id="convert-seqnames-of-batf-gr-object">Convert seqnames of BATF GR object</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Modify seqnames</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="co">#remove_seqnames &lt;- "chrM|chrUn|chrEBV|random"</span></span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a><span class="co">#batf &lt;- batf[!(grepl(remove_seqnames, seqnames(batf)) ) ]</span></span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a><span class="co">#batf &lt;- keepSeqlevels(batf, </span></span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a><span class="co">#           seqlevelsInUse(batf), </span></span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a><span class="co">#          pruning.mode="coarse" #Remove the elements in x where the seqlevels to drop are in use.</span></span>
<span id="cb118-7"><a href="#cb118-7" aria-hidden="true" tabindex="-1"></a><span class="co">#          ) </span></span>
<span id="cb118-8"><a href="#cb118-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb118-9"><a href="#cb118-9" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlevels</span>(batf) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"^chr"</span>, <span class="st">""</span>, <span class="fu">seqlevels</span>(batf))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqnames</span>(batf) <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>.
                   1                   10                   11 
                4462                 2267                 1958 
                  12                   13                   14 
                2522                 1347                 1634 
                  15                   16                   17 
                1450                 1144                 1652 
17_GL000205v2_random                   18                   19 
                   1                  915                  950 
 1_KI270706v1_random                    2                   20 
                   1                 4209                  949 
                  21                   22 22_KI270733v1_random 
                 622                  592                    1 
                   3                    4                    5 
                3375                 2256                 2491 
                   6                    7                    8 
                3390                 2148                 2218 
                   9        Un_GL000195v1        Un_GL000219v1 
                1625                    1                    2 
       Un_GL000220v1        Un_KI270438v1        Un_KI270466v1 
                   1                    1                    1 
       Un_KI270467v1                    X                    Y 
                   3                 1047                    8 </code></pre>
</div>
</div>
<p>We will use the <code>subsetByOverlaps</code> method of GRanges to extract the elements in the query (e.g.&nbsp;TSS) that overlap at least one element in the subject (e.g.&nbsp;batf).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>overlap <span class="ot">&lt;-</span> <span class="fu">subsetByOverlaps</span>(tss_window, batf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .merge_two_Seqinfo_objects(x, y): Each of the 2 combined objects has sequence levels not in the other:
  - in 'x': MT
  - in 'y': 17_GL000205v2_random, 1_KI270706v1_random, 22_KI270733v1_random, Un_GL000195v1, Un_GL000219v1, Un_GL000220v1, Un_KI270438v1, Un_KI270466v1, Un_KI270467v1
  Make sure to always combine/compare objects based on the same reference
  genome (use suppressWarnings() to suppress this warning).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>overlap</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3888 ranges and 1 metadata column:
         seqnames            ranges strand |   gene_name
            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
     [1]        1 12617711-12619210      - |       DHRS3
     [2]        1   8525440-8526939      - |        RERE
     [3]        1   8423167-8424666      - |        RERE
     [4]        1 11805191-11806690      + |       CLCN6
     [5]        1   2526097-2527596      - |       PANK4
     ...      ...               ...    ... .         ...
  [3884]       21 46634642-46636141      + |       PRMT2
  [3885]       21 46634653-46636152      + |       PRMT2
  [3886]       21 46634681-46636180      + |       PRMT2
  [3887]       21 33947926-33949425      + |   LINC00649
  [3888]       21 46285384-46286883      + |        YBEY
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>Can you tell what the warning is about?</p>
</blockquote>
<p>Now we can retrieve the genes, whose promoter is bound by the BATF transcription factor.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> overlap<span class="sc">$</span>gene_name <span class="sc">%&gt;%</span> <span class="fu">unique</span>() <span class="sc">%&gt;%</span> <span class="fu">sort</span>()</span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2709</code></pre>
</div>
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AAAS"  "AAGAB" "AAK1"  "AAMP"  "AARS1" "ABCB1"</code></pre>
</div>
</div>
</section>
<section id="range-based-transformations-are-grouped-in-2-categories" class="level2">
<h2 class="anchored" data-anchor-id="range-based-transformations-are-grouped-in-2-categories">Range-based transformations are grouped in 2 categories:</h2>
<ol type="1">
<li><p>Intra range transformations (e.g.&nbsp;shift()) <strong>transform each range individually</strong> (and independently of the other ranges). They return an object parallel to the input object, that is, where the i-th range corresponds to the i-th range in the input. Those transformations are described in the intra-range-methods man page (see ?<code>intra-range-methods</code>).</p></li>
<li><p>Inter range transformations (e.g.&nbsp;reduce()) transform all the ranges together as a set to produce a new set of ranges. They return an object that is generally NOT parallel to the input object. Those transformations are described below.</p></li>
</ol>
<section id="intra-range-functions" class="level3">
<h3 class="anchored" data-anchor-id="intra-range-functions">Intra-range functions</h3>
<p>Operate on each element of a GRanges object independently of the other ranges in the object.</p>
<p>For instance, the <code>flank</code> method allows you to generate regions flanking the existing ranges in a GRanges object in a strand dependent manner. To capture the 10 bases located upstream of these ranges:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> gr[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="fu">flank</span>(g, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   112-121      - |         1  1.000000
  b     chr2    92-101      + |         2  0.888889
  c     chr2    93-102      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>To include the downstream bases:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">flank</span>(g, <span class="dv">10</span>, <span class="at">start=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1    91-100      - |         1  1.000000
  b     chr2   113-122      + |         2  0.888889
  c     chr2   114-123      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>We can shift the ranges by a specific number of bases with <code>shift</code>. This is performed is a strand independent manner.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Shift (same direction) both start and end by N bp </span></span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a><span class="fu">shift</span>(g, <span class="sc">-</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1    96-106      - |         1  1.000000
  b     chr2    97-107      + |         2  0.888889
  c     chr2    98-108      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>The <code>resize</code> method will extend the ranges by a specified width from the start in a strand dependent manner.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Rize width to N bp (from start)</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a><span class="fu">resize</span>(g, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   107-111      - |         1  1.000000
  b     chr2   102-106      + |         2  0.888889
  c     chr2   103-107      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="inter-range-functions" class="level3">
<h3 class="anchored" data-anchor-id="inter-range-functions">Inter-range functions</h3>
<p>Transform all the ranges together as a set to produce a new set of ranges.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  d     chr2   104-114      * |         4  0.666667
  e     chr1   105-115      * |         5  0.555556
  f     chr1   106-116      + |         6  0.444444
  g     chr3   107-117      + |         7  0.333333
  h     chr3   108-118      + |         8  0.222222
  i     chr3   109-119      - |         9  0.111111
  j     chr3   110-120      - |        10  0.000000
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<ol type="1">
<li>Align the ranges and merge overlapping ranges to produce a simplified set. It maintains information related to the strand.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Merge overlapping intervals</span></span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a><span class="fu">reduce</span>(gr) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(start)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 7 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1   101-111      -
  [2]     chr2   102-113      +
  [3]     chr2   104-114      *
  [4]     chr1   105-115      *
  [5]     chr1   106-116      +
  [6]     chr3   107-118      +
  [7]     chr3   109-120      -
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<ol start="2" type="1">
<li>Retrieve the gaps between the ranges</li>
</ol>
<p>For example, it might be useful to retrieve all regions not bound by a transcription factor to interrogate further.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Retrieve the gaps between the intervals</span></span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gaps</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 7 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1     1-105      +
  [2]     chr1     1-100      -
  [3]     chr1     1-104      *
  [4]     chr2     1-101      +
  [5]     chr2     1-103      *
  [6]     chr3     1-106      +
  [7]     chr3     1-108      -
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<ol start="3" type="1">
<li>A collection of non-overlapping ranges</li>
</ol>
<p>Rather than merging intervals that overlap by 1 bp, it retrieves the complete overlapping plus non-overlapping regions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Returns the non-overlapping regions</span></span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a><span class="fu">disjoin</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 13 ranges and 0 metadata columns:
       seqnames    ranges strand
          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
   [1]     chr1   106-116      +
   [2]     chr1   101-111      -
   [3]     chr1   105-115      *
   [4]     chr2       102      +
   [5]     chr2   103-112      +
   ...      ...       ...    ...
   [9]     chr3   108-117      +
  [10]     chr3       118      +
  [11]     chr3       109      -
  [12]     chr3   110-119      -
  [13]     chr3       120      -
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<ol start="4" type="1">
<li>Quantify the degree of overlap for all the ranges in a GRanges object.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example GRanges object</span></span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr1"</span>, <span class="st">"chr2"</span>),</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="at">start =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>), <span class="at">end =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">7</span>)),</span>
<span id="cb147-4"><a href="#cb147-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">strand =</span> <span class="fu">rep</span>(<span class="st">'+'</span>,<span class="dv">3</span>)</span>
<span id="cb147-5"><a href="#cb147-5" aria-hidden="true" tabindex="-1"></a>             <span class="co">#strand = Rle(strand('+'),3)</span></span>
<span id="cb147-6"><a href="#cb147-6" aria-hidden="true" tabindex="-1"></a>             <span class="co">#strand = Rle(rep('+',3))</span></span>
<span id="cb147-7"><a href="#cb147-7" aria-hidden="true" tabindex="-1"></a>             )</span>
<span id="cb147-8"><a href="#cb147-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb147-9"><a href="#cb147-9" aria-hidden="true" tabindex="-1"></a><span class="co"># = Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2))</span></span>
<span id="cb147-10"><a href="#cb147-10" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1       1-4      +
  [2]     chr1       3-6      +
  [3]     chr2       4-7      +
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">coverage</span>(g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>RleList of length 2
$chr1
integer-Rle of length 6 with 3 runs
  Lengths: 2 2 2
  Values : 1 2 1

$chr2
integer-Rle of length 7 with 2 runs
  Lengths: 3 4
  Values : 0 1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to dataframe</span></span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute coverage (returns an RleList)</span></span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>covg <span class="ot">&lt;-</span> <span class="fu">coverage</span>(g)</span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert coverage to a data frame</span></span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a>df_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(covg), <span class="cf">function</span>(chr) {</span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(</span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqnames =</span> chr,</span>
<span id="cb151-9"><a href="#cb151-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">seq_along</span>(covg[[chr]]),  <span class="co"># Positions along the chromosome</span></span>
<span id="cb151-10"><a href="#cb151-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">coverage =</span> <span class="fu">as.vector</span>(covg[[chr]])   <span class="co"># Convert Rle to numeric vector</span></span>
<span id="cb151-11"><a href="#cb151-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb151-12"><a href="#cb151-12" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb151-13"><a href="#cb151-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-14"><a href="#cb151-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine into a single data frame</span></span>
<span id="cb151-15"><a href="#cb151-15" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, df_list)</span>
<span id="cb151-16"><a href="#cb151-16" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   seqnames position coverage
1      chr1        1        1
2      chr1        2        1
3      chr1        3        2
4      chr1        4        2
5      chr1        5        1
6      chr1        6        1
7      chr2        1        0
8      chr2        2        0
9      chr2        3        0
10     chr2        4        1
11     chr2        5        1
12     chr2        6        1
13     chr2        7        1</code></pre>
</div>
</div>
</section>
</section>
<section id="interval-set-operations-for-granges-objects" class="level2">
<h2 class="anchored" data-anchor-id="interval-set-operations-for-granges-objects">Interval set operations for GRanges objects</h2>
<p>Perform operations between two GRanges objects.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> gr[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> <span class="fu">head</span>(gr, <span class="at">n=</span><span class="dv">2</span>)</span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-4"><a href="#cb153-4" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>g2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Retrieve the union of regions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combined ranges</span></span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>GenomicRanges<span class="sc">::</span><span class="fu">union</span>(g, g2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1   101-111      -
  [2]     chr2   102-113      +
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Retrieve overlapping regions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Overlapping regions</span></span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a>GenomicRanges<span class="sc">::</span><span class="fu">intersect</span>(g, g2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1   101-111      -
  [2]     chr2   102-112      +
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Retrieve the non-overlapping regions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Non overlapping regions</span></span>
<span id="cb161-2"><a href="#cb161-2" aria-hidden="true" tabindex="-1"></a>GenomicRanges<span class="sc">::</span><span class="fu">setdiff</span>(g, g2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr2       113      +
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="splitting-and-combining-granges-objects" class="level2">
<h2 class="anchored" data-anchor-id="splitting-and-combining-granges-objects">Splitting and combining GRanges objects</h2>
<p>GRanges objects can be divided into groups using the <code>split</code> function. This produces a GRangesList object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Split using a factor that defines the grouping `rep(1:2, each=5)`. Use `time` instead of `each`.</span></span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">split</span>(gr, <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">each=</span><span class="dv">5</span>))</span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>sp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRangesList object of length 2:
$`1`
GRanges object with 5 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  d     chr2   104-114      * |         4  0.666667
  e     chr1   105-115      * |         5  0.555556
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths

$`2`
GRanges object with 5 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  f     chr1   106-116      + |         6  0.444444
  g     chr3   107-117      + |         7  0.333333
  h     chr3   108-118      + |         8  0.222222
  i     chr3   109-119      - |         9  0.111111
  j     chr3   110-120      - |        10  0.000000
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Set the names of the two items in the list</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Name the two elements of the lists</span></span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sp) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'A'</span>,<span class="st">'B'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the length of each item</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lengths</span>(sp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A B 
5 5 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Retrieve the first item in the list</span></span>
<span id="cb168-2"><a href="#cb168-2" aria-hidden="true" tabindex="-1"></a><span class="co">#sp[[1]]</span></span>
<span id="cb168-3"><a href="#cb168-3" aria-hidden="true" tabindex="-1"></a>sp<span class="sc">$</span>A</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 5 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  d     chr2   104-114      * |         4  0.666667
  e     chr1   105-115      * |         5  0.555556
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Concatenate the two lists</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Concatenate </span></span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(sp[[<span class="dv">1</span>]], sp[[<span class="dv">2</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  d     chr2   104-114      * |         4  0.666667
  e     chr1   105-115      * |         5  0.555556
  f     chr1   106-116      + |         6  0.444444
  g     chr3   107-117      + |         7  0.333333
  h     chr3   108-118      + |         8  0.222222
  i     chr3   109-119      - |         9  0.111111
  j     chr3   110-120      - |        10  0.000000
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>A useful and fast strategy is to unlist the GRangesList to a GRanges object, operate on the GRanges object, then relist the result.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Unlist into a GRanges object</span></span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unlist</span>(sp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 2 metadata columns:
      seqnames    ranges strand |     score        GC
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  A.a     chr1   101-111      - |         1  1.000000
  A.b     chr2   102-112      + |         2  0.888889
  A.c     chr2   103-113      + |         3  0.777778
  A.d     chr2   104-114      * |         4  0.666667
  A.e     chr1   105-115      * |         5  0.555556
  B.f     chr1   106-116      + |         6  0.444444
  B.g     chr3   107-117      + |         7  0.333333
  B.h     chr3   108-118      + |         8  0.222222
  B.i     chr3   109-119      - |         9  0.111111
  B.j     chr3   110-120      - |        10  0.000000
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Splitting can be performed based on genomic coordinate information or metadata.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Split by chromosome</span></span>
<span id="cb174-2"><a href="#cb174-2" aria-hidden="true" tabindex="-1"></a>singles <span class="ot">&lt;-</span> <span class="fu">split</span>(gr, <span class="fu">seqnames</span>(gr))</span>
<span id="cb174-3"><a href="#cb174-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(singles)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "chr1" "chr2" "chr3"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>singles</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRangesList object of length 3:
$chr1
GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  e     chr1   105-115      * |         5  0.555556
  f     chr1   106-116      + |         6  0.444444
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths

$chr2
GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  d     chr2   104-114      * |         4  0.666667
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths

$chr3
GRanges object with 4 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  g     chr3   107-117      + |         7  0.333333
  h     chr3   108-118      + |         8  0.222222
  i     chr3   109-119      - |         9  0.111111
  j     chr3   110-120      - |        10  0.000000
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Split based on metadata information</span></span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a>singles <span class="ot">&lt;-</span> <span class="fu">split</span>(gr, <span class="fu">mcols</span>(gr)<span class="sc">$</span>score<span class="sc">&lt;</span><span class="dv">4</span>)</span>
<span id="cb178-3"><a href="#cb178-3" aria-hidden="true" tabindex="-1"></a>singles</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRangesList object of length 2:
$`FALSE`
GRanges object with 7 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  d     chr2   104-114      * |         4  0.666667
  e     chr1   105-115      * |         5  0.555556
  f     chr1   106-116      + |         6  0.444444
  g     chr3   107-117      + |         7  0.333333
  h     chr3   108-118      + |         8  0.222222
  i     chr3   109-119      - |         9  0.111111
  j     chr3   110-120      - |        10  0.000000
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths

$`TRUE`
GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>How would you split for start position less than 106 and GC more than 0.7</p>
<p>It is also possible to append an interval to the GRanges object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Append to the ranges</span></span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> gr[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb180-3"><a href="#cb180-3" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">append</span>(g, singles[[<span class="dv">1</span>]][<span class="dv">1</span>])</span>
<span id="cb180-4"><a href="#cb180-4" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 4 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr2   102-112      + |         2  0.888889
  c     chr2   103-113      + |         3  0.777778
  d     chr2   104-114      * |         4  0.666667
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a>singles[[<span class="dv">1</span>]][<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  d     chr2   104-114      * |         4  0.666667
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Or replace a genomic interval</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace a genomic interval</span></span>
<span id="cb184-2"><a href="#cb184-2" aria-hidden="true" tabindex="-1"></a>g[<span class="dv">2</span>] <span class="ot">&lt;-</span> singles[[<span class="dv">1</span>]][<span class="dv">2</span>]</span>
<span id="cb184-3"><a href="#cb184-3" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 4 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr1   105-115      * |         5  0.555556
  c     chr2   103-113      + |         3  0.777778
  d     chr2   104-114      * |         4  0.666667
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb186"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb186-1"><a href="#cb186-1" aria-hidden="true" tabindex="-1"></a>singles[[<span class="dv">1</span>]][<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;
  e     chr1   105-115      * |         5  0.555556
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Replace a value of a genomic interval</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace the strand in the second interval</span></span>
<span id="cb188-2"><a href="#cb188-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(g)<span class="sc">$</span>score[<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb188-3"><a href="#cb188-3" aria-hidden="true" tabindex="-1"></a><span class="fu">strand</span>(g[<span class="dv">2</span>]) <span class="ot">&lt;-</span> <span class="st">'+'</span></span>
<span id="cb188-4"><a href="#cb188-4" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 4 ranges and 2 metadata columns:
    seqnames    ranges strand |     score        GC
       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;
  a     chr1   101-111      - |         1  1.000000
  b     chr1   105-115      + |         5  0.555556
  c     chr2   103-113      + |        10  0.777778
  d     chr2   104-114      * |         4  0.666667
  -------
  seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb190"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a><span class="co"># gr[IRanges(start=c(2,7), end=c(3,8))] # genomic intervals [2:3] and [7:8]: from 2nd to 3rd row and from 7th to 8th row</span></span>
<span id="cb190-2"><a href="#cb190-2" aria-hidden="true" tabindex="-1"></a><span class="co"># gr %&gt;% filter(start&lt;107, start&gt;102)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb191"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a><span class="co">#help("inter-range-methods", package = "GenomicRanges")</span></span>
<span id="cb191-2"><a href="#cb191-2" aria-hidden="true" tabindex="-1"></a><span class="co">#?`inter-range-methods`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>GRanges offers parallel methods when two objects have matching structures, e.g.&nbsp;element 1 of one object corresponds to element 1 of the other. These methods, which start with “p” (for parallel), perform element-wise operations like unions. Both objects must have the same number of elements, with matching seqnames and strand throughout.</p>
</blockquote>
</section>
<section id="online-resources-and-databases-for-omics-data" class="level2">
<h2 class="anchored" data-anchor-id="online-resources-and-databases-for-omics-data">Online Resources and Databases for omics data</h2>
<p>The section below briefly introduce you to the description and usage of some very useful online resources for ChIP-seq public data and also other omics data. Take a closer look to explore them, you might need them in the future in order to complement your future experiments!</p>
<section id="poor-mans-guide-to-omics" class="level3">
<h3 class="anchored" data-anchor-id="poor-mans-guide-to-omics">Poor man’s guide to omics</h3>
<p>When approaching a new problem, you might conceive <strong>multiple hypotheses</strong> that could instruct a wide variety of experimental approaches. However, you have neither the time, nor the money or the inclination to perform all of these experiments. Chances are, though, that somewhere in a server there is a dataset that can help you, be it a ChIP-seq dataset or any other omic. They may not give you an answer outright, but they’ll help you <strong>prune your hypotheses</strong> and only keep the ones that have some foundation. After all, while big data analysis often results in false positives, or at least positives that are contingent to a specific model, it is very useful to avoid pursuing roads for which no evidence is found in datasets.</p>
<p>This section aims to give you <strong>resources</strong> that you can employ to perform simple analyses that prevent you from wasting your time and allow you to focus on the experiments that matter.</p>
<p>Over the years, multiple consortia have assembled these datasets into curated collections, and new datasets are published with almost every new paper, neatly stored into various sites.</p>
<p>Here we provide a handful of <strong>key databases</strong> and characteristics of the main ones, as well as a list of other databases that may be useful for more specific questions.</p>
<section id="encode-the-encyclopedia-of-dna-elements" class="level4">
<h4 class="anchored" data-anchor-id="encode-the-encyclopedia-of-dna-elements">ENCODE: the encyclopedia of DNA elements</h4>
<p>The closest to the object of the workshop, over more than 20 years of activity the ENCODE consortium is the biggest effort in <strong>functional genomics</strong> in the world, assembling thousands of datasets. You can easily browse their portal by theme and find ready-made datasets that pertain to your field. Once you have your dataset (or datasets) of choice, you only have to browse it with the help of this workshop’s scripts or visualize it, for example with IGV.</p>
<p><a href="https://www.encodeproject.org/">Encode project</a></p>
</section>
<section id="geo-the-gene-expression-omnibus" class="level4">
<h4 class="anchored" data-anchor-id="geo-the-gene-expression-omnibus">GEO: The Gene Expression Omnibus</h4>
<p>This portal is almost ubiquitous in publications - in fact, nearly all <strong>gene expression datasets</strong> used in publications, both bulk and single cell, are published on the GEO, a service handled by the NCBI. The GEO datasets can be accessed from papers, but also directly from the portal, and the neat thing is that they are generally well annotated so that you can understand how specific experiments were performed. They have specific files that tell you how the data was generated. This kind of information can be complementary to epigenome data, or can inform derivative work giving you invaluable insight into a model you want to adopt.</p>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus</a></p>
</section>
<section id="tcga-the-cancer-genome-atlas" class="level4">
<h4 class="anchored" data-anchor-id="tcga-the-cancer-genome-atlas">TCGA: The Cancer Genome Atlas</h4>
<p>Among the biggest efforts in contemporary medical research, spearheaded by the NIH. The Cancer Genome Atlas has gathered <strong>transcriptomics, methylation, and chromatin accessibility</strong> data for an incredibly wide array of malignancies, and it keeps expanding with new cohorts of patients, curated by individual contributors. While it is a specific resource pertaining to oncology, its wealth of controls and diversity of malignancies makes it invaluable for anybody working on human biology - after all, most fundamental biology experiments are performed on cancer cell lines.</p>
<p><a href="https://www.cancer.gov/ccg/research/genome-sequencing/tcga">The Cancer Genome Atlas</a></p>
</section>
<section id="the-human-cell-atlas" class="level4">
<h4 class="anchored" data-anchor-id="the-human-cell-atlas">The Human Cell Atlas</h4>
<p>Arguably the biggest ongoing sequencing project in the world, the Human Cell Atlas aims to fully leverage <strong>single cell transcriptomes</strong> to define every <em>cell type</em> in the human body. Currently hosting data for close to <strong>60 million</strong> sequenced cells and still counting, this is the largest single-cell resolved gene expression resource currently available for non-diseased tissue. Allows downloads in formats that are ‘easily’ browsed - meaning you don’t have to do core processing yourself.</p>
<p><a href="https://data.humancellatlas.org/">The Human Cell Atlas - HCA</a></p>
</section>
<section id="additional-links" class="level4">
<h4 class="anchored" data-anchor-id="additional-links">Additional links:</h4>
<p><a href="https://egg2.wustl.edu/roadmap/web_portal/">Roadmap consortium</a> Generated tracks for multiple modifications, and produced ChromHMM to sort them out</p>
<p><a href="https://www.ebi.ac.uk/bioimage-archive/">EBI Bioimage Archive</a> European resource for bioimages. Pretty unique and specific, rare type of datasets.</p>
<p><a href="https://gtexportal.org/home/">GTEX</a> Tissue-specific genotype regulation - a complex database that can tell you whether your gene of interest works in the context you study.</p>
<p><a href="https://libguides.brown.edu/c.php?g=545426&amp;p=3741199">Brown university collection</a> Collection of resources, including some of those we showed above. Take a look when you need to find some new source of data!</p>
<p><a href="https://www.nlm.nih.gov/NIHbmic/domain_specific_repositories.html">NIH’s collection - even better than Brown’s!</a> Includes even more well-categorized resources.</p>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>