<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2025-05-26">

<title>ChIP-seq_analysis_part1 â€“ 2025 MBMM</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-de84f8d6bb715db06a919283c2d1e787.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-de2f664832f57b2c83139acc48228f9e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="../../site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="../../site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">2025 MBMM</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../pages/index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../IntroGR/pages/Intro_to_R.html"> 
<span class="menu-text">Intro_to_R</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../IntroGR/pages/GRanges_gtf.html"> 
<span class="menu-text">Working with genomic coordinates</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_principles.html"> 
<span class="menu-text">ChIP-seq principles</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../ChIP-seq_analysis/pages/ChIP-seq_analysis_part1.html" aria-current="page"> 
<span class="menu-text">ChIP-seq analysis Part1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_analysis_part2.html"> 
<span class="menu-text">ChIP-seq analysis Part2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../Gviz/pages/Genomic_overviews_with_Gviz.html"> 
<span class="menu-text">Genomic visualization</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../PBMC3k/pages/scRNA-seq_PBMC3k.html"> 
<span class="menu-text">scRNA-seq_PBMC3k</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/paganilab"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/LabPagani"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#objectives" id="toc-objectives" class="nav-link active" data-scroll-target="#objectives">Objectives</a></li>
  <li><a href="#install-the-packages" id="toc-install-the-packages" class="nav-link" data-scroll-target="#install-the-packages">Install the packages</a></li>
  <li><a href="#the-chip-seq-dataset" id="toc-the-chip-seq-dataset" class="nav-link" data-scroll-target="#the-chip-seq-dataset">The ChIP-seq dataset</a></li>
  <li><a href="#load-and-explore-the-dataset" id="toc-load-and-explore-the-dataset" class="nav-link" data-scroll-target="#load-and-explore-the-dataset">Load and explore the dataset</a></li>
  <li><a href="#savingloading-files" id="toc-savingloading-files" class="nav-link" data-scroll-target="#savingloading-files">Saving/Loading Files</a></li>
  <li><a href="#brief-recap-of-the-dataset" id="toc-brief-recap-of-the-dataset" class="nav-link" data-scroll-target="#brief-recap-of-the-dataset">Brief recap of the dataset</a></li>
  <li><a href="#data-normalization-removing-uninteresting-differences" id="toc-data-normalization-removing-uninteresting-differences" class="nav-link" data-scroll-target="#data-normalization-removing-uninteresting-differences">Data Normalization: removing uninteresting differences</a></li>
  <li><a href="#about-the-edger-package" id="toc-about-the-edger-package" class="nav-link" data-scroll-target="#about-the-edger-package">About the <code>edgeR</code> Package</a>
  <ul class="collapse">
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  </ul></li>
  <li><a href="#create-a-dgelist-object-with-counts-sample-metadata-and-design-formula" id="toc-create-a-dgelist-object-with-counts-sample-metadata-and-design-formula" class="nav-link" data-scroll-target="#create-a-dgelist-object-with-counts-sample-metadata-and-design-formula">Create a DGEList object with counts, sample metadata and design formula</a>
  <ul class="collapse">
  <li><a href="#behind-the-design-formula" id="toc-behind-the-design-formula" class="nav-link" data-scroll-target="#behind-the-design-formula">Behind The Design Formula</a></li>
  </ul></li>
  <li><a href="#filtering-data" id="toc-filtering-data" class="nav-link" data-scroll-target="#filtering-data">Filtering Data</a></li>
  <li><a href="#normalizing-count-data" id="toc-normalizing-count-data" class="nav-link" data-scroll-target="#normalizing-count-data">Normalizing Count Data</a>
  <ul class="collapse">
  <li><a href="#normalization-discussion" id="toc-normalization-discussion" class="nav-link" data-scroll-target="#normalization-discussion">Normalization discussion</a></li>
  </ul></li>
  <li><a href="#transforming-count-data-for-visualization-purposes" id="toc-transforming-count-data-for-visualization-purposes" class="nav-link" data-scroll-target="#transforming-count-data-for-visualization-purposes">Transforming Count Data for visualization purposes</a></li>
  <li><a href="#assessing-sample-to-sample-relationships" id="toc-assessing-sample-to-sample-relationships" class="nav-link" data-scroll-target="#assessing-sample-to-sample-relationships">Assessing Sample-to-sample Relationships</a>
  <ul class="collapse">
  <li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering">Hierarchical clustering</a></li>
  <li><a href="#sample-to-sample-distances" id="toc-sample-to-sample-distances" class="nav-link" data-scroll-target="#sample-to-sample-distances">Sample-to-sample Distances</a></li>
  <li><a href="#pairwise-clustering-using-euclidean-distances" id="toc-pairwise-clustering-using-euclidean-distances" class="nav-link" data-scroll-target="#pairwise-clustering-using-euclidean-distances">Pairwise clustering using Euclidean distances</a></li>
  <li><a href="#using-rcolorbrewer-palettes" id="toc-using-rcolorbrewer-palettes" class="nav-link" data-scroll-target="#using-rcolorbrewer-palettes">Using RColorBrewer palettes</a></li>
  <li><a href="#dimensionality-reduction-with-principal-component-analysis-pca" id="toc-dimensionality-reduction-with-principal-component-analysis-pca" class="nav-link" data-scroll-target="#dimensionality-reduction-with-principal-component-analysis-pca">Dimensionality reduction with Principal Component Analysis (PCA)</a></li>
  </ul></li>
  <li><a href="#differential-analysis" id="toc-differential-analysis" class="nav-link" data-scroll-target="#differential-analysis">Differential analysis</a>
  <ul class="collapse">
  <li><a href="#the-main-edger-function" id="toc-the-main-edger-function" class="nav-link" data-scroll-target="#the-main-edger-function">The Main <code>edgeR</code> Function</a></li>
  </ul></li>
  <li><a href="#extracting-and-examining-results" id="toc-extracting-and-examining-results" class="nav-link" data-scroll-target="#extracting-and-examining-results">Extracting and examining results</a>
  <ul class="collapse">
  <li><a href="#visualizing-results-with-md-plots" id="toc-visualizing-results-with-md-plots" class="nav-link" data-scroll-target="#visualizing-results-with-md-plots">Visualizing Results With MD Plots</a></li>
  <li><a href="#visualizing-results-with-volcano-plots" id="toc-visualizing-results-with-volcano-plots" class="nav-link" data-scroll-target="#visualizing-results-with-volcano-plots">Visualizing results with volcano plots</a></li>
  <li><a href="#visualizing-results-with-heatmaps" id="toc-visualizing-results-with-heatmaps" class="nav-link" data-scroll-target="#visualizing-results-with-heatmaps">Visualizing results with heatmaps</a></li>
  <li><a href="#visualizing-results-with-violin-and-box-plots" id="toc-visualizing-results-with-violin-and-box-plots" class="nav-link" data-scroll-target="#visualizing-results-with-violin-and-box-plots">Visualizing results with violin and box plots</a></li>
  <li><a href="#save-dds-object" id="toc-save-dds-object" class="nav-link" data-scroll-target="#save-dds-object">Save dds object</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">ChIP-seq_analysis_part1</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 26, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="objectives" class="level2">
<h2 class="anchored" data-anchor-id="objectives">Objectives</h2>
<ul>
<li>Introduction to ChIP-seq</li>
<li>Dataset introduction and exploration</li>
<li>Data normalization with <code>edgeR</code></li>
<li>Diagnostic and exploratory analysis</li>
<li>Resources for ChIP-seq and other omics data</li>
</ul>
</section>
<section id="install-the-packages" class="level2">
<h2 class="anchored" data-anchor-id="install-the-packages">Install the packages</h2>
<p>The analyses we are going to perform require specific <em>functions</em> that are not included in the basic set of functions in R. These functions are collected in specific <em>packages</em>. R packages are extensions to the R programming language that contain code, data and documentation which help us performing standardized workflows. In the chunk below, we instruct R to install the packages that we will need later on through the workshop.</p>
<blockquote class="blockquote">
<p>We have already installed the necessary packages. However, if a library is not present copy the relevant command line and paste it to your R script.</p>
</blockquote>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # Install packages from Bioconductor</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># if (!require("BiocManager", quietly = TRUE))</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   install.packages("BiocManager")</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># # Install packages from CRAN</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># #install.packages("tidyr")</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># #install.packages("dplyr")</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># #install.packages("googledrive")</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># # For differential analysis</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install("vsn")</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install("edgeR")</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("statmod")</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># # For visualizations</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("hexbin")</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("pheatmap")</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co"># #install.packages("RColorBrewer")</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># #install.packages("ggrepel")</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="co"># #install.packages("circlize")</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="co"># # For downstream analysis</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("gprofiler2")</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="co"># # Clean garbage</span></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="co"># gc()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>During installation of packages, you will see many messages being displayed on your R console, you usually need to pay attention to them if they are <strong>red and specify an error</strong>!</p>
<p>If you encounter any of the messages below during installation, follow this procedure here:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># R asks for package updates, answer "n" and type enter</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Question displayed:</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>Update all<span class="sc">/</span>some<span class="sc">/</span>none? [a<span class="sc">/</span>s<span class="sc">/</span>n]<span class="sc">:</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Answer to type:  </span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>n</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># R asks for installation from binary source, answer "no" and type enter</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Question displayed:</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>Do you want to install from sources the packages which need compilation? (Yes<span class="sc">/</span>no<span class="sc">/</span>cancel)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Answer to type:</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>no</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="the-chip-seq-dataset" class="level2">
<h2 class="anchored" data-anchor-id="the-chip-seq-dataset">The ChIP-seq dataset</h2>
<p>The dataset we will analyze comes from this paper published from our lab in 2021:</p>
<blockquote class="blockquote">
<p>Della Chiara, Gervasoni, Fakiola, Godano et al., 2021 - Epigenomic landscape of human colorectal cancer unveils an aberrant core of pan-cancer enhancers orchestrated by YAP/TAZ</p>
</blockquote>
<p>Its main object was to explore the way cancer cells use enhancers as opposed to nonmalignant cells, and we did so by using <strong>ChIP-seq on histone marks</strong> extensively. We generated <strong>organoid lines to model colorectal tumors</strong>, analyzed their epigenome, and compared it to the ATAC-seq-derived chromatin activity data of multiple other malignancies. We found a small number of <strong>enhancer sequences were conserved</strong> across all these types, and identified a transcription factor that linked them - the YAP/TAZ pair - that when perturbed would cause the death of malignant, but not normal, organoids, therefore representing a potential pan-cancer therapeutic target.</p>
<p><img src="../pics/graphical_abstract_yap_taz_1.png" class="img-fluid"></p>
<p><img src="../pics/graphical_abstract_yap_taz_2.png" class="img-fluid"></p>
<blockquote class="blockquote">
<p>âœ… In this workshop, we will re-analyze a portion of the ChIP-seq dataset used in the paper, pertaining to <strong>Histone 3 Lysine 27 Acetylation</strong> - a marker of activation present at active regulatory sequences, enhancers and promoters alike.</p>
</blockquote>
</section>
<section id="load-and-explore-the-dataset" class="level2">
<h2 class="anchored" data-anchor-id="load-and-explore-the-dataset">Load and explore the dataset</h2>
<p>In order to speed up the computations and keep memory usage low, we have <strong>subset the dataset only to chromosome 12</strong>. It will be interesting to see if we can recapitulate similar analyses by looking only at a single chromosome ðŸ¤“.</p>
<p>Data are in this public Google Drive <a href="https://drive.google.com/drive/folders/1tRXeL1ZSyZ8wSR_kUWZVJov1_KHCFHjb?usp=sharing">folder</a>. You will find:</p>
<ol type="1">
<li><code>raw_counts_chr12.matrix</code>: the <strong>peak by sample matrix</strong> containing the number of reads detected for each peak in each sample.</li>
<li><code>colData.txt</code>: a tabular file containing our <strong>metadata</strong> related to columns of the count table, which contains different info about our samples (e.g.&nbsp;the treatment, the sample origin, etc.).</li>
<li><code>peakset.bed</code>: a BED file with the chromosome locations of the entire consensus peakset (i.e.&nbsp;all chromosomes).</li>
<li><code>recurrence_chr12.matrix</code>: this is a table where you can find if any of the intervals that constitute the consensus peakset is also called as peak in each separate sample. Remember that when a consensus peakset is created, usually genomic intervals called as peaks in at least 2 out of all the samples are kept in the consensus. Therefore, in the consensus you might find intervals that are called in 3 or in all the samples. It might be interesting to check how the peaks are distributed.</li>
<li><code>dba_Korg_Ntissue_homer_annot.txt</code>: this file contains information about the annotation of each consensus peakset to the nearest gene TSS. You will understand later its usage.</li>
<li><code>Korg_UP_regions_results.txt</code> and <code>Ncr_UP_regions_results.txt</code>: these files store the differential analysis results for the entire peakset and we will need them to perform downstream functional analyses.</li>
</ol>
<p>Open the Google Drive link above in teh browser, check the presence of the files and download them on your computer. Create a new directory within C:\Users\student&nbsp;to store the data.</p>
<p>Create a new R notebook and save it in the same folder. Check the path to the current working directory and the list of files you have just downloaded. Modify the code below according to the directory where you downloaded the data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">"./"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Import the counts and metadata information for the samples.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../MBMM25_uploadata/raw_counts_chr12.matrix"</span>,<span class="at">text =</span> ., <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../MBMM25_uploadata/colData.txt"</span>, <span class="at">text =</span> ., <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now explore the data that we have just loaded in the current R session to familiarize with it.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check out the counts</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(counts, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2157</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_1990</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2010</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2163</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2204</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2212</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2216</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2222</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2288</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2303</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2298</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2145</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058021</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058022</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058023</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">reg_6364</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">97</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">85</td>
<td style="text-align: right;">34</td>
<td style="text-align: right;">185</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6365</td>
<td style="text-align: right;">26</td>
<td style="text-align: right;">20</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">58</td>
<td style="text-align: right;">67</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">65</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">43</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">79</td>
<td style="text-align: right;">56</td>
<td style="text-align: right;">216</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6366</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">65</td>
<td style="text-align: right;">44</td>
<td style="text-align: right;">169</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6367</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">132</td>
<td style="text-align: right;">66</td>
<td style="text-align: right;">188</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6368</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">56</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6369</td>
<td style="text-align: right;">184</td>
<td style="text-align: right;">304</td>
<td style="text-align: right;">215</td>
<td style="text-align: right;">661</td>
<td style="text-align: right;">278</td>
<td style="text-align: right;">519</td>
<td style="text-align: right;">466</td>
<td style="text-align: right;">435</td>
<td style="text-align: right;">500</td>
<td style="text-align: right;">267</td>
<td style="text-align: right;">1045</td>
<td style="text-align: right;">297</td>
<td style="text-align: right;">1950</td>
<td style="text-align: right;">1562</td>
<td style="text-align: right;">3860</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6370</td>
<td style="text-align: right;">28</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">92</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">50</td>
<td style="text-align: right;">49</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">57</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">91</td>
<td style="text-align: right;">44</td>
<td style="text-align: right;">136</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6371</td>
<td style="text-align: right;">141</td>
<td style="text-align: right;">290</td>
<td style="text-align: right;">281</td>
<td style="text-align: right;">673</td>
<td style="text-align: right;">196</td>
<td style="text-align: right;">398</td>
<td style="text-align: right;">438</td>
<td style="text-align: right;">303</td>
<td style="text-align: right;">451</td>
<td style="text-align: right;">188</td>
<td style="text-align: right;">666</td>
<td style="text-align: right;">222</td>
<td style="text-align: right;">1793</td>
<td style="text-align: right;">1409</td>
<td style="text-align: right;">3416</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6372</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">26</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">39</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">39</td>
<td style="text-align: right;">25</td>
<td style="text-align: right;">222</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6373</td>
<td style="text-align: right;">63</td>
<td style="text-align: right;">86</td>
<td style="text-align: right;">112</td>
<td style="text-align: right;">293</td>
<td style="text-align: right;">70</td>
<td style="text-align: right;">168</td>
<td style="text-align: right;">55</td>
<td style="text-align: right;">99</td>
<td style="text-align: right;">268</td>
<td style="text-align: right;">135</td>
<td style="text-align: right;">187</td>
<td style="text-align: right;">31</td>
<td style="text-align: right;">580</td>
<td style="text-align: right;">229</td>
<td style="text-align: right;">894</td>
</tr>
</tbody>
</table>


</div>
</div>
<p>We can then check the shape of our counts table (i.e.&nbsp;how many different peaks we are detecting and how many different samples?)</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many rows and columns does our count table have?</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] 1581   15</code></pre>
</div>
</div>
<p>We can see that our table contains count information for 1581 peaks and 15 samples.</p>
<p>We can also inspect the <strong>metadata</strong> from the samples which is stored in the <code>samples</code> object we created above.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What does the table look like?</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th">groups</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">sizeFactor</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Tissue</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">SampleID</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Treatment</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">SQ_2157</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.489936</td>
<td style="text-align: left;">4KE</td>
<td style="text-align: left;">SQ_2157</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_1990</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.758060</td>
<td style="text-align: left;">8KE</td>
<td style="text-align: left;">SQ_1990</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2010</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.760352</td>
<td style="text-align: left;">10KE</td>
<td style="text-align: left;">SQ_2010</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2163</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.425892</td>
<td style="text-align: left;">13KE</td>
<td style="text-align: left;">SQ_2163</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2204</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">1.863534</td>
<td style="text-align: left;">18KE</td>
<td style="text-align: left;">SQ_2204</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2212</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.580778</td>
<td style="text-align: left;">11KW</td>
<td style="text-align: left;">SQ_2212</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2216</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.159358</td>
<td style="text-align: left;">24KE</td>
<td style="text-align: left;">SQ_2216</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2222</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.203927</td>
<td style="text-align: left;">22KE</td>
<td style="text-align: left;">SQ_2222</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2288</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.402964</td>
<td style="text-align: left;">36KE</td>
<td style="text-align: left;">SQ_2288</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2303</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">2.430445</td>
<td style="text-align: left;">41KE</td>
<td style="text-align: left;">SQ_2303</td>
<td style="text-align: left;">K_org</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2298</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">3.756529</td>
<td style="text-align: left;">CR_41_mp</td>
<td style="text-align: left;">SQ_2298</td>
<td style="text-align: left;">N_crypts</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2145</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">2.804850</td>
<td style="text-align: left;">CR_28_mp</td>
<td style="text-align: left;">SQ_2145</td>
<td style="text-align: left;">N_crypts</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GSM2058021</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">1.252530</td>
<td style="text-align: left;">CR_28</td>
<td style="text-align: left;">GSM2058021</td>
<td style="text-align: left;">N_crypts</td>
</tr>
<tr class="even">
<td style="text-align: left;">GSM2058022</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">1.000000</td>
<td style="text-align: left;">CR_29</td>
<td style="text-align: left;">GSM2058022</td>
<td style="text-align: left;">N_crypts</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GSM2058023</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">2.476903</td>
<td style="text-align: left;">CR_37</td>
<td style="text-align: left;">GSM2058023</td>
<td style="text-align: left;">N_crypts</td>
</tr>
</tbody>
</table>


</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What is the shape of this samples table?</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] 15  5</code></pre>
</div>
</div>
<p>In this case, this <code>samples</code> table has as many rows (15) as there are samples (which in turn is equal to the number of columns in the <code>counts</code> table), with columns containing different types of information related to each of the sample in the analysis.</p>
<p>We can verify that the samples in the <code>counts</code> table matches the samples in the <code>metadata</code> table.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">colnames</span>(counts)<span class="sc">==</span><span class="fu">rownames</span>(samples))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>
TRUE 
  15 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What if we reverse the order of the samples? </span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">colnames</span>(counts)<span class="sc">==</span><span class="fu">rev</span>(<span class="fu">rownames</span>(samples)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>
FALSE  TRUE 
   14     1 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">#colnames(counts)==rev(rownames(samples))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="savingloading-files" class="level2">
<h2 class="anchored" data-anchor-id="savingloading-files">Saving/Loading Files</h2>
<p><strong>Letâ€™s save this object with samples information</strong> in a file format where columns are separated by commas. You might be familiar with this format if you have worked quite a bit in Excel. In <code>R</code>, we can save tabular data with the <code>write.table()</code> function specifying the location (the file name) we want. This is <strong>useful in the case our <code>R</code> session dies or we decide to interrupt it</strong>. In this case we will not have to run the whole analysis from the beginning and we can just source the file and load it!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(samples, <span class="st">"../results/samples_table.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can <strong>load the object back into the current session</strong> by using the following code line:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../results/samples_table.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Save also the counts</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data again</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../results/counts_table.csv"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../results/samples_table.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>! Troubleshoot</strong></p>
<blockquote class="blockquote">
<p>Where are you? Where is the file located? Check the current directory and navigate through the folders with <code>getwd()</code> and <code>dir()</code></p>
</blockquote>
</section>
<section id="brief-recap-of-the-dataset" class="level2">
<h2 class="anchored" data-anchor-id="brief-recap-of-the-dataset">Brief recap of the dataset</h2>
<p>First of all, letâ€™s make a recap on the dataset that we are going to analyze.</p>
<p>Each row of the <code>counts</code> table corresponds to a genomic interval within chromosome 12. We discussed about a standard analysis that requires a peak caller (such as <strong>MACS</strong>) used on the aligned reads to call peaks. While having called peaks for the sample is useful in a variety of ways, it is possible to perform a quantitative analysis without using called peaks. For example, in some cases, the regions of interest may be known in advance (such as a list of known gene promoters).</p>
<p>This is actually our case, because the genomic regions that weâ€™ll analyze are the active enhancers defined by <a href="https://www.nature.com/articles/nprot.2017.124">ChromHMM</a>. Using this machine-learning approach, we have characterized the chromatin states in CRC organoids and normal colon tissues, by combining multiple histone mark ChIP-seq profiles including H3K27 acetylation. Among the different chromatin states, we were able to identify <strong>genomic regions with active enhancer features</strong>. We have then generated a consensus peakset by merging peaks called in at least two samples.</p>
<center>
<p><img src="../pics/Consensus_peaks.png" class="img-fluid" width="400"></p>
</center>
<p>On these consensus peakset, we then counted reads from ChIP-seq on H3K27Ac for each sample. <em>NB: This is a compute-intensive step that might have taken too long to run in the workshop!</em> Thus we have already perfomed this step and the table represents the counts for each consensus region on chr12 across samples.</p>
</section>
<section id="data-normalization-removing-uninteresting-differences" class="level2">
<h2 class="anchored" data-anchor-id="data-normalization-removing-uninteresting-differences">Data Normalization: removing uninteresting differences</h2>
<p>Since we want to highlight interesting biological differences between groups of samples in our dataset, the first thing to do is to ensure that we take into account and minimize all of the <em>uninteresting</em> differences between samples. This is accomplished by <strong>normalizing the data</strong>, a mandatory step before any differential analysis, if we want to make counts <strong>comparable both across samples and across different peaks</strong>, without any unwanted <em>bias</em>.</p>
<p>The main differences in the dataset that we want to normalize for are:</p>
<ol type="1">
<li>the <strong>Sequencing Depth</strong> (expressed as the total amount of uniquely mapped reads in a sample): even between samples that were sequenced in the same sequencing run as part of the same experiments (but even more likely if samples have been generated in <strong>multiple <em>batches</em></strong>), differences in the sequencing depth are expected. Itâ€™s important to take this into account because differences in sequencing depth can erroneously lead to the perception of genomic intervals having differential signals.</li>
</ol>
<center>
<img src="https://hbctraining.github.io/DGE_workshop/img/normalization_methods_depth.png" width="500">
</center>
<ol start="2" type="1">
<li><strong>Genomic interval length</strong>: this is a potential bias if we want to <strong>compare across group of intervals</strong> other than across samples. Indeed, the larger the genomic region, the higher the amount of reads that will be counted on it, and viceversa.</li>
</ol>
<center>
<img src="https://hbctraining.github.io/DGE_workshop/img/normalization_methods_length.png" width="300">
</center>
<p>During the years, <strong>many approaches</strong> to data normalization have been attempted. In the table below, we summarized three common methods that can be employed to account for differences in library size (i.e.&nbsp;the sequencing depth) and the genomic interval length.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 33%">
<col style="width: 33%">
<col style="width: 33%">
</colgroup>
<thead>
<tr class="header">
<th><strong>Normalization Method</strong></th>
<th><strong>Accounted Factors</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>CPM (counts per million)</td>
<td>Sequencing depth</td>
<td>Counts scaled by total read number</td>
</tr>
<tr class="even">
<td>TPM (transcripts per million)</td>
<td>Sequencing depth and gene length</td>
<td>Counts per length of transcript (kb) per million mapped reads</td>
</tr>
<tr class="odd">
<td>FPKM/RPKM</td>
<td>Sequencing depth and gene length</td>
<td>Counts per kilobase of exon mapped reads per million mapped reads</td>
</tr>
</tbody>
</table>
<blockquote class="blockquote">
<p>Notice the term â€œtranscriptâ€ for TPM or â€œexonâ€ in the method descriptions. Historically, these methods have been developed for RNA-seq data, dealing with <strong>gene expression counts</strong>, but they can be extended for other sequencing technologies. See also the <em>Normalization Discussion</em> section.</p>
</blockquote>
<p>CPM, TPM and FPKM/RPKM are considered <strong>simple normalization methods</strong>, that can be useful to scale NGS data. However, newer and more sophisticated approaches have been developed to take into account technical variability and sample-specific biases. Two of these are the <strong>DESeq2â€™s median of ratios</strong> and the <strong>edgeRâ€™s trimmed mean of M values</strong>. These are indeed more <strong>advanced statistical methods</strong> used for normalization and differential analysis in multiple sequencing data, including RNA-seq and ATAC-seq.</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th><strong>Normalization Method</strong></th>
<th><strong>Accounted Factors</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><em>DESeq2</em>â€™s median of ratios<sup><a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">1</a></sup></td>
<td>Sequencing depth and RNA composition</td>
<td>Counts are divided by a sample-specific size factor</td>
</tr>
<tr class="even">
<td><em>edgeR</em>â€™s trimmed mean of M values<sup><a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">2</a></sup></td>
<td>Sequencing depth, RNA composition and gene length</td>
<td>Weighted trimmed mean of the log ratio of expression between samples</td>
</tr>
</tbody>
</table>
<p>As you can see from the table, with respect to the other methods, these ones can correct for one additional unwanted difference across libraries: the <strong>Composition Bias</strong>. As the name suggests, composition biases are formed when there are differences in the composition of sequences across libraries. <strong>Highly enriched regions consume more sequencing resources</strong> and thereby <strong>suppress the representation of other regions</strong>. Scaling by library size fails to correct for this as composition biases can still occur in libraries of the same size.</p>
<center>
<img src="https://hbctraining.github.io/DGE_workshop/img/normalization_methods_composition.png" width="500">
</center>
<p>Out of all these, we will use one of the more advanced ones provided in the <code>edgeR</code> package which will be now introduced.</p>
</section>
<section id="about-the-edger-package" class="level2">
<h2 class="anchored" data-anchor-id="about-the-edger-package">About the <code>edgeR</code> Package</h2>
<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p>One of the aims of this workshop is to understand which genomic regions have a <strong>differential signal for H3K27Ac</strong> in our dataset, that is, have a <strong>difference in enhancer activty</strong>. In particular, we want to identify regions that are gaining more of this histone modification in CRC organoid samples, that might be <strong>enhancer regions with increased activity compared to normal organoids</strong>. To do so, we will compare H3K27Ac levels across samples and <strong>statistically assess and quantify differences</strong> arising between the conditions represented by our categories of samples (i.e., <strong>normal colon crypts and CRC organoids</strong>). <code>EdgeR</code> is a widely used package for normalization and differential analysis on bulk sequencing data.</p>
<blockquote class="blockquote">
<p><strong>Detailed explanations of the statistical procedures implemented in the package are available in the packageâ€™s <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html"><em>vignette</em></a>.</strong></p>
</blockquote>
<p>We will start by loading the packages.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the package</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can have a look at all the functions contained in the <code>edgeR</code> package by typing the following command:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This should open a popup window in the lower right part of the screen displaying the functions in the package</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>??edgeR</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="create-a-dgelist-object-with-counts-sample-metadata-and-design-formula" class="level2">
<h2 class="anchored" data-anchor-id="create-a-dgelist-object-with-counts-sample-metadata-and-design-formula">Create a DGEList object with counts, sample metadata and design formula</h2>
<p>In order for the package to read and understand our data and correctly perform the analysis, we need to <strong>organize our data</strong> in a way that the functions of the package can handle. This <strong>new object</strong> that we are going to create is called <code>DGEList</code> and there is a utility function to create one starting from the <em>ingredients</em> we currently have, (1) a table of counts (our counts object), (2) a table with sample information (our samples object) and (3) what comparisons we want to perform, this is called a <strong>design formula</strong>.</p>
<section id="behind-the-design-formula" class="level3">
<h3 class="anchored" data-anchor-id="behind-the-design-formula">Behind The Design Formula</h3>
<p>The design formula should <strong>contain the name of a column of interest in our table of samples</strong> (that we can call <strong>factor</strong>) which stores the information related to the <strong>levels</strong> (or categories) we want to contrast. Letâ€™s say that we have a dataset with two conditions (condition_1 vs condition_2) that we want to compare. The samples table will look like this, with three replicates for each of the two conditions:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Sample Code</th>
<th>Patient</th>
<th>Condition</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>SA1</td>
<td>Pt1</td>
<td>Condition_1</td>
</tr>
<tr class="even">
<td>SA2</td>
<td>Pt2</td>
<td>Condition_1</td>
</tr>
<tr class="odd">
<td>SA3</td>
<td>Pt3</td>
<td>Condition_1</td>
</tr>
<tr class="even">
<td>SA4</td>
<td>Pt1</td>
<td>Condition_2</td>
</tr>
<tr class="odd">
<td>SA5</td>
<td>Pt2</td>
<td>Condition_2</td>
</tr>
<tr class="even">
<td>SA6</td>
<td>Pt3</td>
<td>Condition_2</td>
</tr>
</tbody>
</table>
<section id="paired-analyses" class="level4">
<h4 class="anchored" data-anchor-id="paired-analyses">Paired Analyses</h4>
<p>The optimal setting for the analysis (decided experimentally) is to have paired samples. This might be a somewhat difficult concept to grasp, but for our table above this means that every Patient <strong>contributes equally to the two categories</strong> in the Condition columns that we are interested in. In this setting, we are fully capable of exploiting the statistics behind the tools we use for differential analysis by correcting for the uninteresting differences arising between patients. This aspect greatly helps the analysis and improves the statistical validity of the results.</p>
<p>Remember, this is something achieved by deciding the experimental setting <strong>beforehand</strong>! Ideally this should be done through a collaborative effort between bioinformaticians/statisticians and bench scientists!</p>
<p>If we are interested in performing a differential expression analysis comparing <code>condition_1</code> versus <code>condition_2</code>, then our design formula should specify the <code>Condition</code> factor.</p>
<blockquote class="blockquote">
<p>ðŸ’¡ <strong>What is the column that we are interested in when specifying the design formula using in our <code>samples</code> table?</strong></p>
</blockquote>
<blockquote class="blockquote">
<p>ðŸ’¡ <strong>What is the factor and levels in our case?</strong></p>
</blockquote>
<p>Now that we also understand the design formula, we can create the <code>DGEList</code> object with the data that we loaded beforehand, but first we need to check that the columns of the <code>counts</code> table are in the same order of the rows of the <code>sample</code> table, this is important since <strong>we want to be sure that the right levels of expression are associated to the right sample</strong>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">rownames</span>(samples) <span class="sc">==</span> <span class="fu">colnames</span>(counts))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] TRUE</code></pre>
</div>
</div>
<p>Further, we will get rid of non-useful columns in our sample metadata:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">select</span>(samples, <span class="sc">-</span><span class="fu">c</span>(sizeFactor, Treatment))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And now we can build the object:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a design formula. Notice that we are creating an object type known as "factor"</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>sample_group <span class="ot">&lt;-</span> <span class="fu">factor</span>(samples<span class="sc">$</span>groups, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"N_crypts"</span>, <span class="st">"K_org"</span>))</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> sample_group)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a `DGEList` object and call it dds</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts =</span> counts, <span class="at">samples =</span> samples)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's save the `design` in the dds object</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>dds<span class="sc">$</span>design <span class="ot">&lt;-</span> design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Remember that <code>dds</code> is just a <strong>list</strong> in R which can be updated with different elements.</p>
<p>We can remove the counts table from our R environment since that information is stored in our DGEList object now. This is useful to save on memory space!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove original `counts` table to save memory space</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(counts)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Great! You have created a DGEList object which we called dds, this <strong>contains all the information related to the counts table and the sample information table in one spot</strong>. We can have a look at the sample information table and the counts table in the dds object like so:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the table with sample information</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dds<span class="sc">$</span>samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th">group</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">lib.size</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">norm.factors</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">groups</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Tissue</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">SampleID</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">SQ_2157</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">69284</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">4KE</td>
<td style="text-align: left;">SQ_2157</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_1990</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">96388</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">8KE</td>
<td style="text-align: left;">SQ_1990</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2010</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">112029</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">10KE</td>
<td style="text-align: left;">SQ_2010</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2163</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">169859</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">13KE</td>
<td style="text-align: left;">SQ_2163</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2204</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">108070</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">18KE</td>
<td style="text-align: left;">SQ_2204</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2212</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">162626</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">11KW</td>
<td style="text-align: left;">SQ_2212</td>
</tr>
</tbody>
</table>


</div>
</div>
<p>We can see that some new columns were added to the samples table present in our DGEList object when we created it (the <code>group</code>, <code>lib.size</code>, <code>norm.factors</code> columns)! These will be used by edgeR later on for data normalization!</p>
<blockquote class="blockquote">
<p>Can you compute the library sizes yourself and compare them to the ones generated by edgeR?</p>
</blockquote>
<p>We can also take a look at the table containing the counts, which is just another element of our DGEList object:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the table with count information</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dds<span class="sc">$</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2157</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_1990</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2010</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2163</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2204</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2212</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2216</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2222</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2288</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2303</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2298</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2145</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058021</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058022</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058023</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">reg_6364</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">97</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">85</td>
<td style="text-align: right;">34</td>
<td style="text-align: right;">185</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6365</td>
<td style="text-align: right;">26</td>
<td style="text-align: right;">20</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">58</td>
<td style="text-align: right;">67</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">65</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">43</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">79</td>
<td style="text-align: right;">56</td>
<td style="text-align: right;">216</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6366</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">65</td>
<td style="text-align: right;">44</td>
<td style="text-align: right;">169</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6367</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">132</td>
<td style="text-align: right;">66</td>
<td style="text-align: right;">188</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6368</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">56</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6369</td>
<td style="text-align: right;">184</td>
<td style="text-align: right;">304</td>
<td style="text-align: right;">215</td>
<td style="text-align: right;">661</td>
<td style="text-align: right;">278</td>
<td style="text-align: right;">519</td>
<td style="text-align: right;">466</td>
<td style="text-align: right;">435</td>
<td style="text-align: right;">500</td>
<td style="text-align: right;">267</td>
<td style="text-align: right;">1045</td>
<td style="text-align: right;">297</td>
<td style="text-align: right;">1950</td>
<td style="text-align: right;">1562</td>
<td style="text-align: right;">3860</td>
</tr>
</tbody>
</table>


</div>
</div>
<blockquote class="blockquote">
<p>ðŸ’¡ In R, list elements are accessible with the <code>$</code> accessor. Our dds object is indeed a list made up of three elements, the counts table, the samples table and the design table, these are accessible using <code>$</code> like we did above.</p>
</blockquote>
</section>
</section>
</section>
<section id="filtering-data" class="level2">
<h2 class="anchored" data-anchor-id="filtering-data">Filtering Data</h2>
<p>Genes with very low counts across all libraries provide little evidence for differential expression. In addition, the pronounced discreteness of these counts interferes with some of the statistical approximations that are used later in the pipeline. These genes should be filtered out prior to further analysis. Users can have their own definition of expression but in general counts are dropped if they are not expressed in a minimum number of samples based on the minimum group size or if it does not achieve a minimum number of counts across all samples.</p>
<p>edgeR provides the <code>filterByExpr()</code> function that retains genes with sufficiently large counts for statistical analysis.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Dimensions of the object before filtering</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] 1581   15</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform filtering, subset the dds object and check the dimensions</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dds, <span class="at">group=</span>dds<span class="sc">$</span>samples<span class="sc">$</span>group)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> dds[keep, , keep.lib.sizes<span class="ot">=</span><span class="cn">FALSE</span>]</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] 1490   15</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many regions were filterd out?</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>counts[<span class="sc">!</span>keep,] <span class="sc">%&gt;%</span> <span class="fu">dim</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] 91 15</code></pre>
</div>
</div>
<p>We can verify that the count distribution of the filtered regions is lower compared to those that were kept.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Density plot for filtered regions</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>counts[<span class="sc">!</span>keep,] <span class="sc">%&gt;%</span> </span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">Region=</span><span class="fu">rownames</span>(.)) <span class="sc">%&gt;%</span>  </span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">gather</span>(., samples, value, <span class="sc">-</span>Region) <span class="sc">%&gt;%</span> </span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>        <span class="co">#head()</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggplot</span>(., <span class="fu">aes</span>(<span class="fu">log2</span>(value), <span class="at">fill=</span>samples)) <span class="sc">+</span> <span class="fu">geom_density</span>() <span class="sc">+</span> <span class="fu">theme_light</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>Check the count distribution of the retained regions</p>
</blockquote>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It is more helpful to compare them on the same plot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>counts <span class="sc">%&gt;%</span> </span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">Region=</span><span class="fu">rownames</span>(.),</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">Keep=</span>keep) <span class="sc">%&gt;%</span>  </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">gather</span>(., value, variable, <span class="sc">-</span><span class="fu">c</span>(Region,Keep)) <span class="sc">%&gt;%</span> </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>        <span class="co">#head()</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggplot</span>(., <span class="fu">aes</span>(<span class="fu">log2</span>(variable), <span class="at">fill=</span>Keep)) <span class="sc">+</span> <span class="fu">geom_density</span>() <span class="sc">+</span> <span class="fu">theme_light</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>Try to create a boxplot instead of a histogram.</p>
</blockquote>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="normalizing-count-data" class="level2">
<h2 class="anchored" data-anchor-id="normalizing-count-data">Normalizing Count Data</h2>
<p>As we have discussed above, normalization is an integral step to the downstream analysis and necessary for differential comparison. In this section we will normalize our data using the <code>calcNormFactors</code> function of the package. As we have previously introduced, edgeR uses the <strong>trimmed mean of M-values (TMM) method</strong> to calculate a set of <strong>size factors</strong> to <strong>minimize the log-fold change differences occurring between samples</strong> (<em>uninteresting</em>) for the majority of genomic intervals. The <strong>counts for each sample get then multiplied by the scaling factors</strong> to generate what is referred to as <em>effective library size</em>, which will be used for all downstream analyses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Call the function to normalize count data</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can check the values of the computed size factors by doing the following, note how there are as many size factors as there are samples and they are inserted in a column of the samples table named <code>norm.factors</code> in our <code>DGEList</code> object:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>dds<span class="sc">$</span>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th">group</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">lib.size</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">norm.factors</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">groups</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Tissue</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">SampleID</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">SQ_2157</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">68628</td>
<td style="text-align: right;">0.7617531</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">4KE</td>
<td style="text-align: left;">SQ_2157</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_1990</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">95645</td>
<td style="text-align: right;">0.9517516</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">8KE</td>
<td style="text-align: left;">SQ_1990</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2010</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">111383</td>
<td style="text-align: right;">1.0635417</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">10KE</td>
<td style="text-align: left;">SQ_2010</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2163</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">168819</td>
<td style="text-align: right;">1.0359859</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">13KE</td>
<td style="text-align: left;">SQ_2163</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2204</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">107640</td>
<td style="text-align: right;">0.9422863</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">18KE</td>
<td style="text-align: left;">SQ_2204</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2212</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">161795</td>
<td style="text-align: right;">0.8542493</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">11KW</td>
<td style="text-align: left;">SQ_2212</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2216</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">166417</td>
<td style="text-align: right;">0.8582417</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">24KE</td>
<td style="text-align: left;">SQ_2216</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2222</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">125528</td>
<td style="text-align: right;">1.0216942</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">22KE</td>
<td style="text-align: left;">SQ_2222</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2288</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">192107</td>
<td style="text-align: right;">1.1007166</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">36KE</td>
<td style="text-align: left;">SQ_2288</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2303</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: right;">91710</td>
<td style="text-align: right;">0.8959298</td>
<td style="text-align: left;">K_org</td>
<td style="text-align: left;">41KE</td>
<td style="text-align: left;">SQ_2303</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SQ_2298</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">105375</td>
<td style="text-align: right;">1.3540033</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: left;">CR_41_mp</td>
<td style="text-align: left;">SQ_2298</td>
</tr>
<tr class="even">
<td style="text-align: left;">SQ_2145</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">31148</td>
<td style="text-align: right;">1.2742282</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: left;">CR_28_mp</td>
<td style="text-align: left;">SQ_2145</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GSM2058021</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">175916</td>
<td style="text-align: right;">1.0590448</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: left;">CR_28</td>
<td style="text-align: left;">GSM2058021</td>
</tr>
<tr class="even">
<td style="text-align: left;">GSM2058022</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">119435</td>
<td style="text-align: right;">0.9694495</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: left;">CR_29</td>
<td style="text-align: left;">GSM2058022</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GSM2058023</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: right;">285704</td>
<td style="text-align: right;">1.0153128</td>
<td style="text-align: left;">N_crypts</td>
<td style="text-align: left;">CR_37</td>
<td style="text-align: left;">GSM2058023</td>
</tr>
</tbody>
</table>


</div>
</div>
<blockquote class="blockquote">
<p>ðŸ’¡ NOTE: Although <code>edgeR</code> does not use normalized counts as input for the differential analysis (the normalization process happens inside automatically), the normalized counts we just generated are useful when plotting results and performing clustering.</p>
</blockquote>
<section id="normalization-discussion" class="level3">
<h3 class="anchored" data-anchor-id="normalization-discussion">Normalization discussion</h3>
<p>Normalization of experimental data is particularly important in ChIP-seq (and ATAC-seq) analysis, and may require <strong>more careful consideration than needed for RNA-seq analysis</strong>. This is because the range of ChIP-seq experiments covers more cases than RNA-seq, which usually involve a similar set of possible expressed genes and/or transcripts, many of which are not expected to significantly change expression. ChIP, ATAC, and similar enrichment-based sequencing data may not follow the assumptions inherent in popular methods for normalizing RNA-seq data, as well as exhibiting different types of efficiency and other biases.</p>
<section id="reference-reads" class="level4">
<h4 class="anchored" data-anchor-id="reference-reads">Reference reads</h4>
<p>The reads used as the basis for scaling or normalizing the data. These can be total reads per sample, a subset of reads, or even a reference distribution derived from all samples.</p>
<p>While in RNA-seq experiments the expression matrix is normalized based on the <strong>reads that uniquely overlap genes</strong> or transcripts, this does not necessarily apply to a count matrix based on a consensus peakset.</p>
<p>The <code>DiffBind</code> package <a href="https://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf"></a> is a tool that has been designed specifically for ChIP-seq data analysis. Using this tools, you can normalize your data using two different sets of reference reads:</p>
<ul>
<li><p>one uses all the reads (the <em>full</em> library size)</p></li>
<li><p>the other normalizes bases only on the total number of reads in the consensus peakset (namely, the <em>effective</em> library size).</p></li>
</ul>
<p>The major difference is that the full library size takes into account also â€œbackgroundâ€ reads, that are most of the reads in a ChIP-seq sample library! In brief, the reads that are used as the reference for normalizing can be as important as the normalization method itself (e.g.&nbsp;background vs.&nbsp;enriched consensus reads ). The fact that we consider the background might reflect in a less biased normalization calculation.</p>
<p><strong>Take home message:</strong> &gt; In this workshop, to make things less complex, we used a â€œcommonâ€ normalization method without any prior in-depth analysis on the best strategy to normalize these specific data in this context. But just keep in mind that some normalization methods might be more â€œcorrectâ€ than others in some circumstances.</p>
</section>
</section>
</section>
<section id="transforming-count-data-for-visualization-purposes" class="level2">
<h2 class="anchored" data-anchor-id="transforming-count-data-for-visualization-purposes">Transforming Count Data for visualization purposes</h2>
<p>After we have normalized our data, we need to perform a <strong>transformation</strong>. There are many ways to transform count data but all of them achieve the <strong>goal of removing the dependence of the <em>variance</em> on the <em>mean</em> counts across samples</strong> (something called <em>homoscedasticity</em>) in order to highlight interesting and biologically relevant expression trends even for genes expressed at lower values. This is because many downstream analyses (like PCA, clustering, distance calculation) assume homoscedasticity â€” equal variance across features. If we do not stabilize for the mean-variance relationship, these analyses will be dominated by highly expressed genes â€” even if those genes might not be biologically interesting. Low and moderately expressed genes (which might carry the biological signal) are ignored, because they contribute much less to the variance or distances.</p>
<p>We transform the data using a function provided in the <code>edgeR</code> package called <code>cpm()</code> which also performs a <strong>logarithmic transformation</strong> which has the effect of reshaping the data to achieve gene-wise distributions which resemble a normal distribution. Without getting too much into the details of the workings of the function, we will transform the data and then look at how the gene-wise relationship between the mean and variance in our normalized data changes <em>before and after the transformation</em>. The purpose of this procedure is to allow proper data visualization later in the analysis, <strong>the transformed data is NOT used for the differential expression analysis which instead starts from <em>raw counts</em></strong>!</p>
<p>The following code is used to plot the mean/standard deviation relationship of every gene before the transformation.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vsn)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot before data transformation</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="fu">meanSdPlot</span>(dds<span class="sc">$</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Check the distribution of counts, the mean and sd for each region.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Quantiles of counts</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>dds<span class="sc">$</span>counts <span class="sc">%&gt;%</span> <span class="fu">quantile</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>  0%  25%  50%  75% 100% 
   1    9   30   83 7000 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Quantiles of standard deviation across regions</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(dds<span class="sc">$</span>counts, <span class="dv">1</span>, sd) <span class="sc">%&gt;%</span> <span class="fu">quantile</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>         0%         25%         50%         75%        100% 
   6.023762   21.026570   37.602235   71.282102 1668.777091 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Quantiles of mean across regions</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(dds<span class="sc">$</span>counts, <span class="dv">1</span>, mean) <span class="sc">%&gt;%</span> <span class="fu">quantile</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>         0%         25%         50%         75%        100% 
   6.533333   20.550000   38.200000   83.700000 2323.133333 </code></pre>
</div>
</div>
<p>Transform the data and then plot the mean/standard deviation relationship after the transformation.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Transform the data with a log2 transform (watch how we create a new variable for it)</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>log2dds <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dds, <span class="at">log=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>log2dds <span class="sc">%&gt;%</span> <span class="fu">quantile</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>       0%       25%       50%       75%      100% 
 4.196226  6.533113  8.032246  9.339823 14.817662 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(log2dds, <span class="dv">1</span>, sd) <span class="sc">%&gt;%</span> <span class="fu">quantile</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>       0%       25%       50%       75%      100% 
0.3110070 0.9600368 1.2826842 1.6385618 3.8225115 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(log2dds, <span class="dv">1</span>, mean) <span class="sc">%&gt;%</span> <span class="fu">quantile</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>       0%       25%       50%       75%      100% 
 5.473874  6.704483  7.623939  8.872750 14.010599 </code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check out the transformed values (notice how we now have floating point values)</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(log2dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2157</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_1990</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2010</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2163</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2204</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2212</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2216</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2222</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2288</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2303</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2298</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2145</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058021</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058022</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058023</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">reg_6364</td>
<td style="text-align: right;">7.391532</td>
<td style="text-align: right;">4.693150</td>
<td style="text-align: right;">4.543824</td>
<td style="text-align: right;">9.153524</td>
<td style="text-align: right;">4.628941</td>
<td style="text-align: right;">6.767372</td>
<td style="text-align: right;">5.165425</td>
<td style="text-align: right;">6.760853</td>
<td style="text-align: right;">7.004273</td>
<td style="text-align: right;">7.943201</td>
<td style="text-align: right;">8.427468</td>
<td style="text-align: right;">6.853782</td>
<td style="text-align: right;">8.879940</td>
<td style="text-align: right;">8.269220</td>
<td style="text-align: right;">9.350105</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6365</td>
<td style="text-align: right;">9.000603</td>
<td style="text-align: right;">7.873964</td>
<td style="text-align: right;">7.923632</td>
<td style="text-align: right;">8.147764</td>
<td style="text-align: right;">9.196486</td>
<td style="text-align: right;">8.964709</td>
<td style="text-align: right;">8.426038</td>
<td style="text-align: right;">8.715365</td>
<td style="text-align: right;">8.332118</td>
<td style="text-align: right;">8.079257</td>
<td style="text-align: right;">8.304934</td>
<td style="text-align: right;">8.189953</td>
<td style="text-align: right;">8.777787</td>
<td style="text-align: right;">8.961506</td>
<td style="text-align: right;">9.568881</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6366</td>
<td style="text-align: right;">5.087972</td>
<td style="text-align: right;">6.829712</td>
<td style="text-align: right;">4.543824</td>
<td style="text-align: right;">5.620275</td>
<td style="text-align: right;">4.628941</td>
<td style="text-align: right;">4.467208</td>
<td style="text-align: right;">4.451886</td>
<td style="text-align: right;">6.118166</td>
<td style="text-align: right;">4.293732</td>
<td style="text-align: right;">4.757775</td>
<td style="text-align: right;">8.367501</td>
<td style="text-align: right;">5.324762</td>
<td style="text-align: right;">8.506873</td>
<td style="text-align: right;">8.625283</td>
<td style="text-align: right;">9.222714</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6367</td>
<td style="text-align: right;">5.087972</td>
<td style="text-align: right;">4.693150</td>
<td style="text-align: right;">4.543824</td>
<td style="text-align: right;">7.076313</td>
<td style="text-align: right;">5.112776</td>
<td style="text-align: right;">6.860139</td>
<td style="text-align: right;">5.422542</td>
<td style="text-align: right;">5.751377</td>
<td style="text-align: right;">4.605370</td>
<td style="text-align: right;">4.757775</td>
<td style="text-align: right;">8.567309</td>
<td style="text-align: right;">9.542125</td>
<td style="text-align: right;">9.498622</td>
<td style="text-align: right;">9.192004</td>
<td style="text-align: right;">9.372787</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6368</td>
<td style="text-align: right;">5.087972</td>
<td style="text-align: right;">4.693150</td>
<td style="text-align: right;">7.144706</td>
<td style="text-align: right;">6.808420</td>
<td style="text-align: right;">6.210186</td>
<td style="text-align: right;">6.446814</td>
<td style="text-align: right;">5.830150</td>
<td style="text-align: right;">4.503405</td>
<td style="text-align: right;">6.063995</td>
<td style="text-align: right;">6.242883</td>
<td style="text-align: right;">6.523221</td>
<td style="text-align: right;">6.499337</td>
<td style="text-align: right;">6.208006</td>
<td style="text-align: right;">7.738227</td>
<td style="text-align: right;">7.699948</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6369</td>
<td style="text-align: right;">11.787271</td>
<td style="text-align: right;">11.711810</td>
<td style="text-align: right;">10.837456</td>
<td style="text-align: right;">11.889581</td>
<td style="text-align: right;">11.428188</td>
<td style="text-align: right;">11.880287</td>
<td style="text-align: right;">11.678379</td>
<td style="text-align: right;">11.734101</td>
<td style="text-align: right;">11.216369</td>
<td style="text-align: right;">11.672566</td>
<td style="text-align: right;">12.841337</td>
<td style="text-align: right;">12.872236</td>
<td style="text-align: right;">13.355543</td>
<td style="text-align: right;">13.721191</td>
<td style="text-align: right;">13.701438</td>
</tr>
</tbody>
</table>


</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(counts) <span class="sc">%&gt;%</span> <span class="fu">kbl</span>() <span class="sc">%&gt;%</span> <span class="fu">kable_styling</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2157</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_1990</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2010</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2163</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2204</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2212</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2216</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2222</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2288</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2303</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2298</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2145</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058021</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058022</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058023</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">reg_6364</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">97</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">85</td>
<td style="text-align: right;">34</td>
<td style="text-align: right;">185</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6365</td>
<td style="text-align: right;">26</td>
<td style="text-align: right;">20</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">58</td>
<td style="text-align: right;">67</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">65</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">43</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">79</td>
<td style="text-align: right;">56</td>
<td style="text-align: right;">216</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6366</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">65</td>
<td style="text-align: right;">44</td>
<td style="text-align: right;">169</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6367</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">132</td>
<td style="text-align: right;">66</td>
<td style="text-align: right;">188</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6368</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">56</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6369</td>
<td style="text-align: right;">184</td>
<td style="text-align: right;">304</td>
<td style="text-align: right;">215</td>
<td style="text-align: right;">661</td>
<td style="text-align: right;">278</td>
<td style="text-align: right;">519</td>
<td style="text-align: right;">466</td>
<td style="text-align: right;">435</td>
<td style="text-align: right;">500</td>
<td style="text-align: right;">267</td>
<td style="text-align: right;">1045</td>
<td style="text-align: right;">297</td>
<td style="text-align: right;">1950</td>
<td style="text-align: right;">1562</td>
<td style="text-align: right;">3860</td>
</tr>
</tbody>
</table>


</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># let's plot the transformed values</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">meanSdPlot</span>(log2dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>It is clear how regions with high mean signal (on the right) are now comparable in terms of standard deviation to regions with lower mean signal (on the left). Normally this plot should look slightly different in terms of density, but remember that we are considering only a subset of the dataset.</p>
</blockquote>
<blockquote class="blockquote">
<p>After the transformation, the genes with the same mean do not have exactly the same standard deviations, but the experiment-wide trend (of variance over mean) has flattened.The genes with variance above the trend will allow us to cluster samples into interesting groups.</p>
</blockquote>
</section>
<section id="assessing-sample-to-sample-relationships" class="level2">
<h2 class="anchored" data-anchor-id="assessing-sample-to-sample-relationships">Assessing Sample-to-sample Relationships</h2>
<p>One way to understand trends in our data and the presence of poor quality or <em>outlier</em> samples is to perform exploratory analyses and visualize the results.</p>
<p>Of particular interest is the presence of technical effects in the experiment, such as <strong>batch effects</strong>.</p>
<p>In R in general, <strong>data visualization</strong> is aided by the presence of many packages which can handle diverse data visualization tasks (from traditional plots to visualizing tabular data through heatmaps). We will use two of these packages, one is <code>ggplot2</code> and the other one is <code>pheatmap</code>.</p>
<section id="hierarchical-clustering" class="level3">
<h3 class="anchored" data-anchor-id="hierarchical-clustering">Hierarchical clustering</h3>
<p>One of the main strategies for checking the consistency of our dataset is to <strong>cluster samples based on their complete H3K27 acetylation profile</strong> (considering all 1490 genomic regions in our dataset). This will allow us to <strong>spot the presence of <em>outliers</em> in the data</strong> and <strong>look for consistent profiles of H3K27ac across biological replicates</strong>, which we expect. Use the code below to <strong>plot a <em>heatmap</em> of normalized (and transformed) count values</strong> for our samples. Since plotting the full count table can be computationally expensive, we might want to subset it to the 200 genomic regions with stronger H3K27ac signal in the dataset.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"pheatmap"</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Take the top 200 genomic regions with the highest mean in the dataset. Returns the index of regions with the highest mean across samples.</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>select <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">rowMeans</span>(log2dds),</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">decreasing=</span><span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">200</span>] <span class="co"># Select number of regions. </span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create another table for annotating the heatmap with colors</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(samples[,<span class="fu">c</span>(<span class="st">"groups"</span>,<span class="st">"Tissue"</span>)])</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Draw the heatmap using the `pheatmap` package</span></span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(log2dds[select,], <span class="at">cluster_rows=</span><span class="cn">FALSE</span>, <span class="at">show_rownames=</span><span class="cn">FALSE</span>,</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols=</span><span class="cn">TRUE</span>, <span class="at">annotation_col=</span>df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>What type of assessment would you make about the consistency of the samples across these top active 200 regions ? Do they cluster (a synonym for similar) based on the biological condition of our interest?</p>
</blockquote>
</section>
<section id="sample-to-sample-distances" class="level3">
<h3 class="anchored" data-anchor-id="sample-to-sample-distances">Sample-to-sample Distances</h3>
<p>Another way to get a sense of the global relationship between samples is to check for <strong>how distant samples are between themselves</strong>. This analysis of <em>pairwise distances</em> looks at the acetylation signal of all 1490 genomic regions in the dataset and determines which samples have a more or less similar or different signal value for each. We expect biologically similar samples to have little difference.</p>
</section>
<section id="pairwise-clustering-using-euclidean-distances" class="level3">
<h3 class="anchored" data-anchor-id="pairwise-clustering-using-euclidean-distances">Pairwise clustering using Euclidean distances</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute distances</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>sampleDists <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(log2dds))</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Organize</span></span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>sampleDistMatrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(sampleDists)</span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>( <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">"Blues"</span>)) )(<span class="dv">255</span>)</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with `pheatmap`</span></span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(sampleDistMatrix,</span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_distance_rows=</span>sampleDists,</span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_distance_cols=</span>sampleDists,</span>
<span id="cb56-15"><a href="#cb56-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">color =</span> colors,</span>
<span id="cb56-16"><a href="#cb56-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> df</span>
<span id="cb56-17"><a href="#cb56-17" aria-hidden="true" tabindex="-1"></a>         )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>ðŸ’¡ What type of assessment would you make about the heatmap you just produced? Is the analysis highlighting biologically relevant differences in the dataset?</p>
</blockquote>
</section>
<section id="using-rcolorbrewer-palettes" class="level3">
<h3 class="anchored" data-anchor-id="using-rcolorbrewer-palettes">Using RColorBrewer palettes</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Draw the color palettes associated to RColorBrewer</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.all</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-58-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co">#display.brewer.pal(n = 8, name = 'RdBu')</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">8</span>, <span class="at">name =</span> <span class="st">'RdBu'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] "#B2182B" "#D6604D" "#F4A582" "#FDDBC7" "#D1E5F0" "#92C5DE" "#4393C3"
[8] "#2166AC"</code></pre>
</div>
</div>
</section>
<section id="dimensionality-reduction-with-principal-component-analysis-pca" class="level3">
<h3 class="anchored" data-anchor-id="dimensionality-reduction-with-principal-component-analysis-pca">Dimensionality reduction with Principal Component Analysis (PCA)</h3>
<p>Another useful approach for <strong>understanding the main variability axes in our data</strong> is to perform <a href="https://en.wikipedia.org/wiki/Principal_component_analysis"><strong>PCA</strong></a> and plot the results. PCA takes our H3K27Ac data and outputs its <strong>principal components</strong>, which encode the <strong><em>main sources of variability in the data</em></strong>. Ideally, <strong>we want the samples to have variability caused by the biological effect of our interest</strong> (in this case the differences between normal colon and tumor organoids), but this might not be the case. By plotting and annotating the points by different <strong><em>covariates</em></strong> (i.e.&nbsp;subject or condition) we are able to understand where the variability comes from and if there is any detectable <a href="https://en.wikipedia.org/wiki/Batch_effect"><strong>batch effect</strong></a>. Use the code below to generate a <em>scatter plot</em> of PCA coordinates and annotate them to understand what causes the variability in the data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate principal components and percentage of variance</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>pcs <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(log2dds, <span class="at">scale =</span> <span class="cn">TRUE</span>)</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>percentVar <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> <span class="fu">summary</span>(pcs)<span class="sc">$</span>importance[<span class="dv">2</span>,])</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>pcaData <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(pcs<span class="sc">$</span>rotation) <span class="sc">%&gt;%</span> <span class="fu">merge</span>(samples, <span class="at">by=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the summary of the pcs object</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pcs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>Importance of components:
                          PC1    PC2     PC3     PC4     PC5     PC6     PC7
Standard deviation     2.9465 1.5096 0.76219 0.72368 0.68918 0.61819 0.59742
Proportion of Variance 0.5788 0.1519 0.03873 0.03491 0.03166 0.02548 0.02379
Cumulative Proportion  0.5788 0.7307 0.76945 0.80437 0.83603 0.86151 0.88530
                           PC8     PC9    PC10    PC11    PC12    PC13    PC14
Standard deviation     0.55961 0.55433 0.50767 0.48353 0.45745 0.44858 0.32425
Proportion of Variance 0.02088 0.02049 0.01718 0.01559 0.01395 0.01342 0.00701
Cumulative Proportion  0.90618 0.92666 0.94385 0.95943 0.97338 0.98680 0.99381
                          PC15
Standard deviation     0.30476
Proportion of Variance 0.00619
Cumulative Proportion  1.00000</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot (this time with ggplot2!!)</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pcaData, <span class="fu">aes</span>(PC1, PC2, <span class="at">color=</span>groups)) <span class="sc">+</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC1: "</span>,percentVar[<span class="dv">1</span>],<span class="st">"% variance"</span>)) <span class="sc">+</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC2: "</span>,percentVar[<span class="dv">2</span>],<span class="st">"% variance"</span>)) <span class="sc">+</span> </span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_linedraw</span>()<span class="sc">+</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">aspect.ratio=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>ðŸ’¡ What can you say about this PCA? Are the samples positioning themselves based on their biological condition? How does this information relate to the other plots we produced above? Can you project both the Patient ID and the groups levels on the PCA space?</p>
</blockquote>
</section>
</section>
<section id="differential-analysis" class="level2">
<h2 class="anchored" data-anchor-id="differential-analysis">Differential analysis</h2>
<p>Now that we have plotted all the main diagnostic information related to the dataset, we can start thinking about testing for differentially active enhancer regions.</p>
<p>The main concept behind it is to <strong>contrast tumor from normal organoids</strong> and check which enhancer regions are predominantly (defined in a <em>statistical</em> sense) enriched for H3K27ac in one condition as opposed to the other. Remember that this is a proxy for <strong>enhancer activity</strong>. We have already given instructions to <code>edgeR</code> on which comparison to perform through the design formula when we created the DGEList object called <code>dds</code>.</p>
<blockquote class="blockquote">
<p>In case your dataset has a source of variability due to <strong>batch effect</strong> or other factors, the design formula is the right place where you can define whether you want to correct for these possible uniteresting differences. For instance, if you found that in your dataset part of the variability is due to the different patients from which the biological replicates come from, but you want to assess other biologically driven differences due to a certain treatment, you could modify your design formula in order to <strong>check differences due to the treatment, while controlling simultaneously for patient-to-patient variability</strong>.</p>
</blockquote>
<center>
<p><img src="../pics/Batch_effect.png" class="img-fluid" width="400"></p>
<p><a href="https://scilifelab.github.io/courses/ngsintro/1905/slides/rnaseq/presentation.html#7">Source</a></p>
</center>
<blockquote class="blockquote">
<p>In this case, the design formula would be something similar to:</p>
</blockquote>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Design formula that takes into account also variability across patients</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="sc">~</span> Patient <span class="sc">+</span> Treatment</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>This design formula works ideally if your dataset is <strong>paired</strong>, that is, there is equal contribution for all the patients to the contrasted condition (i.e., the treatment).</p>
</blockquote>
<p>Our <code>design</code> that we previously built is actually an object of â€œ<strong>matrix</strong>â€ data type, which is similar to a <code>data.frame</code> object.</p>
<blockquote class="blockquote">
<p>The main difference between matrices and data frames is that matrices store only a single class of data, while data frames can consist of many different classes of data. You can also think about matrices as vectors that additionally contain the dimension attribute.</p>
</blockquote>
<p>We can inspect more closely the <code>design</code> matrix:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: right;" data-quarto-table-cell-role="th">(Intercept)</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">sample_groupK_org</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>


</div>
</div>
<p>If we take a look at it we can see how the <strong>second column</strong> of this matrix is the one related to our comparison of interest, in particular to the tumor condition that we want to compare against the <em>reference</em> group of the normal condition.</p>
<blockquote class="blockquote">
<p>The design matrix typically consists of <strong><em>binary</em> values</strong>, where â€˜1â€™ represents the samples of interest (tumor condition), and â€˜0â€™ represents the reference group (normal condition). This binary setup allows us to construct a contrast between the two conditions.</p>
</blockquote>
<section id="the-main-edger-function" class="level3">
<h3 class="anchored" data-anchor-id="the-main-edger-function">The Main <code>edgeR</code> Function</h3>
<p>Letâ€™s perform differential expression analysis with <code>edgeR</code> on our dataset using the main function for the task in the package, <code>glmTest()</code>. Without going into the mathematical details, this function fits a <a href="https://en.wikipedia.org/wiki/Generalized_linear_model"><strong>generalized linear model (GLM)</strong></a> to the data in order to perform inference and <strong>decide which genomic regions have a <em>statistically significant difference</em> in H3K27ac signal</strong>.</p>
<blockquote class="blockquote">
<p>GLMs are an extension of classical linear models to <strong>non-normally distributed data</strong>, and are used to specify probability distributions according to their mean-variance relationship.</p>
</blockquote>
<p>To make statistical inferences, we need to properly account for the biological variability that adds extra noise beyond pure sampling randomness or technical variability. This overdispersion (variance grows faster than the mean or more spread than expected by the mean) can be modeled by a Negative Binomial distribution. Thus, using negative binomial modeling we can compute the probability of seeing a certain count, depending on the expected mean and how much the counts vary between replicates.</p>
<p>We first need to compute <strong>region-wise dispersion estimates</strong> with the function <code>estimateDisp()</code>. We can visually inspect the fit of the dispersion estimates below.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First we fit region-wise dispersion estimates to accomodate the theoretical assumptions of the model</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateDisp</span>(dds, design, <span class="at">robust=</span><span class="cn">TRUE</span>)</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the fitted dispersion values</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotBCV</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-68-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>From the dispersion estimate we can see that we are capturing and modelling efficiently the region-wise dispersion in the dataset which is <strong>intrinsically present due to variation</strong>. This variation is quantified in <code>edgeR</code> with a <strong>BCV</strong> or a <em>B</em>iological <em>C</em>oefficient of <em>V</em>ariation which generally <strong>takes into account both <em>unwanted</em> biological variability (that you can specify in the design) and technical variation</strong>.</p>
<p>Now that we have estimated dispersion of the data, we can start from the <strong>raw count data</strong> to assess statistically significant differences.</p>
<blockquote class="blockquote">
<p>Therefore, we will not use normalized and transformed data, but we know that <code>edgeR</code> will perform internally the normalization when performing the differential testing.</p>
</blockquote>
<p>Given raw counts, NB dispersion(s) and a design matrix, glmFit() fits the negative binomial GLM and produces an object of class DGEGLM with some new components. This DGEGLM object can then be passed to glmLRT() to carry out the likelihood ratio test.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the GLM</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">glmFit</span>(dds, design)</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform differential testing given the design formula we wrote previously</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>lrt <span class="ot">&lt;-</span> <span class="fu">glmLRT</span>(fit, <span class="at">coef=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>The reason why we specified <code>coef=2</code> in the code above is because we are referring to the 2nd column of the design matrix, the one contrasting tumor versus normal samples.</p>
</blockquote>
</section>
</section>
<section id="extracting-and-examining-results" class="level2">
<h2 class="anchored" data-anchor-id="extracting-and-examining-results">Extracting and examining results</h2>
<p>After having used the main edgeR function, we can actively explore the results of the analysis for the comparisons of our interest. We can later filter the results based adjusted <em>P</em>-value used to <strong>accept or reject the null hypothesis</strong> (<span class="math inline">\(H_{0}\)</span>) <strong>of a genomic region NOT being differentially enriched between the two conditions</strong>.</p>
<p>With the code below we can extract a table that we call <code>res</code> which contains the results for every single enhancer, stored in separate rows.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the results</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(lrt<span class="sc">$</span>table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now <strong>check out our results object</strong>, which will be a <code>data.frame</code>, a table.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check out results object</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: right;" data-quarto-table-cell-role="th">logFC</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">logCPM</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">LR</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">PValue</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">reg_6364</td>
<td style="text-align: right;">-1.4943851</td>
<td style="text-align: right;">7.812858</td>
<td style="text-align: right;">3.916181</td>
<td style="text-align: right;">0.0478234</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6365</td>
<td style="text-align: right;">-0.3455046</td>
<td style="text-align: right;">8.651352</td>
<td style="text-align: right;">0.681276</td>
<td style="text-align: right;">0.4091477</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6366</td>
<td style="text-align: right;">-3.6917987</td>
<td style="text-align: right;">7.187410</td>
<td style="text-align: right;">27.092621</td>
<td style="text-align: right;">0.0000002</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6367</td>
<td style="text-align: right;">-3.9379829</td>
<td style="text-align: right;">7.907521</td>
<td style="text-align: right;">40.470868</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6368</td>
<td style="text-align: right;">-1.1175852</td>
<td style="text-align: right;">6.537113</td>
<td style="text-align: right;">3.700392</td>
<td style="text-align: right;">0.0543997</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6369</td>
<td style="text-align: right;">-1.7393659</td>
<td style="text-align: right;">12.442966</td>
<td style="text-align: right;">36.882154</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6370</td>
<td style="text-align: right;">-0.8737474</td>
<td style="text-align: right;">8.181011</td>
<td style="text-align: right;">1.578901</td>
<td style="text-align: right;">0.2089190</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6371</td>
<td style="text-align: right;">-1.7123683</td>
<td style="text-align: right;">12.205290</td>
<td style="text-align: right;">31.240509</td>
<td style="text-align: right;">0.0000000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6372</td>
<td style="text-align: right;">-3.3769355</td>
<td style="text-align: right;">7.095959</td>
<td style="text-align: right;">18.885159</td>
<td style="text-align: right;">0.0000139</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6373</td>
<td style="text-align: right;">-0.9411323</td>
<td style="text-align: right;">10.468915</td>
<td style="text-align: right;">5.458547</td>
<td style="text-align: right;">0.0194728</td>
</tr>
</tbody>
</table>


</div>
</div>
<p>This table shows the <strong>log-fold change</strong> levels of enrichment/depletion of histone mark in our genomic regions. Keep in mind that a <strong>log-fold change of 1 corresponds to a difference in raw gene expression value of 2 times since the log has a base of 2</strong>. So, to recap, all of the enhancers with log-fold change of 1 or more are twice as enriched in H3K27ac in one condition compared to the other and we will later filter regions based on the fold-change value.</p>
<p>Additionally, we can see the <strong>Likelihood Ratio</strong>, (â€œLRâ€) for each region. This is another statistical measure used in <code>edgeR</code> used to assess significance: higher LR values indicate stronger evidence against the null hypothesis of no difference between conditions.</p>
<p>We can additionally <strong>print out a summary of the results of the differential analysis at a <em>P</em>-value &lt; 0.01</strong> by using the following code:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">decideTests</span>(lrt, <span class="at">p.value =</span> <span class="fl">0.01</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>       sample_groupK_org
Down                 149
NotSig              1184
Up                   157</code></pre>
</div>
</div>
<p>Here we can see (1) the type of comparison we are performing (vs the reference, in our case normal colon crypts), (2) the number of genomic regions with significantly higher (â€œUpâ€) signal in tumor organoids and the number of regions with significantly lower signal in tumor organoids (i.e., higher signal in normal condition).</p>
<p>Save the results of the differential analysis</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>res <span class="sc">%&gt;%</span> </span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="at">var=</span><span class="st">'PeakID'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(., <span class="st">'../results/res_chr12.csv'</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualizing-results-with-md-plots" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-results-with-md-plots">Visualizing Results With MD Plots</h3>
<p><strong>MD plots are used to get a sense of the proportions of up- and down-regulated features</strong> between two conditions and the number of counts per million (CPM) of each feature, to check if regions with higher counts are statistically preferred to be also differential.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the MD Plot </span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="co"># abline() enables you to draw straight lines on a plot</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMD</span>(lrt)</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="at">col=</span><span class="st">"gray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-77-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>With the gray line we indicate a fold-change of +/- 1 which, if you recall, stands for an <strong>actual magnitude of change of value 2</strong>.</p>
</section>
<section id="visualizing-results-with-volcano-plots" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-results-with-volcano-plots">Visualizing results with volcano plots</h3>
<p>Results from a differential analysis can actually be visualized in many ways, in order to <strong>emphasize different messages of interest within them</strong>. One popular way is to associate statistical significance (P-value) and magnitude of effect (fold change) related to each region in the dataset for the specific comparison we are evaluating.</p>
<blockquote class="blockquote">
<p>This is also the way results have been visualized in the paper from our study.</p>
</blockquote>
<p>Letâ€™s plot a <em>scatterplot</em>, called â€œvolcanoâ€ for the typical vulcanic shape, to summarize the results, also adding some <strong>thresholds</strong> to P-values and log-fold changes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>volcano_corr <span class="ot">&lt;-</span> res <span class="sc">%&gt;%</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>( <span class="at">threshold=</span><span class="fu">ifelse</span>(logFC <span class="sc">&gt;=</span> <span class="dv">2</span> <span class="sc">&amp;</span> PValue <span class="sc">&lt;</span> <span class="fl">0.01</span>,<span class="st">"A"</span>, </span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">ifelse</span>(logFC <span class="sc">&lt;=</span> <span class="sc">-</span><span class="dv">2</span> <span class="sc">&amp;</span> PValue <span class="sc">&lt;</span> <span class="fl">0.01</span>, <span class="st">"B"</span>, <span class="st">"C"</span>))</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>volc_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(volcano_corr, <span class="fu">aes</span>(<span class="at">x=</span>logFC, <span class="at">y =</span><span class="sc">-</span><span class="fu">log10</span>(PValue), <span class="at">color=</span>threshold)) <span class="sc">+</span></span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.4</span>, <span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a><span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>( <span class="st">"A"</span><span class="ot">=</span><span class="st">"#D90416"</span>,<span class="st">"B"</span><span class="ot">=</span><span class="st">"#033E8C"</span>, <span class="st">"C"</span><span class="ot">=</span><span class="st">"grey"</span>)) <span class="sc">+</span></span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a><span class="fu">xlab</span>(<span class="st">"log2(Fold Change)"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"-log10(adj p-value)"</span>) <span class="sc">+</span></span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb74-11"><a href="#cb74-11" aria-hidden="true" tabindex="-1"></a><span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb74-12"><a href="#cb74-12" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span>, <span class="at">colour=</span><span class="st">"#990000"</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>) <span class="sc">+</span> </span>
<span id="cb74-13"><a href="#cb74-13" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>, <span class="at">colour=</span><span class="st">"#990000"</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>) <span class="sc">+</span> </span>
<span id="cb74-14"><a href="#cb74-14" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">colour=</span><span class="st">"#990000"</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb74-15"><a href="#cb74-15" aria-hidden="true" tabindex="-1"></a><span class="fu">scale_y_continuous</span>(<span class="at">trans =</span> <span class="st">"log1p"</span>)</span>
<span id="cb74-16"><a href="#cb74-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-17"><a href="#cb74-17" aria-hidden="true" tabindex="-1"></a>volc_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-78-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualizing-results-with-heatmaps" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-results-with-heatmaps">Visualizing results with heatmaps</h3>
<p>We can also plot differentially enriched regions in the two conditions of our interest using heatmaps. In this case we select genomic intervals based on their significance (P-value &lt; 0.01) and visualize how their H3K27ac values change across samples just like we have done earlier.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take differential regions</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>diffs <span class="ot">&lt;-</span> <span class="fu">rbind</span>(volcano_corr[volcano_corr<span class="sc">$</span>threshold <span class="sc">==</span> <span class="st">"A"</span>,], volcano_corr[volcano_corr<span class="sc">$</span>threshold <span class="sc">==</span> <span class="st">"B"</span>,]) <span class="sc">%&gt;%</span> <span class="fu">rownames</span>()</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset matrix for regions of interest</span></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>mtx <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dds)[diffs,]</span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create another table for annotating the heatmap with colors</span></span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(samples[,<span class="fu">c</span>(<span class="st">"groups"</span>), <span class="at">drop=</span><span class="cn">FALSE</span>])</span>
<span id="cb75-9"><a href="#cb75-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-10"><a href="#cb75-10" aria-hidden="true" tabindex="-1"></a>diff_heat <span class="ot">&lt;-</span> <span class="fu">pheatmap</span>(mtx, <span class="at">cluster_rows=</span><span class="cn">TRUE</span>, <span class="at">show_rownames=</span><span class="cn">FALSE</span>,<span class="at">cluster_cols=</span><span class="cn">TRUE</span>, <span class="at">annotation_col=</span>df, <span class="at">scale =</span> <span class="st">"row"</span>)</span>
<span id="cb75-11"><a href="#cb75-11" aria-hidden="true" tabindex="-1"></a>diff_heat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-79-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualizing-results-with-violin-and-box-plots" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-results-with-violin-and-box-plots">Visualizing results with violin and box plots</h3>
<p>Last, we can visualize the results by looking at the <strong>distribution of counts</strong> aggregated across conditions. We can plot the counts specifically for genomic regions up-regulated in tumor compared to normal condition and use a violin plot together with a box plot. The box plot highlights the <strong>interquartile range and the median of the data</strong>, while the volcano also show the <strong>kernel probability density of the data at different values</strong>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract up-regulated regions which correspond to all the dots in the right part of the volcano</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>up_regions <span class="ot">&lt;-</span> <span class="fu">rownames</span>(volcano_corr[volcano_corr<span class="sc">$</span>threshold <span class="sc">==</span> <span class="st">"A"</span>,])</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract normalized and log-scaled counts for up-regulated regions</span></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>up_mtx <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dds, <span class="at">log=</span><span class="cn">TRUE</span>, <span class="at">normalized.lib.sizes =</span> <span class="cn">TRUE</span>)[up_regions,]</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge information about samples with the counts in order to group counts based on the condition of interest</span></span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>merged_long_up_mtx <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(<span class="fu">as.data.frame</span>(up_mtx), <span class="at">cols =</span> <span class="fu">everything</span>(), <span class="at">names_to =</span> <span class="st">'id'</span>) <span class="sc">%&gt;%</span> <span class="fu">merge</span>(samples, <span class="at">by.x =</span> <span class="st">'id'</span>, <span class="at">by.y =</span> <span class="dv">0</span>)</span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Make the plot with ggplot2</span></span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>violins <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(merged_long_up_mtx, <span class="fu">aes</span>(groups, value, <span class="at">fill =</span> groups)) <span class="sc">+</span> <span class="fu">geom_violin</span>(<span class="at">trim=</span><span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">fill=</span><span class="st">'white'</span>, <span class="at">width=</span><span class="fl">0.1</span>) <span class="sc">+</span> </span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span> </span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'log CPM'</span>)</span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a>violins</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part1_files/figure-html/unnamed-chunk-80-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="save-dds-object" class="level3">
<h3 class="anchored" data-anchor-id="save-dds-object">Save dds object</h3>
<p>Let us save the object in case we need to retrieve it later on.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the environment</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(dds, <span class="st">'../results/dds_object.rds'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>