<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2025-05-26">

<title>ChIP-seq_analysis_part2 ‚Äì 2025 MBMM</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-de84f8d6bb715db06a919283c2d1e787.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-de2f664832f57b2c83139acc48228f9e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="../../site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="../../site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">2025 MBMM</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../pages/index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../IntroGR/pages/Intro_to_R.html"> 
<span class="menu-text">Intro_to_R</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../IntroGR/pages/GRanges_gtf.html"> 
<span class="menu-text">Working with genomic coordinates</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_principles.html"> 
<span class="menu-text">ChIP-seq principles</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../ChIP-seq_analysis/pages/ChIP-seq_analysis_part1.html"> 
<span class="menu-text">ChIP-seq analysis Part1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../ChIP-seq_analysis/pages/ChIP-seq_analysis_part2.html" aria-current="page"> 
<span class="menu-text">ChIP-seq analysis Part2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../Gviz/pages/Genomic_overviews_with_Gviz.html"> 
<span class="menu-text">Genomic visualization</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../PBMC3k/pages/scRNA-seq_PBMC3k.html"> 
<span class="menu-text">scRNA-seq_PBMC3k</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/paganilab"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/LabPagani"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#objetives" id="toc-objetives" class="nav-link active" data-scroll-target="#objetives">Objetives</a></li>
  <li><a href="#further-downstream-analyses" id="toc-further-downstream-analyses" class="nav-link" data-scroll-target="#further-downstream-analyses">Further downstream analyses</a></li>
  <li><a href="#understanding-recurrence-of-enhancer-regions-across-patients" id="toc-understanding-recurrence-of-enhancer-regions-across-patients" class="nav-link" data-scroll-target="#understanding-recurrence-of-enhancer-regions-across-patients">Understanding recurrence of enhancer regions across patients</a></li>
  <li><a href="#annotating-genomic-regions-to-genes" id="toc-annotating-genomic-regions-to-genes" class="nav-link" data-scroll-target="#annotating-genomic-regions-to-genes">Annotating genomic regions to genes</a>
  <ul class="collapse">
  <li><a href="#enhancer-distribution-across-known-genomic-features" id="toc-enhancer-distribution-across-known-genomic-features" class="nav-link" data-scroll-target="#enhancer-distribution-across-known-genomic-features">Enhancer distribution across known genomic features</a></li>
  </ul></li>
  <li><a href="#gene-ontology-analysis" id="toc-gene-ontology-analysis" class="nav-link" data-scroll-target="#gene-ontology-analysis">Gene Ontology Analysis</a></li>
  <li><a href="#enrichment-analysis-using-gsea" id="toc-enrichment-analysis-using-gsea" class="nav-link" data-scroll-target="#enrichment-analysis-using-gsea">Enrichment analysis using GSEA</a></li>
  <li><a href="#region-annotation-with-chippeakanno" id="toc-region-annotation-with-chippeakanno" class="nav-link" data-scroll-target="#region-annotation-with-chippeakanno">Region annotation with ChIPpeakAnno</a></li>
  <li><a href="#take-home-messages" id="toc-take-home-messages" class="nav-link" data-scroll-target="#take-home-messages">Take-home Messages üè†</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">ChIP-seq_analysis_part2</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 26, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="objetives" class="level2">
<h2 class="anchored" data-anchor-id="objetives">Objetives</h2>
<ul>
<li>Downstream analysis of differential expression data</li>
<li>Perform gene ontology analysis on interesting gene groups</li>
<li>Perform gene set enrichment analysis</li>
<li>Inspecting data with Integrative Genome Viewer</li>
</ul>
</section>
<section id="further-downstream-analyses" class="level2">
<h2 class="anchored" data-anchor-id="further-downstream-analyses">Further downstream analyses</h2>
<p>Once we have our differentially enriched regions, we can perform various downstream analyses to check the <strong>functional aspects</strong> of the group of regions which are up- or down-regulated in our condition of interest.</p>
<section id="loading-tables" class="level4">
<h4 class="anchored" data-anchor-id="loading-tables">Loading tables</h4>
<p>For these analyses, we will need to upload additional tables, that have been generated separately by us in order to speed up some computations and save memory. We introduced the content of these files on the first day, here we‚Äôll do a brief recap.</p>
<ol type="1">
<li><p><code>recurrence_chr12.matrix</code>: this is a table where you can find if any of the regions that constitute the <strong>consensus set of enhancer regions</strong> is identified as enhancers also <strong>in the individual samples</strong>.</p></li>
<li><p><code>Korg_UP_regions_results.txt</code> and <code>Ncr_UP_regions_results.txt</code>: these files store the differential analysis results for the <strong>entire set of enhancers across all the genome</strong> and not only chr12 and, importantly, the regions listed in these files are annotated to gene symbols (we will talk about genomic region annotation later).</p></li>
</ol>
<p>Import these files as R objects</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>recurrence_table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../MBMM25_uploadata//recurrence_chr12.matrix"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>tumor_up_res <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../MBMM25_uploadata/Korg_UP_regions_results.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">quote =</span> <span class="st">""</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>tumor_down_res <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../MBMM25_uploadata/Ncr_UP_regions_results.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">quote =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="understanding-recurrence-of-enhancer-regions-across-patients" class="level2">
<h2 class="anchored" data-anchor-id="understanding-recurrence-of-enhancer-regions-across-patients">Understanding recurrence of enhancer regions across patients</h2>
<p>As the paper presenting our studies reports:</p>
<blockquote class="blockquote">
<p>‚ÄúTo identify common epigenetic blueprints across the organoid library, we looked at the <strong>concordance</strong> of tumor-enriched enhancers in the PDOs (Patient-Derived Organoids).‚Äù</p>
</blockquote>
<p>One of the downstream analyses that has been performed on the tumor-enriched enhancers (i.e.&nbsp;the enhancers with significantly higher H3K27ac signal compared to normal colon tissues) was aimed at understanding if there is a <strong>signature of genomic regions that are highly conserved and consistently active in the entire library of PDOs</strong>, which represent an especially interesting subset of regions being the <strong>most conserved enhancers</strong>.</p>
<p>In order to perform this analysis, we need to start from a binary table that indicates for each genomic region (i.e., the rows), if it was called as enhancer in each separate sample (i.e., the columns). This is our recurrence table:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(recurrence_table)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(recurrence_table, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2157</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_1990</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2010</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2163</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2204</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2212</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2216</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2222</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2288</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2303</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2298</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">SQ_2145</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058021</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058022</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">GSM2058023</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Called_Korg</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Called_Ncr</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">reg_6364</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6365</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6366</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">reg_6367</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">reg_6368</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>


</div>
</div>
<p>In order to get the number of PDO samples and normal samples that share each genomic region in the consensus, we need to calculate the <strong>row-wise sum</strong> of all the ‚Äò1‚Äôs in the table. &gt; We have already performed this step for you: you can see the row-sums in the last two columns of the table, named ‚ÄôCalled_Korg‚Äô and ‚ÄòCalled_Ncr‚Äô respectively for tumor and normal sample counts.</p>
<p>We will first extract the upregulated regions for chr12 using certain thresholds. Then we will retrieve information regarding the recurrence of these regions across all PDOs.</p>
<p>At this point we can use this code to generate a <strong>pie chart</strong> that illustrates the recurrence of the up-regulated regions across samples.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../results/res_chr12.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(res) <span class="ot">&lt;-</span> res<span class="sc">$</span>PeakID</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>           PeakID      logFC    logCPM        LR       PValue
reg_6364 reg_6364 -1.4943851  7.812858  3.916181 4.782339e-02
reg_6365 reg_6365 -0.3455046  8.651352  0.681276 4.091477e-01
reg_6366 reg_6366 -3.6917987  7.187409 27.092621 1.939368e-07
reg_6367 reg_6367 -3.9379829  7.907521 40.470868 1.995695e-10
reg_6368 reg_6368 -1.1175852  6.537113  3.700392 5.439969e-02
reg_6369 reg_6369 -1.7393659 12.442966 36.882154 1.254894e-09</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>up_regions <span class="ot">&lt;-</span> res <span class="sc">%&gt;%</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(logFC <span class="sc">&gt;=</span> <span class="dv">2</span> <span class="sc">&amp;</span> PValue <span class="sc">&lt;</span> <span class="fl">0.01</span>) <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Tabulate information on the proportion of enhancers in each category.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter the recurrence table for the significantly enriched enhancers in tumor samples and</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># select only the columns with the row-sums</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>rec_df <span class="ot">&lt;-</span> recurrence_table[up_regions, ] <span class="sc">%&gt;%</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">contains</span>(<span class="st">"Called"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Create a new column that group regions based on the extent of recurrence across samples</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">case_when</span>(Called_Korg <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) <span class="sc">~</span> <span class="st">"1-4"</span>, Called_Korg <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">5</span><span class="sc">:</span><span class="dv">7</span>) <span class="sc">~</span> <span class="st">"5-7"</span>,</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"8-10"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Aggregate regions based on the group</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">group_by</span>(group) <span class="sc">%&gt;%</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">n =</span> <span class="fu">as.numeric</span>(n)) <span class="sc">%&gt;%</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(<span class="fu">desc</span>(group)) <span class="sc">%&gt;%</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate frequencies for each group of regions</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">prop =</span> n<span class="sc">/</span><span class="fu">sum</span>(n) <span class="sc">*</span> <span class="dv">100</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Compute cumulative sum</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">cumsum =</span> <span class="fu">cumsum</span>(prop)) <span class="sc">%&gt;%</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">ypos =</span> <span class="fu">cumsum</span>(prop) <span class="sc">-</span> <span class="fl">0.5</span> <span class="sc">*</span> prop)  <span class="co"># To position the label at the middle of the slice </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Draw the plot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make plot</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>rec_pie <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(rec_df, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">""</span>, <span class="at">y=</span>prop, <span class="at">fill=</span>group)) <span class="sc">+</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">'identity'</span>, <span class="at">width=</span><span class="dv">1</span>, <span class="at">color=</span><span class="st">'white'</span>) <span class="sc">+</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="st">"y"</span>, <span class="at">start=</span><span class="dv">0</span>, <span class="at">direction=</span><span class="dv">1</span>) <span class="sc">+</span>  <span class="co"># theta=variable to map angle to | start=Offset of starting point from 12 o'clock in radians | direction = clockwise or anti-clockwise.</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span>  <span class="co"># try to plot without it</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">y =</span> ypos, <span class="at">label =</span> group), <span class="at">color =</span> <span class="st">"white"</span>, <span class="at">size=</span><span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">'orange2'</span>,<span class="st">'coral3'</span>,<span class="st">'red4'</span>))</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>rec_pie</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part2_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>A pie chart is a stacked bar chart with polar coordinates (<strong>a 2D coordinate system</strong> where each point is defined by its <strong>distance</strong> from a central point (the pole) and the <strong>angle</strong> it makes with a reference line (the polar axis).</p>
<blockquote class="blockquote">
<p>Try to draw the stacked bar chart alone</p>
</blockquote>
<p>We can print the frequencies as percentages, as we have calculated them in the above code.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>rec_df <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(group, n, prop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table caption-top table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">group</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">n</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">prop</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">8-10</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">20.45455</td>
</tr>
<tr class="even">
<td style="text-align: left;">5-7</td>
<td style="text-align: right;">51</td>
<td style="text-align: right;">28.97727</td>
</tr>
<tr class="odd">
<td style="text-align: left;">1-4</td>
<td style="text-align: right;">89</td>
<td style="text-align: right;">50.56818</td>
</tr>
</tbody>
</table>


</div>
</div>
<p>As we can see from the table, there is a subset of enhancers, 20% of the upregulated enhancers in chromosome 12, which are consistently highly enriched in 8-to-10 tumor organoids compared to the normal colon (10 is the total number of PDOs examined in thatstudy). This <em>core</em> set of regions constitute a particularly interesting signature that likely <strong>drives the regulation of relevant genes in CRC tumor cells</strong>.</p>
</section>
<section id="annotating-genomic-regions-to-genes" class="level2">
<h2 class="anchored" data-anchor-id="annotating-genomic-regions-to-genes">Annotating genomic regions to genes</h2>
<p>One of the key steps when dealing with genomic features is their <strong>annotation</strong> to actual genes. Annotating genomic regions to genes involves the process of <strong>identifying and labeling specific regions of the genome with the genes they correspond to</strong>. This facilitates the process of understanding the genomic context of genes, including their regulation, and function.</p>
<p>One of the most <em>reliable</em> ways to annotate genomic regions to genes is <strong>through experimental techniques</strong> that directly measure interactions or associations between genomic elements and genes. <strong>Chromosome Conformation Capture</strong> (3C) techniques, in particular, can capture physical interactions between distant genomic regions, helping to identify enhancer-promoter interactions and other long-range interactions.</p>
<blockquote class="blockquote">
<p>Annotating genomic regions to genes is a complex task that often requires a combination of experimental data and statistical analysis to establish associations.</p>
</blockquote>
<p>Often, however, it is not easy to retrieve experimental data to annotate genomic regions to genes. In such case, the best strategy is to associate these regions to <strong>nearby genes</strong>. The annotation by <strong>TSS proximity</strong> is an approximation of the reality, but nevertheless serves as a good starting point for subsequent interpretations of the functional implications derived by the annotation.</p>
<blockquote class="blockquote">
<p>For instance, researchers may investigate how regulatory elements affect gene expression or how genetic variants within regulatory regions impact gene function.</p>
</blockquote>
<section id="enhancer-distribution-across-known-genomic-features" class="level3">
<h3 class="anchored" data-anchor-id="enhancer-distribution-across-known-genomic-features">Enhancer distribution across known genomic features</h3>
<p>In our case, the annotation to nearby TSS has already been performed using an external tool called <code>Homer</code>, which enables the annotation of regions to genes, but also the <strong>annotation of their genomic position with respect to relevant genomic features</strong>, like introns, exons, promoter-TSS regions, etc.</p>
<p>We will now check how our upregulated enhancers in tumor organoids are distributed across these genomic features, by plotting another pie chart.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset upregulated regions</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>df_anno <span class="ot">&lt;-</span> tumor_up_res <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(annotation) <span class="sc">%&gt;%</span> <span class="fu">tally</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(annotation<span class="sc">!=</span><span class="st">'NA'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">fraction =</span> n<span class="sc">/</span><span class="fu">sum</span>(n)) <span class="sc">%&gt;%</span>     <span class="co"># Compute percentages</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ymax=</span><span class="fu">cumsum</span>(fraction)) <span class="sc">%&gt;%</span>   <span class="co"># Compute the cumulative percentages (top of each rectangle)</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ymin=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">head</span>(ymax, <span class="at">n=</span><span class="sc">-</span><span class="dv">1</span>))) <span class="co"># Compute the bottom of each rectangle</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>df_anno</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code># A tibble: 8 √ó 5
  annotation       n fraction   ymax   ymin
  &lt;chr&gt;        &lt;int&gt;    &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 3'              40  0.0101  0.0101 0     
2 5'               4  0.00101 0.0111 0.0101
3 Intergenic    2224  0.560   0.571  0.0111
4 TTS             34  0.00856 0.579  0.571 
5 exon            23  0.00579 0.585  0.579 
6 intron        1605  0.404   0.989  0.585 
7 non-coding      23  0.00579 0.995  0.989 
8 promoter-TSS    21  0.00528 1      0.995 </code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make the plot</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>anno_rect <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_anno, <span class="fu">aes</span>(<span class="at">ymax=</span>ymax, <span class="at">ymin=</span>ymin, <span class="at">xmax=</span><span class="dv">4</span>, <span class="at">xmin=</span><span class="dv">3</span>, <span class="at">fill=</span>annotation)) <span class="sc">+</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rect</span>() <span class="sc">+</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="at">theta=</span><span class="st">"y"</span>) <span class="sc">+</span> <span class="co"># Try to remove that to understand how the chart is built initially</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>)) <span class="sc">+</span> <span class="co"># Try to remove that to see how to make a pie chart</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette=</span><span class="dv">4</span>) </span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="co">#theme_void()</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Print plot</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>anno_rect</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part2_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Try to draw the same plot as stacked barplot</p>
<blockquote class="blockquote">
<p>What is the genomic feature that harbors the majority of upregulated enhancer regions?</p>
</blockquote>
</section>
</section>
<section id="gene-ontology-analysis" class="level2">
<h2 class="anchored" data-anchor-id="gene-ontology-analysis">Gene Ontology Analysis</h2>
<p>Now that we have the enhancers annotated to known human genes, we could perform a plethora of analyses in order to get other <strong>biological insights related to gene regulation</strong>. One of these is the Gene Ontology Enrichment analysis. We will try to get a more <em>unsupervised</em> look at what kind of biological processes are captured by the upregulated enhancers in CRC PDOs. We will do this using the <code>gProfiler</code> package in <code>R</code>.</p>
<blockquote class="blockquote">
<p><strong>Gene Ontology</strong> is a standardized system for <strong>annotating genes with terms</strong> describing their biological attributes. These terms are organized into three main categories: <strong>Molecular Function</strong> (the biochemical activity of the gene product), <strong>Biological Process</strong> (the broader biological objectives the gene contributes to), and <strong>Cellular Component</strong> (the location where the gene product is active).</p>
</blockquote>
<p>The <em>enrichment</em> is evaluated in this way: a list of genes is compared against a background set of genes (e.g., all genes in the genome) to identify GO terms that are <strong>significantly overrepresented in the list of interest</strong>.</p>
<blockquote class="blockquote">
<p>Statistical tests, such as Fisher‚Äôs exact test or hypergeometric test, are commonly used to determine whether the observed number of genes associated with a particular GO term in the gene list is significantly higher than expected by chance.</p>
</blockquote>
<p>The output of GO enrichment analysis includes a list of significantly enriched GO terms along with statistical metrics, such as p-values or false discovery rates (FDR).</p>
<blockquote class="blockquote">
<p>This information helps to prioritize genes for further study and provides further context to the experimental results!</p>
</blockquote>
<p>First, we will <strong>create a custom function</strong> that takes as input a list of genes and automatically run the <code>gProfiler</code> function responsible for calulating the enrichment and also creating two plots: one that generally describes the categories of enriched terms, and another one more specific for <strong>enriched pathways from KEGG</strong> (aka, <em>Kyoto Encyclopedia of Genes and Genomes</em>).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Result data.frames will be stored in this object</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>gprof <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># genes: provide a character vector with gene names geneListName: a character identifying the</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># list of genes that will be used to name the data.frame stored in res and to create the pdf</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>gprofiler <span class="ot">&lt;-</span> <span class="cf">function</span>(genes, geneListName) {</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Parameters you might want to change: ordered_query: if the gene list provided is ranked</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># evcodes: if you want to have the gene ids that intersect between the query and the term</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># custom_bg: the gene universe used as a background</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    gostres <span class="ot">&lt;-</span> <span class="fu">gost</span>(<span class="at">query =</span> <span class="fu">unique</span>(<span class="fu">as.character</span>(genes)), <span class="at">organism =</span> <span class="st">"hsapiens"</span>, <span class="at">ordered_query =</span> <span class="cn">FALSE</span>,</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">multi_query =</span> <span class="cn">FALSE</span>, <span class="at">significant =</span> <span class="cn">TRUE</span>, <span class="at">exclude_iea =</span> <span class="cn">FALSE</span>, <span class="at">measure_underrepresentation =</span> <span class="cn">FALSE</span>,</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">evcodes =</span> <span class="cn">TRUE</span>, <span class="at">user_threshold =</span> <span class="fl">0.05</span>, <span class="at">correction_method =</span> <span class="st">"g_SCS"</span>, <span class="at">domain_scope =</span> <span class="st">"annotated"</span>,</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">custom_bg =</span> <span class="cn">NULL</span>, <span class="at">numeric_ns =</span> <span class="st">""</span>, <span class="at">sources =</span> <span class="cn">NULL</span>, <span class="at">as_short_link =</span> <span class="cn">FALSE</span>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Create the overview plot (not interactive)</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>    gostplot <span class="ot">&lt;-</span> <span class="fu">gostplot</span>(gostres, <span class="at">capped =</span> <span class="cn">TRUE</span>, <span class="at">interactive =</span> F)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Keep only useful columns</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>    gp_mod <span class="ot">&lt;-</span> gostres<span class="sc">$</span>result[, <span class="fu">c</span>(<span class="st">"query"</span>, <span class="st">"source"</span>, <span class="st">"term_id"</span>, <span class="st">"term_name"</span>, <span class="st">"p_value"</span>, <span class="st">"query_size"</span>,</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>        <span class="st">"intersection_size"</span>, <span class="st">"term_size"</span>, <span class="st">"effective_domain_size"</span>, <span class="st">"intersection"</span>)]</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>    gp_mod<span class="sc">$</span>query <span class="ot">&lt;-</span> geneListName</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate GeneRatio for the dotplot: number of genes intersecting the term/ total number</span></span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>    <span class="co"># of unique genes provided</span></span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>    gp_mod<span class="sc">$</span>GeneRatio <span class="ot">&lt;-</span> gp_mod<span class="sc">$</span>intersection_size<span class="sc">/</span>gp_mod<span class="sc">$</span>query_size</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Number of genes within the term / number of all unique genes across all terms (universe)</span></span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>    gp_mod<span class="sc">$</span>BgRatio <span class="ot">&lt;-</span> <span class="fu">paste0</span>(gp_mod<span class="sc">$</span>term_size, <span class="st">"/"</span>, gp_mod<span class="sc">$</span>effective_domain_size)</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Rename columns</span></span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(gp_mod) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Cluster"</span>, <span class="st">"Category"</span>, <span class="st">"ID"</span>, <span class="st">"Description"</span>, <span class="st">"p.adjust"</span>, <span class="st">"query_size"</span>, <span class="st">"Intersection_size"</span>,</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a>        <span class="st">"term_size"</span>, <span class="st">"effective_domain_size"</span>, <span class="st">"intersection"</span>, <span class="st">"GeneRatio"</span>, <span class="st">"BgRatio"</span>)</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Save the results data.frame in res</span></span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a>    gprof[[geneListName]] <span class="ot">&lt;&lt;-</span> gp_mod</span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Remove possible duplicate terms for plotting</span></span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a>    gp_mod <span class="sc">%&gt;%</span></span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a>        <span class="fu">group_by</span>(Description) <span class="sc">%&gt;%</span></span>
<span id="cb13-44"><a href="#cb13-44" aria-hidden="true" tabindex="-1"></a>        <span class="fu">filter</span>(<span class="fu">row_number</span>() <span class="sc">==</span> <span class="dv">1</span>)</span>
<span id="cb13-45"><a href="#cb13-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-46"><a href="#cb13-46" aria-hidden="true" tabindex="-1"></a>    <span class="co"># omit_ids &lt;- gp_mod[duplicated(gp_mod$Description), ] omit_ids_list &lt;-</span></span>
<span id="cb13-47"><a href="#cb13-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># omit_ids$Description gp_mod &lt;- gp_mod[!gp_mod$Description %in% omit_ids_list,]</span></span>
<span id="cb13-48"><a href="#cb13-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-49"><a href="#cb13-49" aria-hidden="true" tabindex="-1"></a>    go_table_pathways <span class="ot">&lt;-</span> <span class="fu">filter</span>(gp_mod, Category <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"KEGG"</span>))</span>
<span id="cb13-50"><a href="#cb13-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-51"><a href="#cb13-51" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate negLog P-Value and rank terms based on this value</span></span>
<span id="cb13-52"><a href="#cb13-52" aria-hidden="true" tabindex="-1"></a>    go_table_pathways<span class="sc">$</span>negLogPval <span class="ot">=</span> <span class="sc">-</span><span class="fu">log10</span>(go_table_pathways<span class="sc">$</span>p.adjust)</span>
<span id="cb13-53"><a href="#cb13-53" aria-hidden="true" tabindex="-1"></a>    go_table_pathways <span class="ot">&lt;-</span> go_table_pathways[<span class="fu">order</span>(<span class="sc">-</span>go_table_pathways<span class="sc">$</span>negLogPval), ]</span>
<span id="cb13-54"><a href="#cb13-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-55"><a href="#cb13-55" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Make dot plot</span></span>
<span id="cb13-56"><a href="#cb13-56" aria-hidden="true" tabindex="-1"></a>    dotPlot <span class="ot">&lt;-</span> <span class="fu">arrange</span>(go_table_pathways, negLogPval) <span class="sc">%&gt;%</span></span>
<span id="cb13-57"><a href="#cb13-57" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">Description =</span> <span class="fu">factor</span>(.<span class="sc">$</span>Description, <span class="at">levels =</span> .<span class="sc">$</span>Description)) <span class="sc">%&gt;%</span></span>
<span id="cb13-58"><a href="#cb13-58" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggplot</span>(<span class="fu">aes</span>(negLogPval, Description)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> negLogPval, <span class="at">size =</span> GeneRatio)) <span class="sc">+</span></span>
<span id="cb13-59"><a href="#cb13-59" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_size</span>(<span class="at">range =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">9</span>)) <span class="sc">+</span> <span class="fu">scale_color_gradient</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">high =</span> <span class="st">"red"</span>) <span class="sc">+</span> <span class="fu">theme_light</span>() <span class="sc">+</span></span>
<span id="cb13-60"><a href="#cb13-60" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ylab</span>(<span class="st">"Pathway"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>), <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>),</span>
<span id="cb13-61"><a href="#cb13-61" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">7</span>), <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"bold"</span>))</span>
<span id="cb13-62"><a href="#cb13-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-63"><a href="#cb13-63" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Print on display</span></span>
<span id="cb13-64"><a href="#cb13-64" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">list</span>(gostplot, dotPlot))</span>
<span id="cb13-65"><a href="#cb13-65" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let‚Äôs extract a <em>vector</em> of genes associated to the most recurrent regulatory enhancers up-regulated in tumor samples, and perform the analysis:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract regions with high recurrence (from all chromosomes)</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>all_enh_recur <span class="ot">&lt;-</span> tumor_up_res <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> log2FoldChange <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> Called_Korg <span class="sc">&gt;=</span> <span class="dv">8</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract corresponding genes</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>genes_up_recur <span class="ot">&lt;-</span> all_enh_recur <span class="sc">%&gt;%</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(Gene.Name)) <span class="sc">%&gt;%</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(Gene.Name) <span class="sc">%&gt;%</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can check the number of genes that we have retrieved. We‚Äôll relate this to the number of regions we start with.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">"Number of recurrent enhancers:"</span>, <span class="fu">length</span>(all_enh_recur<span class="sc">$</span>PeakID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] "Number of recurrent enhancers: 687"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">"Number of genes associated to the recurrent enhancers:"</span>, <span class="fu">length</span>(genes_up_recur))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] "Number of genes associated to the recurrent enhancers: 582"</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>üí° Can you make a consideration about the differences between the number of regions and the number of corresponding genes?</p>
</blockquote>
<p>We can now run the function created above to obtain our enriched biological pathways.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load package</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gprofiler2)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Run custom function</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">gprofiler</span>(<span class="at">genes =</span> genes_up_recur, <span class="at">geneListName =</span> <span class="st">"genes_up_recur"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part2_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>
[[2]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part2_files/figure-html/unnamed-chunk-21-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here, interestingly, we find that the two most enriched biological pathways are ‚Äúcolorectal cancer‚Äù and ‚ÄúHippo signalling pathway‚Äù.</p>
<p>From this insight we can start to <strong>generate new hypotheses</strong>, like the one tested in the study about the relevance of YAP/TAZ factors, which are indeed key downstream effectors of the Hippo signalling.</p>
<blockquote class="blockquote">
<p>üí° GO analyses might highlight very interesting patterns and generate hypotheses, but are many times quite hard to interpret depending also on the biological system we are studying.</p>
</blockquote>
<p>What information did we retrieve from the analysis?</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>gprof<span class="sc">$</span>genes_up_recur <span class="sc">%&gt;%</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code> [1] "Cluster"               "Category"              "ID"                   
 [4] "Description"           "p.adjust"              "query_size"           
 [7] "Intersection_size"     "term_size"             "effective_domain_size"
[10] "intersection"          "GeneRatio"             "BgRatio"              </code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(gprof<span class="sc">$</span>genes_up_recur<span class="sc">$</span>Category)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>
CORUM GO:BP GO:CC GO:MF    HP  KEGG  REAC    WP 
    1   109    17     1     1     4     1     3 </code></pre>
</div>
</div>
<p>Can you retrieve the differentially expressed genes that are part of the Hippo signaling pathway?</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] "BMP4,CTNNA2,TGFBR2,GSK3B,WWTR1,AREG,TCF7,FGF1,SERPINE1,WNT16,PPP2R2A,YWHAZ"</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] "JUN,GADD45A,DCC,MAP2K2,TGFBR2,GSK3B,AREG,PIK3R1,TCF7"</code></pre>
</div>
</div>
<p>Run enrichment analysis for regions with recurrence in at least 5/10 tumor samples</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part2_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>
[[2]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part2_files/figure-html/unnamed-chunk-26-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What class is gprof?</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(gprof)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] "list"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What are the dimensions of the elements in gprof?</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(gprof, dim)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>$genes_up_recur
[1] 137  12

$genes_up_recur5
[1] 239  12</code></pre>
</div>
</div>
</section>
<section id="enrichment-analysis-using-gsea" class="level2">
<h2 class="anchored" data-anchor-id="enrichment-analysis-using-gsea">Enrichment analysis using GSEA</h2>
<p>Normally we would perform GSEA using all regions tested for differential analysis, ranked by their padj value x log2FC sign.</p>
<p>Here, we will merge the upregulated and downregulated regions and rank them based by their -log10(p-adjusted value) * the sign of the log2FoldChange, e.g.&nbsp;significantly upregulated genes are at the top of the rank whilst significantly downregulated ones are at the bottom. Given that we are dealing with regions and not genes, it is possible that multiple regions are annotated to the same gene. Thus, duplicate genes need to be removed from the ranking list. In doing so, we will keep the region with the most significant padj value, e.g.&nbsp;max(abs(padj x FC)).</p>
<p>We will retrieve gene sets from the Molecular Signatures Database (MSigDB)</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Rank the genes based on their padj x sign(log2FC)</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>ranked <span class="ot">&lt;-</span> <span class="fu">rbind</span>(tumor_up_res, tumor_down_res) <span class="sc">%&gt;%</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(Gene.Name, log2FoldChange, padjxFC) <span class="sc">%&gt;%</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span>Gene.Name <span class="sc">==</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">abs =</span> <span class="fu">abs</span>(padjxFC)) <span class="sc">%&gt;%</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># arrange(desc(abs)) %&gt;%</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(<span class="fu">desc</span>(padjxFC)) <span class="sc">%&gt;%</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(Gene.Name) <span class="sc">%&gt;%</span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="fu">row_number</span>() <span class="sc">==</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># arrange(desc(padjxFC)) %&gt;% tibble::column_to_rownames('Gene.Name') %&gt;%</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># filter(Gene.Name=='ARHGEF16')</span></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a><span class="fu">pull</span>(padjxFC, Gene.Name)</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ranked)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>INSIG1   KRR1 PHLDA1 ACTL7B  IGFL4   SCG2 
21.093 17.403 16.816 15.909 15.447 14.599 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(ranked)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>    FAM90A26        UBE2K LOC101928269         MRC2       SNAR-I           GC 
     -12.327      -12.395      -13.549      -13.684      -13.886      -14.254 </code></pre>
</div>
</div>
<p>Install and import libraries for fgsea and msigdbr</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(fgsea, <span class="at">quiet =</span> <span class="cn">TRUE</span>)) BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"fgsea"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(msigdbr, <span class="at">quiet =</span> <span class="cn">TRUE</span>)) <span class="fu">install.packages</span>(<span class="st">"msigdbr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Extract the curated gene sets (C2) from the MSigDB database</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(msigdbr)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract specific gene sets from the MSigDB database</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>curated_gsets <span class="ot">&lt;-</span> <span class="fu">msigdbr</span>(<span class="at">species =</span> <span class="st">"human"</span>, <span class="at">category =</span> <span class="st">"C2"</span>, <span class="at">subcategory =</span> <span class="st">"CGP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Inspect the information and gene sets we retrieved.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(curated_gsets)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code> [1] "gene_symbol"        "ncbi_gene"          "ensembl_gene"      
 [4] "db_gene_symbol"     "db_ncbi_gene"       "db_ensembl_gene"   
 [7] "source_gene"        "gs_id"              "gs_name"           
[10] "gs_collection"      "gs_subcollection"   "gs_collection_name"
[13] "gs_description"     "gs_source_species"  "gs_pmid"           
[16] "gs_geoid"           "gs_exact_source"    "gs_url"            
[19] "db_version"         "db_target_species"  "entrez_gene"       
[22] "gs_cat"             "gs_subcat"         </code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>curated_gsets <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(gs_name) <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>() <span class="sc">%&gt;%</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">head</span>(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>[1] "ABBUD_LIF_SIGNALING_1_DN" "ABBUD_LIF_SIGNALING_1_UP"</code></pre>
</div>
</div>
<p>Check if there are any gene sets related to Colon Cancer.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>curated_gsets <span class="sc">%&gt;%</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="fu">grepl</span>(<span class="st">"COLON_AND_RECTAL"</span>, gs_name)) <span class="sc">%&gt;%</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>  gene_symbol ncbi_gene    ensembl_gene db_gene_symbol db_ncbi_gene
1       ACTL8     81569 ENSG00000117148          ACTL8        81569
2    ADAMTSL4     54507 ENSG00000143382       ADAMTSL4        54507
3      AKR7A2      8574 ENSG00000053371         AKR7A2         8574
4        AOC2       314 ENSG00000131480           AOC2          314
5    B3GALNT1      8706 ENSG00000169255       B3GALNT1         8706
6   C14orf132     56967 ENSG00000227051      C14orf132        56967
  db_ensembl_gene source_gene  gs_id                          gs_name
1 ENSG00000117148    LOC81569 M15780 GRADE_COLON_AND_RECTAL_CANCER_DN
2 ENSG00000143382       TSRC1 M15780 GRADE_COLON_AND_RECTAL_CANCER_DN
3 ENSG00000053371      AKR7A2 M15780 GRADE_COLON_AND_RECTAL_CANCER_DN
4 ENSG00000131480        AOC2 M15780 GRADE_COLON_AND_RECTAL_CANCER_DN
5 ENSG00000169255     B3GALT3 M15780 GRADE_COLON_AND_RECTAL_CANCER_DN
6 ENSG00000227051   C14orf132 M15780 GRADE_COLON_AND_RECTAL_CANCER_DN
  gs_collection gs_subcollection                 gs_collection_name
1            C2              CGP Chemical and Genetic Perturbations
2            C2              CGP Chemical and Genetic Perturbations
3            C2              CGP Chemical and Genetic Perturbations
4            C2              CGP Chemical and Genetic Perturbations
5            C2              CGP Chemical and Genetic Perturbations
6            C2              CGP Chemical and Genetic Perturbations
                                                                              gs_description
1 Down-regulated genes in both rectal and colon carcinoma compared to normal mucosa samples.
2 Down-regulated genes in both rectal and colon carcinoma compared to normal mucosa samples.
3 Down-regulated genes in both rectal and colon carcinoma compared to normal mucosa samples.
4 Down-regulated genes in both rectal and colon carcinoma compared to normal mucosa samples.
5 Down-regulated genes in both rectal and colon carcinoma compared to normal mucosa samples.
6 Down-regulated genes in both rectal and colon carcinoma compared to normal mucosa samples.
  gs_source_species  gs_pmid gs_geoid                  gs_exact_source gs_url
1                HS 17210682          Table 5S: Colon &lt; 1 &amp; Rectum &lt; 1       
2                HS 17210682          Table 5S: Colon &lt; 1 &amp; Rectum &lt; 1       
3                HS 17210682          Table 5S: Colon &lt; 1 &amp; Rectum &lt; 1       
4                HS 17210682          Table 5S: Colon &lt; 1 &amp; Rectum &lt; 1       
5                HS 17210682          Table 5S: Colon &lt; 1 &amp; Rectum &lt; 1       
6                HS 17210682          Table 5S: Colon &lt; 1 &amp; Rectum &lt; 1       
  db_version db_target_species entrez_gene gs_cat gs_subcat
1  2024.1.Hs                HS       81569     C2       CGP
2  2024.1.Hs                HS       54507     C2       CGP
3  2024.1.Hs                HS        8574     C2       CGP
4  2024.1.Hs                HS         314     C2       CGP
5  2024.1.Hs                HS        8706     C2       CGP
6  2024.1.Hs                HS       56967     C2       CGP</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter the curated gene sets for the gene set of our interest</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>gene_set_name <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GRADE_COLON_AND_RECTAL_CANCER_UP"</span>, <span class="st">"GRADE_COLON_AND_RECTAL_CANCER_DN"</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>tex_sig_df <span class="ot">&lt;-</span> curated_gsets <span class="sc">%&gt;%</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(gs_name <span class="sc">%in%</span> gene_set_name)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tex_sig_df, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>  gene_symbol ncbi_gene    ensembl_gene db_gene_symbol db_ncbi_gene
1       ACTL8     81569 ENSG00000117148          ACTL8        81569
2    ADAMTSL4     54507 ENSG00000143382       ADAMTSL4        54507
  db_ensembl_gene source_gene  gs_id                          gs_name
1 ENSG00000117148    LOC81569 M15780 GRADE_COLON_AND_RECTAL_CANCER_DN
2 ENSG00000143382       TSRC1 M15780 GRADE_COLON_AND_RECTAL_CANCER_DN
  gs_collection gs_subcollection                 gs_collection_name
1            C2              CGP Chemical and Genetic Perturbations
2            C2              CGP Chemical and Genetic Perturbations
                                                                              gs_description
1 Down-regulated genes in both rectal and colon carcinoma compared to normal mucosa samples.
2 Down-regulated genes in both rectal and colon carcinoma compared to normal mucosa samples.
  gs_source_species  gs_pmid gs_geoid                  gs_exact_source gs_url
1                HS 17210682          Table 5S: Colon &lt; 1 &amp; Rectum &lt; 1       
2                HS 17210682          Table 5S: Colon &lt; 1 &amp; Rectum &lt; 1       
  db_version db_target_species entrez_gene gs_cat gs_subcat
1  2024.1.Hs                HS       81569     C2       CGP
2  2024.1.Hs                HS       54507     C2       CGP</code></pre>
</div>
</div>
<p>We will prepare a gene set list retrieving the genes that are contained in each gene set (shown in the <code>gs_name</code> column).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create an empty gene set list</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>gset <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="co"># For each gene set in the data frame retrieve the gene symbols and store them in the gset</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="co"># list.</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gs <span class="cf">in</span> <span class="fu">unique</span>(tex_sig_df<span class="sc">$</span>gs_name)) {</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># subset the table for a gene set</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>    tmp <span class="ot">&lt;-</span> tex_sig_df <span class="sc">%&gt;%</span></span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">filter</span>(gs_name <span class="sc">==</span> gs)</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># retrieve gene symbols</span></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>    tmp_gset <span class="ot">&lt;-</span> <span class="fu">list</span>(tmp<span class="sc">$</span>gene_symbol)</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># names(tmp_gset) &lt;- gene_set_name</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add the genes in the gset list</span></span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a>    gset[gs] <span class="ot">&lt;-</span> tmp_gset</span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(gset, length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>$GRADE_COLON_AND_RECTAL_CANCER_DN
[1] 102

$GRADE_COLON_AND_RECTAL_CANCER_UP
[1] 291</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>gset</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>$GRADE_COLON_AND_RECTAL_CANCER_DN
  [1] "ACTL8"      "ADAMTSL4"   "AKR7A2"     "AOC2"       "B3GALNT1"  
  [6] "C14orf132"  "C2orf88"    "CA5A"       "CACNA2D2"   "CBS"       
 [11] "CCDC71"     "CCR9"       "CD28"       "CDK12"      "CFD"       
 [16] "CHST15"     "CHST8"      "COL4A5"     "CPN2"       "CX3CR1"    
 [21] "CYB5D1"     "DAP"        "DHPS"       "FABP7"      "FAN1"      
 [26] "FHL1"       "FKRP"       "GFI1"       "GGA1"       "GNA11"     
 [31] "GNPDA1"     "GPC1"       "GPR87"      "GREB1"      "GTF3C4"    
 [36] "HAUS5"      "HIC2"       "HMGXB4"     "ITPKB"      "JAM3"      
 [41] "KRT84"      "LDB2"       "LGALS2"     "LRFN1"      "LRTOMT"    
 [46] "LYPD3"      "MEF2D"      "MITF"       "MSRA"       "MYCN"      
 [51] "NAT8L"      "NAV1"       "NR1H4"      "NR3C1"      "NUDT11"    
 [56] "OSGIN2"     "PARM1"      "PARVB"      "PCDHB5"     "PDK4"      
 [61] "PGC"        "PLAC8"      "PLD1"       "PLPP1"      "PRKACB"    
 [66] "PRRT2"      "PTPRZ1"     "RERE"       "SASH1"      "SCAMP5"    
 [71] "SERTAD4"    "SESN2"      "SIRPB1"     "SIX1"       "SLC16A9"   
 [76] "SLC30A3"    "SLC30A6"    "SMIM14"     "SNCA"       "SOCS2"     
 [81] "SPIN4"      "ST6GALNAC6" "STAR"       "STAT5A"     "SYNC"      
 [86] "TBC1D27P"   "TGFBR3"     "TIMM22"     "TLE4"       "TLN2"      
 [91] "TM9SF1"     "TOX2"       "TRPS1"      "TYRP1"      "UBAP1"     
 [96] "UGDH"       "UNC5B"      "XPNPEP3"    "YY1AP1"     "ZBTB4"     
[101] "ZNF135"     "ZNF787"    

$GRADE_COLON_AND_RECTAL_CANCER_UP
  [1] "AAMP"      "AARS1"     "ABCB8"     "ABCF1"     "ACBD6"     "ADGRG1"   
  [7] "AHCY"      "ALG1"      "AMPD3"     "ANXA5"     "ATIC"      "ATP6AP2"  
 [13] "AURKA"     "AZGP1"     "BACE2"     "BORA"      "BPHL"      "C4BPB"    
 [19] "CAMSAP1"   "CARHSP1"   "CCDC191"   "CCDC85B"   "CCDC92"    "CCNB1"    
 [25] "CCT5"      "CCT7"      "CCT8"      "CD44"      "CD46"      "CDC25B"   
 [31] "CDC45"     "CDC6"      "CDCA7"     "CDK1"      "CDK7"      "CEP57L1"  
 [37] "CFB"       "CHCHD2"    "CKS2"      "CNBP"      "COL4A1"    "COPS8"    
 [43] "CSNK1E"    "CSTF1"     "CTSB"      "DAP3"      "DCAF7"     "DDIT4"    
 [49] "DDX1"      "DDX27"     "DDX39A"    "DDX39B"    "DEF8"      "DGAT2"    
 [55] "DHX8"      "DHX9"      "DNAJA3"    "DNAJC2"    "DNTTIP1"   "DOLK"     
 [61] "DPM1"      "DPY19L1"   "DRG1"      "DUSP18"    "EEF1E1"    "EFTUD2"   
 [67] "EIF2S2"    "EIF3B"     "EIF3E"     "EIF4A1"    "EIF4B"     "EIF6"     
 [73] "EMC8"      "ENG"       "EPHB2"     "EPRS1"     "ERI1"      "ETS2"     
 [79] "ETV4"      "EXOC2"     "FAM210B"   "FBL"       "FBXO11"    "FDFT1"    
 [85] "FERMT1"    "FXYD5"     "GARS1"     "GCLC"      "GCSH"      "GDF15"    
 [91] "GEMIN6"    "GLO1"      "GPSM2"     "GSK3B"     "GTF2F2"    "GTF3A"    
 [97] "GTPBP10"   "GYG2"      "H2AZ1"     "HELLS"     "HILPDA"    "HJURP"    
[103] "HM13"      "HMGB1"     "HMGB2"     "HNRNPA1"   "HNRNPA2B1" "HNRNPD"   
[109] "HNRNPK"    "HOXB5"     "HSPA4"     "HSPA9"     "HSPE1"     "ICA1"     
[115] "IFITM1"    "IFRD1"     "INTS7"     "IRAK1"     "ITGB5"     "KARS1"    
[121] "KBTBD2"    "KIF23"     "KNTC1"     "KPNA3"     "L1CAM"     "LAMP1"    
[127] "LAPTM4B"   "LDHA"      "LDHB"      "LRP8"      "LST1"      "LY6E"     
[133] "LYAR"      "MACROH2A1" "MAGED2"    "MARS1"     "MBTPS2"    "MCM2"     
[139] "MCM3"      "MCM4"      "MCM5"      "MCM7"      "MDC1"      "MELK"     
[145] "METTL2B"   "METTL3"    "METTL5"    "MKLN1"     "MLST8"     "MRGBP"    
[151] "MRPL17"    "MRPL3"     "MRPL9"     "MTHFD1"    "MTHFD2"    "MYC"      
[157] "NAA10"     "NDUFAF7"   "NEBL"      "NIP7"      "NIT2"      "NMI"      
[163] "NOB1"      "NOL8"      "NONO"      "NOP56"     "NPM1"      "NQO2"     
[169] "NSMCE2"    "NUSAP1"    "OLA1"      "OSER1"     "PA2G4"     "PABPC3"   
[175] "PAFAH1B3"  "PCNA"      "PDCD2L"    "PDCD6"     "PDK3"      "PDXK"     
[181] "PERP"      "PFDN2"     "PKM"       "PLAAT3"    "PLAU"      "PLCB4"    
[187] "PLK1"      "PMPCB"     "PNPO"      "PNPT1"     "POFUT1"    "POLD2"    
[193] "POLQ"      "POLR1D"    "POLR2D"    "POMP"      "PPA1"      "PPIA"     
[199] "PRDX1"     "PRDX4"     "PRKDC"     "PSMA1"     "PSMA2"     "PSMA3"    
[205] "PSMB1"     "PSMB3"     "PSMB7"     "PSMD14"    "PSMD4"     "PSMG1"    
[211] "PTCD3"     "PTDSS1"    "PTPN12"    "PXDN"      "RAB4B"     "RAN"      
[217] "RBM26"     "RCC1"      "RCN1"      "RFC2"      "RFC5"      "RHEB"     
[223] "RLIM"      "RMI2"      "RRM1"      "RRM2"      "RTF2"      "RTKN"     
[229] "S100A14"   "SAP18"     "SEPHS1"    "SERPINB5"  "SET"       "SFT2D3"   
[235] "SH3GL1"    "SHISA5"    "SKA3"      "SLC12A2"   "SMPD4"     "SNRPB"    
[241] "SOD2"      "SOX9"      "SPARC"     "SPATA13"   "SPATA13"   "SPP1"     
[247] "SRP72"     "SRSF1"     "SRSF2"     "SSB"       "SSRP1"     "SYAP1"    
[253] "SYNCRIP"   "TALDO1"    "TARS1"     "TBCE"      "TCF7"      "TFB1M"    
[259] "TGDS"      "TGFBI"     "THEM6"     "THOC3"     "THY1"      "TIMELESS" 
[265] "TIMP1"     "TKT"       "TMEM97"    "TMSB4Y"    "TPI1"      "TRAF5"    
[271] "TRAP1"     "TRUB2"     "TTPAL"     "UBD"       "UBE2N"     "UBR4"     
[277] "URI1"      "USP7"      "USP9X"     "UTP11"     "VBP1"      "VEGFA"    
[283] "VKORC1L1"  "WDR74"     "WDR75"     "XRN2"      "ZDHHC9"    "ZFAND1"   
[289] "ZMYND8"    "ZNF84"     "ZNG1B"    </code></pre>
</div>
</div>
<p>Run GSEA on our ranked list of genes using the gene sets we retrieve from MSigDB. Here, we set the number of permutations to 1000.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fgsea)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Run GSEA</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>fgseaRes <span class="ot">&lt;-</span> <span class="fu">fgsea</span>(<span class="at">pathways =</span> gset, <span class="at">stats =</span> ranked, <span class="at">nperm =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Inspect the results</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at results</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>fgseaRes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>                            pathway       pval       padj        ES       NES
                             &lt;char&gt;      &lt;num&gt;      &lt;num&gt;     &lt;num&gt;     &lt;num&gt;
1: GRADE_COLON_AND_RECTAL_CANCER_DN 0.51594747 0.51594747 0.2120043 0.9682577
2: GRADE_COLON_AND_RECTAL_CANCER_UP 0.03208556 0.06417112 0.3032790 1.5517402
   nMoreExtreme  size  leadingEdge
          &lt;num&gt; &lt;int&gt;       &lt;list&gt;
1:          274    41 CHST15, ....
2:           17    66 PLAU, TC....</code></pre>
</div>
</div>
<p>Plot GSEA results</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot GSEA results</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gs <span class="cf">in</span> <span class="fu">unique</span>(gene_set_name)) {</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">plotEnrichment</span>(gset[[gs]], ranked) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> gs))</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part2_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="576"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ChIP-seq_analysis_part2_files/figure-html/unnamed-chunk-40-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>Can you retrieve the leading edge genes from the reuslts?</p>
</blockquote>
</section>
<section id="region-annotation-with-chippeakanno" class="level2">
<h2 class="anchored" data-anchor-id="region-annotation-with-chippeakanno">Region annotation with ChIPpeakAnno</h2>
<p>We will try an alternative approach to annotate our enhancer regions leveraging the BioConductor package ChIPpeakAnno. First we create a GRanges object of the upregulated regions.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>up.gr <span class="ot">&lt;-</span> <span class="fu">makeGRangesFromDataFrame</span>(tumor_up_res, <span class="at">keep.extra.columns =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>up.gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>GRanges object with 3974 ranges and 15 metadata columns:
         seqnames              ranges strand |      PeakID       deseq
            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
     [1]     chr1     3291000-3294200      * |      reg_56    deseq_up
     [2]     chr1     3328800-3331400      * |      reg_58    deseq_up
     [3]     chr1     3494000-3496400      * |      reg_66    deseq_up
     [4]     chr1     3509400-3511000      * |      reg_68    deseq_up
     [5]     chr1     3513400-3515200      * |      reg_69    deseq_up
     ...      ...                 ...    ... .         ...         ...
  [3970]     chrX 151896800-151898400      * |   reg_33047    deseq_up
  [3971]     chrX 151900600-151904600      * |   reg_33048    deseq_up
  [3972]     chrX 154368400-154369200      * |   reg_33071    deseq_up
  [3973]     chrY     9176000-9177600      * |   reg_33097    deseq_up
  [3974]     chrY     9278000-9279800      * |   reg_33099    deseq_up
          baseMean log2FoldChange     lfcSE      stat      pvalue        padj
         &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
     [1]  11.53169        3.68190  0.775310   4.74894 2.04489e-06 4.04021e-05
     [2]  11.22958        1.93821  0.735416   2.63553 8.40071e-03 2.66628e-02
     [3]   8.20774        2.49910  1.033260   2.41865 1.55780e-02 4.27478e-02
     [4]   6.61575        2.98599  0.956157   3.12290 1.79076e-03 8.15651e-03
     [5]   6.80467        4.30521  1.071856   4.01659 5.90454e-05 5.82429e-04
     ...       ...            ...       ...       ...         ...         ...
  [3970]   7.35329        3.39137  0.880766   3.85047 0.000117890 0.000998584
  [3971]  16.29193        2.19479  0.765732   2.86627 0.004153423 0.015505174
  [3972]   4.58607        2.22117  0.882322   2.51741 0.011821996 0.034497587
  [3973]   3.78489        3.39099  1.217671   2.78482 0.005355755 0.018851664
  [3974]   6.76525        4.25332  1.267348   3.35608 0.000790553 0.004332179
           padjxFC                 region  annotation   Gene.Name
         &lt;numeric&gt;            &lt;character&gt; &lt;character&gt; &lt;character&gt;
     [1]     4.394   chr1:3291000-3294200      intron    ARHGEF16
     [2]     1.574   chr1:3328800-3331400      intron    ARHGEF16
     [3]     1.369   chr1:3494000-3496400      intron    ARHGEF16
     [4]     2.088   chr1:3509400-3511000      intron     MIR551A
     [5]     3.235   chr1:3513400-3515200      intron     MIR551A
     ...       ...                    ...         ...         ...
  [3970]     3.001 chrX:151896800-15189..  Intergenic  MAGEA4-AS1
  [3971]     1.810 chrX:151900600-15190..  Intergenic  MAGEA4-AS1
  [3972]     1.462 chrX:154368400-15436..      intron        FLNA
  [3973]     1.725   chrY:9176000-9177600  Intergenic    RBMY1A3P
  [3974]     2.363   chrY:9278000-9279800  Intergenic    RBMY1A3P
              Gene.Type Called_Korg Called_Ncr
            &lt;character&gt;   &lt;integer&gt;  &lt;integer&gt;
     [1] protein-coding           4          0
     [2] protein-coding           4          0
     [3] protein-coding           2          0
     [4]          ncRNA           2          0
     [5]          ncRNA           2          0
     ...            ...         ...        ...
  [3970]          ncRNA           3          0
  [3971]          ncRNA           5          0
  [3972] protein-coding           2          0
  [3973]         pseudo           2          0
  [3974]         pseudo           3          0
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the library</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="co"># if (!require(ChIPpeakAnno)) BiocManager::install('ChIPpeakAnno')</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ChIPpeakAnno)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To perform the annotation we need to incorporate information on gene structure and positions. We will import the annotation package for human assembly hg38 version 86 from Ensembl.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use Ensembl annotation package: if (!require(EnsDb.Hsapiens.v86))</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install('EnsDb.Hsapiens.v86')</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnsDb.Hsapiens.v86)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ensembl.hs86.transcript &lt;- transcripts(EnsDb.Hsapiens.v86)</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>ensembl.hs86.gene <span class="ot">&lt;-</span> <span class="fu">genes</span>(EnsDb.Hsapiens.v86)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use Ensembl annotation package:</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>up_peak_ensembl <span class="ot">&lt;-</span> <span class="fu">annotatePeakInBatch</span>(up.gr, <span class="at">AnnotationData =</span> ensembl.hs86.gene)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(up_peak_ensembl, <span class="at">n =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>GRanges object with 2 ranges and 24 metadata columns:
                        seqnames          ranges strand |      PeakID
                           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
  X0001.ENSG00000272235     chr1 3291000-3294200      * |      reg_56
  X0002.ENSG00000272235     chr1 3328800-3331400      * |      reg_58
                              deseq  baseMean log2FoldChange     lfcSE
                        &lt;character&gt; &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;
  X0001.ENSG00000272235    deseq_up   11.5317        3.68190  0.775310
  X0002.ENSG00000272235    deseq_up   11.2296        1.93821  0.735416
                             stat      pvalue        padj   padjxFC
                        &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;
  X0001.ENSG00000272235   4.74894 2.04489e-06 4.04021e-05     4.394
  X0002.ENSG00000272235   2.63553 8.40071e-03 2.66628e-02     1.574
                                      region  annotation   Gene.Name
                                 &lt;character&gt; &lt;character&gt; &lt;character&gt;
  X0001.ENSG00000272235 chr1:3291000-3294200      intron    ARHGEF16
  X0002.ENSG00000272235 chr1:3328800-3331400      intron    ARHGEF16
                             Gene.Type Called_Korg Called_Ncr        peak
                           &lt;character&gt;   &lt;integer&gt;  &lt;integer&gt; &lt;character&gt;
  X0001.ENSG00000272235 protein-coding           4          0        0001
  X0002.ENSG00000272235 protein-coding           4          0        0002
                                feature start_position end_position
                            &lt;character&gt;      &lt;integer&gt;    &lt;integer&gt;
  X0001.ENSG00000272235 ENSG00000272235        3306636      3310096
  X0002.ENSG00000272235 ENSG00000272235        3306636      3310096
                        feature_strand insideFeature distancetoFeature
                           &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
  X0001.ENSG00000272235              -    downstream             19096
  X0002.ENSG00000272235              -      upstream            -18704
                        shortestDistance fromOverlappingOrNearest
                               &lt;integer&gt;              &lt;character&gt;
  X0001.ENSG00000272235            12436          NearestLocation
  X0002.ENSG00000272235            18704          NearestLocation
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>We can now annotate the Ensembl IDs with the corresponding gene symbols. Here, we will use the org.Hs.eg.db database.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"org.Hs.eg.db"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"org.Hs.eg.db"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Use Ensembl annotation package:</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>up_peak_ensembl <span class="ot">&lt;-</span> <span class="fu">addGeneIDs</span>(<span class="at">annotatedPeak =</span> up_peak_ensembl, </span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">orgAnn =</span> <span class="st">"org.Hs.eg.db"</span>,</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>                                <span class="co">#mart = mart,</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>                                <span class="at">feature_id_type =</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>                                <span class="at">IDs2Add =</span> <span class="st">"symbol"</span>)</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(up_peak_ensembl, <span class="at">n =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>GRanges object with 2 ranges and 25 metadata columns:
                        seqnames          ranges strand |      PeakID
                           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
  X0001.ENSG00000272235     chr1 3291000-3294200      * |      reg_56
  X0002.ENSG00000272235     chr1 3328800-3331400      * |      reg_58
                              deseq  baseMean log2FoldChange     lfcSE
                        &lt;character&gt; &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;
  X0001.ENSG00000272235    deseq_up   11.5317        3.68190  0.775310
  X0002.ENSG00000272235    deseq_up   11.2296        1.93821  0.735416
                             stat      pvalue        padj   padjxFC
                        &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;
  X0001.ENSG00000272235   4.74894 2.04489e-06 4.04021e-05     4.394
  X0002.ENSG00000272235   2.63553 8.40071e-03 2.66628e-02     1.574
                                      region  annotation   Gene.Name
                                 &lt;character&gt; &lt;character&gt; &lt;character&gt;
  X0001.ENSG00000272235 chr1:3291000-3294200      intron    ARHGEF16
  X0002.ENSG00000272235 chr1:3328800-3331400      intron    ARHGEF16
                             Gene.Type Called_Korg Called_Ncr        peak
                           &lt;character&gt;   &lt;integer&gt;  &lt;integer&gt; &lt;character&gt;
  X0001.ENSG00000272235 protein-coding           4          0        0001
  X0002.ENSG00000272235 protein-coding           4          0        0002
                                feature start_position end_position
                            &lt;character&gt;      &lt;integer&gt;    &lt;integer&gt;
  X0001.ENSG00000272235 ENSG00000272235        3306636      3310096
  X0002.ENSG00000272235 ENSG00000272235        3306636      3310096
                        feature_strand insideFeature distancetoFeature
                           &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
  X0001.ENSG00000272235              -    downstream             19096
  X0002.ENSG00000272235              -      upstream            -18704
                        shortestDistance fromOverlappingOrNearest      symbol
                               &lt;integer&gt;              &lt;character&gt; &lt;character&gt;
  X0001.ENSG00000272235            12436          NearestLocation        &lt;NA&gt;
  X0002.ENSG00000272235            18704          NearestLocation        &lt;NA&gt;
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Compare the annotation we got from HOMER (<code>Gene.Name</code> column) to the annotation from ChIPpeakAnno (<code>symbol</code> column). You can generate a data.frame object for better visualization.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>up_peak_ensembl <span class="sc">%&gt;%</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(region, annotation, Gene.Name, symbol)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="bordered"><code>GRanges object with 3990 ranges and 4 metadata columns:
                        seqnames              ranges strand |
                           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |
  X0001.ENSG00000272235     chr1     3291000-3294200      * |
  X0002.ENSG00000272235     chr1     3328800-3331400      * |
  X0003.ENSG00000272088     chr1     3494000-3496400      * |
  X0004.ENSG00000272088     chr1     3509400-3511000      * |
  X0005.ENSG00000272088     chr1     3513400-3515200      * |
                    ...      ...                 ...    ... .
  X3970.ENSG00000147381     chrX 151896800-151898400      * |
  X3971.ENSG00000147381     chrX 151900600-151904600      * |
  X3972.ENSG00000196924     chrX 154368400-154369200      * |
  X3973.ENSG00000224408     chrY     9176000-9177600      * |
  X3974.ENSG00000197038     chrY     9278000-9279800      * |
                                        region  annotation   Gene.Name
                                   &lt;character&gt; &lt;character&gt; &lt;character&gt;
  X0001.ENSG00000272235   chr1:3291000-3294200      intron    ARHGEF16
  X0002.ENSG00000272235   chr1:3328800-3331400      intron    ARHGEF16
  X0003.ENSG00000272088   chr1:3494000-3496400      intron    ARHGEF16
  X0004.ENSG00000272088   chr1:3509400-3511000      intron     MIR551A
  X0005.ENSG00000272088   chr1:3513400-3515200      intron     MIR551A
                    ...                    ...         ...         ...
  X3970.ENSG00000147381 chrX:151896800-15189..  Intergenic  MAGEA4-AS1
  X3971.ENSG00000147381 chrX:151900600-15190..  Intergenic  MAGEA4-AS1
  X3972.ENSG00000196924 chrX:154368400-15436..      intron        FLNA
  X3973.ENSG00000224408   chrY:9176000-9177600  Intergenic    RBMY1A3P
  X3974.ENSG00000197038   chrY:9278000-9279800  Intergenic    RBMY1A3P
                             symbol
                        &lt;character&gt;
  X0001.ENSG00000272235        &lt;NA&gt;
  X0002.ENSG00000272235        &lt;NA&gt;
  X0003.ENSG00000272088        &lt;NA&gt;
  X0004.ENSG00000272088        &lt;NA&gt;
  X0005.ENSG00000272088        &lt;NA&gt;
                    ...         ...
  X3970.ENSG00000147381      MAGEA4
  X3971.ENSG00000147381      MAGEA4
  X3972.ENSG00000196924        FLNA
  X3973.ENSG00000224408        &lt;NA&gt;
  X3974.ENSG00000197038        &lt;NA&gt;
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>Why do you think there are differences in the annotation?</p>
</blockquote>
</section>
<section id="take-home-messages" class="level2">
<h2 class="anchored" data-anchor-id="take-home-messages">Take-home Messages üè†</h2>
<p><strong>Congratulations!</strong> You got the end of the course and now hopefully know some crucial aspects of a ChIP-seq analysis workflow! Some of the key concepts that we have explored during the course can enable us to reach some distilled points of interest:</p>
<ul>
<li><p>Design your experiments carefully with data analysis in mind!</p></li>
<li><p>Data needs to be carefully explored to avoid systematic errors in the analyses!</p></li>
<li><p>Plot and Visualize as much as possible!</p></li>
<li><p>Not all information is useful, remember that it all depends on the biological question!</p></li>
<li><p>Omics outputs are immensely rich and one experiment can be used to answer a plethora of questions!</p></li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>